// @mistralai/mistralai/src/models/operations/agentsapiv1conversationsdelete.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type AgentsApiV1ConversationsDeleteRequest = {
  /**
   * ID of the conversation from which we are fetching metadata.
   */
  conversationId: string;
};

/** @internal */
export const AgentsApiV1ConversationsDeleteRequest$inboundSchema: z.ZodType<
  AgentsApiV1ConversationsDeleteRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  conversation_id: z.string(),
}).transform((v) => {
  return remap$(v, {
    "conversation_id": "conversationId",
  });
});

/** @internal */
export type AgentsApiV1ConversationsDeleteRequest$Outbound = {
  conversation_id: string;
};

/** @internal */
export const AgentsApiV1ConversationsDeleteRequest$outboundSchema: z.ZodType<
  AgentsApiV1ConversationsDeleteRequest$Outbound,
  z.ZodTypeDef,
  AgentsApiV1ConversationsDeleteRequest
> = z.object({
  conversationId: z.string(),
}).transform((v) => {
  return remap$(v, {
    conversationId: "conversation_id",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace AgentsApiV1ConversationsDeleteRequest$ {
  /** @deprecated use `AgentsApiV1ConversationsDeleteRequest$inboundSchema` instead. */
  export const inboundSchema =
    AgentsApiV1ConversationsDeleteRequest$inboundSchema;
  /** @deprecated use `AgentsApiV1ConversationsDeleteRequest$outboundSchema` instead. */
  export const outboundSchema =
    AgentsApiV1ConversationsDeleteRequest$outboundSchema;
  /** @deprecated use `AgentsApiV1ConversationsDeleteRequest$Outbound` instead. */
  export type Outbound = AgentsApiV1ConversationsDeleteRequest$Outbound;
}

export function agentsApiV1ConversationsDeleteRequestToJSON(
  agentsApiV1ConversationsDeleteRequest: AgentsApiV1ConversationsDeleteRequest,
): string {
  return JSON.stringify(
    AgentsApiV1ConversationsDeleteRequest$outboundSchema.parse(
      agentsApiV1ConversationsDeleteRequest,
    ),
  );
}

export function agentsApiV1ConversationsDeleteRequestFromJSON(
  jsonString: string,
): SafeParseResult<AgentsApiV1ConversationsDeleteRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) =>
      AgentsApiV1ConversationsDeleteRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'AgentsApiV1ConversationsDeleteRequest' from JSON`,
  );
}
