// @cloudbase/agent-tools/dist/browser/browser_tool_unified.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnifiedBrowserTool = void 0;
exports.createLocalBrowserTool = createLocalBrowserTool;
exports.createSandboxBrowserTool = createSandboxBrowserTool;
const v4_1 = require("zod/v4");
const utils_1 = require("../utils");
const browser_client_unified_1 = require("./browser_client_unified");
/**
 * Browser tool input schema
 */
const BrowserToolInputSchema = v4_1.z.object({
    /**
     * The name of the Playwright MCP tool to call
     * Examples: browser_navigate, browser_click, browser_snapshot, etc.
     */
    tool: v4_1.z.string(),
    /**
     * Arguments to pass to the tool
     */
    arguments: v4_1.z.record(v4_1.z.string(), v4_1.z.any()).optional(),
});
/**
 * Unified Browser Tool
 *
 * Supports both local and sandbox modes:
 * - Local mode: Uses local Playwright MCP server
 * - Sandbox mode: Connects to E2B sandbox browser via CDP
 */
class UnifiedBrowserTool extends utils_1.BaseTool {
    constructor(config) {
        super({
            name: "browser",
            description: `Browser automation tool (${config.mode} mode) - supports all standard Playwright MCP tools like browser_navigate, browser_click, browser_snapshot, etc.`,
            schema: BrowserToolInputSchema,
        });
        this.config = config;
        this.client = new browser_client_unified_1.UnifiedBrowserClient(config);
    }
    async _invoke(input) {
        try {
            // Ensure client is connected
            if (!this.client.isReady()) {
                await this.client.start();
            }
            // Call the Playwright MCP tool
            const mcpCall = {
                name: input.tool,
                arguments: input.arguments || {},
            };
            const result = await this.client.callTool(mcpCall);
            return new utils_1.ToolResult({
                success: !result.isError,
                data: {
                    success: !result.isError,
                    tool: input.tool,
                    content: result.content,
                    error: result.isError ? "Tool execution failed" : undefined,
                },
            });
        }
        catch (error) {
            return new utils_1.ToolResult({
                success: false,
                error: error instanceof Error ? error.message : String(error),
                data: {
                    success: false,
                    tool: input.tool,
                    error: error instanceof Error ? error.message : String(error),
                },
            });
        }
    }
    /**
     * List all available Playwright MCP tools
     */
    async listAvailableTools() {
        if (!this.client.isReady()) {
            await this.client.start();
        }
        return await this.client.listTools();
    }
    // Convenience methods for common operations
    async navigate(url) {
        return this.invoke({
            tool: "browser_navigate",
            arguments: { url },
        });
    }
    async snapshot() {
        return this.invoke({
            tool: "browser_snapshot",
        });
    }
    async click(element, ref) {
        return this.invoke({
            tool: "browser_click",
            arguments: { element, ref },
        });
    }
    async type(element, ref, text, options) {
        return this.invoke({
            tool: "browser_type",
            arguments: { element, ref, text, ...options },
        });
    }
    async screenshot(options) {
        return this.invoke({
            tool: "browser_take_screenshot",
            arguments: options || {},
        });
    }
    async waitFor(options) {
        return this.invoke({
            tool: "browser_wait_for",
            arguments: options,
        });
    }
    async manageTabs(action, index) {
        return this.invoke({
            tool: "browser_tabs",
            arguments: { action, index },
        });
    }
    async fillForm(fields) {
        return this.invoke({
            tool: "browser_fill_form",
            arguments: { fields },
        });
    }
    // Mode-specific methods
    /**
     * Get VNC URL (sandbox mode only)
     */
    getVncUrl() {
        return this.client.getVncUrl();
    }
    /**
     * Get CDP URL (sandbox mode only)
     */
    getCdpUrl() {
        return this.client.getCdpUrl();
    }
    /**
     * Get connection status
     */
    getStatus() {
        return {
            isReady: this.client.isReady(),
            mode: this.config.mode,
            config: this.config,
        };
    }
    /**
     * Close the client connection
     */
    async close() {
        await this.client.stop();
    }
}
exports.UnifiedBrowserTool = UnifiedBrowserTool;
// Export factory functions for convenience
/**
 * Create a local browser tool
 */
function createLocalBrowserTool(config = {}) {
    return new UnifiedBrowserTool({
        ...config,
        mode: "local",
    });
}
/**
 * Create a sandbox browser tool
 */
function createSandboxBrowserTool(config) {
    return new UnifiedBrowserTool({
        ...config,
        mode: "sandbox",
    });
}
