// @cloudbase/agent-tools/dist/fs/operator/local-operator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalFileOperator = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const base_operator_1 = require("./base-operator");
class LocalFileOperator extends base_operator_1.BaseFileOperator {
    constructor(config) {
        super();
        this.workspaceRoot =
            path_1.default.resolve(config?.workspaceRoot || "") || process.cwd();
    }
    resolvePath(filePath) {
        const pathStr = filePath.toString();
        if (path_1.default.isAbsolute(pathStr)) {
            return pathStr;
        }
        if (this.workspaceRoot) {
            return path_1.default.resolve(this.workspaceRoot, pathStr);
        }
        return path_1.default.resolve(pathStr);
    }
    async access(path) {
        return fs_1.default.promises.access(this.resolvePath(path));
    }
    async readFile(path, encoding) {
        const resolvedPath = this.resolvePath(path);
        if (encoding) {
            return fs_1.default.promises.readFile(resolvedPath, encoding);
        }
        return fs_1.default.promises.readFile(resolvedPath);
    }
    async mkdir(path, options) {
        return fs_1.default.promises.mkdir(this.resolvePath(path), options);
    }
    async writeFile(path, data, options) {
        return fs_1.default.promises.writeFile(this.resolvePath(path), data, options);
    }
    async readdir(path, options) {
        return fs_1.default.promises.readdir(this.resolvePath(path), options);
    }
    async stat(path) {
        return fs_1.default.promises.stat(this.resolvePath(path));
    }
    async exists(path) {
        return fs_1.default.existsSync(this.resolvePath(path));
    }
    async unlink(path) {
        return fs_1.default.promises.unlink(this.resolvePath(path));
    }
    async rmdir(path) {
        return fs_1.default.promises.rmdir(this.resolvePath(path));
    }
    async rename(oldPath, newPath) {
        return fs_1.default.promises.rename(this.resolvePath(oldPath), this.resolvePath(newPath));
    }
}
exports.LocalFileOperator = LocalFileOperator;
