// @mistralai/mistralai/models/components/codeinterpretertool.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeInterpreterTool$outboundSchema = exports.CodeInterpreterTool$inboundSchema = exports.CodeInterpreterToolType$outboundSchema = exports.CodeInterpreterToolType$inboundSchema = exports.CodeInterpreterToolType = void 0;
exports.codeInterpreterToolToJSON = codeInterpreterToolToJSON;
exports.codeInterpreterToolFromJSON = codeInterpreterToolFromJSON;
const z = __importStar(require("zod/v3"));
const schemas_js_1 = require("../../lib/schemas.js");
exports.CodeInterpreterToolType = {
    CodeInterpreter: "code_interpreter",
};
/** @internal */
exports.CodeInterpreterToolType$inboundSchema = z.nativeEnum(exports.CodeInterpreterToolType);
/** @internal */
exports.CodeInterpreterToolType$outboundSchema = exports.CodeInterpreterToolType$inboundSchema;
/** @internal */
exports.CodeInterpreterTool$inboundSchema = z.object({
    type: exports.CodeInterpreterToolType$inboundSchema.default("code_interpreter"),
});
/** @internal */
exports.CodeInterpreterTool$outboundSchema = z.object({
    type: exports.CodeInterpreterToolType$outboundSchema.default("code_interpreter"),
});
function codeInterpreterToolToJSON(codeInterpreterTool) {
    return JSON.stringify(exports.CodeInterpreterTool$outboundSchema.parse(codeInterpreterTool));
}
function codeInterpreterToolFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.CodeInterpreterTool$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'CodeInterpreterTool' from JSON`);
}
