// @cloudbase/ai/dist/esm/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { AI } from './AI';
import { Bot } from './bot';
import * as models from './models';
import * as utils from './utils';
var COMPONENT_NAME = 'ai';
var MODELS = models.MODELS, restModels = __rest(models, ["MODELS"]);
var createRequestInstanceWithAccessTokenHeader = function (_a) {
    var getAccessToken = _a.getAccessToken, req = _a.req;
    var fetch = function (options) { return __awaiter(void 0, void 0, void 0, function () {
        var token, _a, headers, restOptions, accessToken;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (typeof req.fetch !== 'function') {
                        throw new Error('req.fetch is not a function');
                    }
                    token = options.token, _a = options.headers, headers = _a === void 0 ? {} : _a, restOptions = __rest(options, ["token", "headers"]);
                    if (!(token != null)) return [3, 1];
                    accessToken = token;
                    return [3, 3];
                case 1: return [4, getAccessToken()];
                case 2:
                    accessToken = (_b.sent()).accessToken;
                    _b.label = 3;
                case 3: return [2, req.fetch(__assign({ headers: __assign({ Authorization: "Bearer ".concat(accessToken) }, headers) }, restOptions))];
            }
        });
    }); };
    return {
        download: req.download,
        post: req.post,
        upload: req.upload,
        fetch: fetch,
    };
};
function createAI(_a) {
    var env = _a.env, baseUrl = _a.baseUrl, req = _a.req, getAccessToken = _a.getAccessToken, handleReqInstance = _a.handleReqInstance, i18n = _a.i18n;
    var getBaseUrl = function () {
        if (baseUrl != null) {
            return baseUrl;
        }
        if (env == null) {
            throw new Error('`env` is required when `baseUrl` is not provided!');
        }
        return "https://".concat(env, ".api.tcloudbasegateway.com/v1");
    };
    var getReq = function () {
        if (handleReqInstance == null) {
            if (getAccessToken == null) {
                throw new Error('`getAccessToken` is required when `handleReqInstance` is not provided!');
            }
            return createRequestInstanceWithAccessTokenHeader({ req: req, getAccessToken: getAccessToken });
        }
        return handleReqInstance({ req: req });
    };
    return new AI(getReq(), getBaseUrl(), i18n);
}
var createAi = createAI;
function ai(options) {
    var _this = this;
    var _a;
    var req = this.request;
    if (req.fetch == null) {
        throw new Error('cloudbase.request.fetch() unimplemented!');
    }
    var getUrlFromCloud = function () {
        var _a = _this.getEndPointWithKey('GATEWAY'), BASE_URL = _a.BASE_URL, PROTOCOL = _a.PROTOCOL;
        return "".concat(PROTOCOL).concat(BASE_URL);
    };
    var baseUrl = (_a = options === null || options === void 0 ? void 0 : options.baseUrl) !== null && _a !== void 0 ? _a : getUrlFromCloud();
    return createAI({ req: req, baseUrl: baseUrl, handleReqInstance: function (_a) {
            var req = _a.req;
            return req;
        }, i18n: this.config.i18n });
}
var component = {
    name: COMPONENT_NAME,
    entity: { ai: ai },
};
function registerAi(app) {
    try {
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
export { AI, Bot, createAi, createAI, restModels as models, registerAi, utils };
export * from './type';
try {
    window.registerAi = registerAi;
}
catch (e) { }
