// @cloudbase/agent-tools/dist/fs/multi-read-tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.multiRead = multiRead;
exports.createMultiReadTool = createMultiReadTool;
const zod_1 = require("zod");
const utils_1 = require("../utils");
const read_tool_1 = require("./read-tool");
/**
 * Read multiple files concurrently
 */
async function multiRead(context, params) {
    const { reads = [] } = params;
    const startTime = Date.now();
    try {
        // Process all files concurrently
        const promises = reads.map(async (fileReadRequest) => {
            try {
                const result = await (0, read_tool_1.read)(context, fileReadRequest);
                return {
                    file_path: fileReadRequest.file_path,
                    ...result,
                };
            }
            catch (error) {
                return {
                    file_path: fileReadRequest.file_path,
                    success: false,
                    error: error instanceof Error ? error.message : String(error),
                };
            }
        });
        const results = await Promise.all(promises);
        const successfulCount = results.filter((r) => r.success).length;
        const failedCount = results.length - successfulCount;
        const duration = Date.now() - startTime;
        const summary = `Successfully read ${successfulCount} file(s), failed ${failedCount} file(s) out of ${reads.length} total in ${duration}ms`;
        context.messageHandler?.currentText.append(`\n[multi-read] Completed: ${summary}\n`);
        const multipleFilesResult = {
            files: results,
            total: reads.length,
            successfulCount,
            failedCount,
        };
        return new utils_1.ToolResult({
            success: true,
            data: multipleFilesResult,
        });
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        context.messageHandler?.currentText.append(`\n[multi-read] Failed: ${errorMessage}\n`);
        return (0, utils_1.handleToolError)(error, "Multi-read tool execution", "execution");
    }
}
/**
 * Create multi-read tool with execution context
 */
function createMultiReadTool(context) {
    return (0, utils_1.tool)((params) => multiRead(context, params), {
        name: "MultiRead",
        description: `Reads multiple files concurrently.
**Usage:**
  - Provide an array of file paths to read
  - All files are processed concurrently
  - Each file can have its own offset and limit for pagination
  - For large files, use pagination:
    * First read: offset=0, limit=50 (reads lines 1-50)
    * Second read: offset=50, limit=50 (reads lines 51-100)
    * Default: offset=0, limit=1000
    * IMPORTANT: Do NOT read the same file multiple times with the same offset. Always increment offset to read next page.
  - Returns detailed results for each file including success/failure status`,
        schema: zod_1.z.object({
            reads: zod_1.z.array(read_tool_1.readParametersSchema),
        }),
        getDisplay: ({ name, input }) => {
            return `> Using ${name} for ${input.reads.length} files`;
        },
    });
}
