// @cloudbase/ai/src/models/HunYuan/type.ts
type HunYuanUserMessage = {
  role: 'user'
  content: string
}

type HunYuanSystemMessage = { role: 'system'; content: string }

type HunYuanToolMessage = { role: 'tool'; content: string; tool_call_id: string }

type HunYuanAssistantMessage =
  | { role: 'assistant'; content: string }
  | {
    role: 'assistant'
    tool_calls: Array<HunYuanToolCall>
    content?: string
  }

type HunYuanToolCall = {
  id: string
  type: 'function'
  function: HunYuanToolCallFunction
}

type HunYuanToolCallFunction = {
  name: string
  arguments: string
}

export type HunYuanMessage = HunYuanUserMessage | HunYuanSystemMessage | HunYuanToolMessage | HunYuanAssistantMessage

type HunYuanTool = {
  // TODO: 混元文档说当前只支持 function
  type: 'function'
  function: HunYuanToolFunction
}

type HunYuanToolFunction = {
  name: string
  parameters: object
  description?: string
}

export type HunYuanInputData = {
  model:
  | (string & {})
  | 'hunyuan-lite'
  | 'hunyuan-standard'
  | 'hunyuan-standard-256K'
  | 'hunyuan-pro'
  | 'hunyuan-code'
  | 'hunyuan-role'
  | 'hunyuan-functioncall'
  | 'hunyuan-vision'
  messages: Array<HunYuanMessage>
  stream_moderation?: boolean
  top_p?: number
  temperature?: number
  enable_enhancement?: boolean
  tools?: Array<HunYuanTool>
  tool_choice?: 'none' | 'auto' | 'custom'
  custom_tool?: HunYuanTool
  search_info?: boolean
}

type Usage = { prompt_tokens: number; completion_tokens: number; total_tokens: number }

export type HunYuanGenerateTextOutput = {
  choices?: Array<GenerateChoice>
  usage?: Usage
}

type GenerateChoice = {
  finish_reason?: string
  message: HunYuanMessage
}

type StreamChoice = {
  finish_reason?: string
  delta: HunYuanAssistantMessage
}

export type HunYuanStreamTextOutput = {
  choices?: Array<StreamChoice>
  usage?: Usage
}
