// @cloudbase/agent-tools/dist/fs/operator/transaction/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SandboxSnapshotStrategy = exports.MemorySnapshotStrategy = exports.LocalSnapshotStrategy = exports.SnapshotStrategy = exports.EOperatorType = exports.TransactionFileOperator = void 0;
// Transaction-related exports
var transaction_operator_1 = require("./transaction-operator");
Object.defineProperty(exports, "TransactionFileOperator", { enumerable: true, get: function () { return transaction_operator_1.TransactionFileOperator; } });
Object.defineProperty(exports, "EOperatorType", { enumerable: true, get: function () { return transaction_operator_1.EOperatorType; } });
var snapshot_strategies_1 = require("./snapshot-strategies");
Object.defineProperty(exports, "SnapshotStrategy", { enumerable: true, get: function () { return snapshot_strategies_1.SnapshotStrategy; } });
Object.defineProperty(exports, "LocalSnapshotStrategy", { enumerable: true, get: function () { return snapshot_strategies_1.LocalSnapshotStrategy; } });
Object.defineProperty(exports, "MemorySnapshotStrategy", { enumerable: true, get: function () { return snapshot_strategies_1.MemorySnapshotStrategy; } });
Object.defineProperty(exports, "SandboxSnapshotStrategy", { enumerable: true, get: function () { return snapshot_strategies_1.SandboxSnapshotStrategy; } });
__exportStar(require("./transaction-utils"), exports);
