// @mistralai/mistralai/src/models/components/conversationrestartrequest.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ClosedEnum } from "../../types/enums.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  CompletionArgs,
  CompletionArgs$inboundSchema,
  CompletionArgs$Outbound,
  CompletionArgs$outboundSchema,
} from "./completionargs.js";
import {
  ConversationInputs,
  ConversationInputs$inboundSchema,
  ConversationInputs$Outbound,
  ConversationInputs$outboundSchema,
} from "./conversationinputs.js";

export const ConversationRestartRequestHandoffExecution = {
  Client: "client",
  Server: "server",
} as const;
export type ConversationRestartRequestHandoffExecution = ClosedEnum<
  typeof ConversationRestartRequestHandoffExecution
>;

/**
 * Request to restart a new conversation from a given entry in the conversation.
 */
export type ConversationRestartRequest = {
  inputs: ConversationInputs;
  stream?: boolean | undefined;
  /**
   * Whether to store the results into our servers or not.
   */
  store?: boolean | undefined;
  handoffExecution?: ConversationRestartRequestHandoffExecution | undefined;
  /**
   * White-listed arguments from the completion API
   */
  completionArgs?: CompletionArgs | undefined;
  /**
   * Custom metadata for the conversation.
   */
  metadata?: { [k: string]: any } | null | undefined;
  fromEntryId: string;
  /**
   * Specific version of the agent to use when restarting. If not provided, uses the current version.
   */
  agentVersion?: number | null | undefined;
};

/** @internal */
export const ConversationRestartRequestHandoffExecution$inboundSchema:
  z.ZodNativeEnum<typeof ConversationRestartRequestHandoffExecution> = z
    .nativeEnum(ConversationRestartRequestHandoffExecution);

/** @internal */
export const ConversationRestartRequestHandoffExecution$outboundSchema:
  z.ZodNativeEnum<typeof ConversationRestartRequestHandoffExecution> =
    ConversationRestartRequestHandoffExecution$inboundSchema;

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace ConversationRestartRequestHandoffExecution$ {
  /** @deprecated use `ConversationRestartRequestHandoffExecution$inboundSchema` instead. */
  export const inboundSchema =
    ConversationRestartRequestHandoffExecution$inboundSchema;
  /** @deprecated use `ConversationRestartRequestHandoffExecution$outboundSchema` instead. */
  export const outboundSchema =
    ConversationRestartRequestHandoffExecution$outboundSchema;
}

/** @internal */
export const ConversationRestartRequest$inboundSchema: z.ZodType<
  ConversationRestartRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  inputs: ConversationInputs$inboundSchema,
  stream: z.boolean().default(false),
  store: z.boolean().default(true),
  handoff_execution: ConversationRestartRequestHandoffExecution$inboundSchema
    .default("server"),
  completion_args: CompletionArgs$inboundSchema.optional(),
  metadata: z.nullable(z.record(z.any())).optional(),
  from_entry_id: z.string(),
  agent_version: z.nullable(z.number().int()).optional(),
}).transform((v) => {
  return remap$(v, {
    "handoff_execution": "handoffExecution",
    "completion_args": "completionArgs",
    "from_entry_id": "fromEntryId",
    "agent_version": "agentVersion",
  });
});

/** @internal */
export type ConversationRestartRequest$Outbound = {
  inputs: ConversationInputs$Outbound;
  stream: boolean;
  store: boolean;
  handoff_execution: string;
  completion_args?: CompletionArgs$Outbound | undefined;
  metadata?: { [k: string]: any } | null | undefined;
  from_entry_id: string;
  agent_version?: number | null | undefined;
};

/** @internal */
export const ConversationRestartRequest$outboundSchema: z.ZodType<
  ConversationRestartRequest$Outbound,
  z.ZodTypeDef,
  ConversationRestartRequest
> = z.object({
  inputs: ConversationInputs$outboundSchema,
  stream: z.boolean().default(false),
  store: z.boolean().default(true),
  handoffExecution: ConversationRestartRequestHandoffExecution$outboundSchema
    .default("server"),
  completionArgs: CompletionArgs$outboundSchema.optional(),
  metadata: z.nullable(z.record(z.any())).optional(),
  fromEntryId: z.string(),
  agentVersion: z.nullable(z.number().int()).optional(),
}).transform((v) => {
  return remap$(v, {
    handoffExecution: "handoff_execution",
    completionArgs: "completion_args",
    fromEntryId: "from_entry_id",
    agentVersion: "agent_version",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace ConversationRestartRequest$ {
  /** @deprecated use `ConversationRestartRequest$inboundSchema` instead. */
  export const inboundSchema = ConversationRestartRequest$inboundSchema;
  /** @deprecated use `ConversationRestartRequest$outboundSchema` instead. */
  export const outboundSchema = ConversationRestartRequest$outboundSchema;
  /** @deprecated use `ConversationRestartRequest$Outbound` instead. */
  export type Outbound = ConversationRestartRequest$Outbound;
}

export function conversationRestartRequestToJSON(
  conversationRestartRequest: ConversationRestartRequest,
): string {
  return JSON.stringify(
    ConversationRestartRequest$outboundSchema.parse(conversationRestartRequest),
  );
}

export function conversationRestartRequestFromJSON(
  jsonString: string,
): SafeParseResult<ConversationRestartRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ConversationRestartRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ConversationRestartRequest' from JSON`,
  );
}
