// @mistralai/mistralai/src/models/components/modelcapabilities.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type ModelCapabilities = {
  completionChat?: boolean | undefined;
  functionCalling?: boolean | undefined;
  completionFim?: boolean | undefined;
  fineTuning?: boolean | undefined;
  vision?: boolean | undefined;
  ocr?: boolean | undefined;
  classification?: boolean | undefined;
  moderation?: boolean | undefined;
  audio?: boolean | undefined;
};

/** @internal */
export const ModelCapabilities$inboundSchema: z.ZodType<
  ModelCapabilities,
  z.ZodTypeDef,
  unknown
> = z.object({
  completion_chat: z.boolean().default(false),
  function_calling: z.boolean().default(false),
  completion_fim: z.boolean().default(false),
  fine_tuning: z.boolean().default(false),
  vision: z.boolean().default(false),
  ocr: z.boolean().default(false),
  classification: z.boolean().default(false),
  moderation: z.boolean().default(false),
  audio: z.boolean().default(false),
}).transform((v) => {
  return remap$(v, {
    "completion_chat": "completionChat",
    "function_calling": "functionCalling",
    "completion_fim": "completionFim",
    "fine_tuning": "fineTuning",
  });
});

/** @internal */
export type ModelCapabilities$Outbound = {
  completion_chat: boolean;
  function_calling: boolean;
  completion_fim: boolean;
  fine_tuning: boolean;
  vision: boolean;
  ocr: boolean;
  classification: boolean;
  moderation: boolean;
  audio: boolean;
};

/** @internal */
export const ModelCapabilities$outboundSchema: z.ZodType<
  ModelCapabilities$Outbound,
  z.ZodTypeDef,
  ModelCapabilities
> = z.object({
  completionChat: z.boolean().default(false),
  functionCalling: z.boolean().default(false),
  completionFim: z.boolean().default(false),
  fineTuning: z.boolean().default(false),
  vision: z.boolean().default(false),
  ocr: z.boolean().default(false),
  classification: z.boolean().default(false),
  moderation: z.boolean().default(false),
  audio: z.boolean().default(false),
}).transform((v) => {
  return remap$(v, {
    completionChat: "completion_chat",
    functionCalling: "function_calling",
    completionFim: "completion_fim",
    fineTuning: "fine_tuning",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace ModelCapabilities$ {
  /** @deprecated use `ModelCapabilities$inboundSchema` instead. */
  export const inboundSchema = ModelCapabilities$inboundSchema;
  /** @deprecated use `ModelCapabilities$outboundSchema` instead. */
  export const outboundSchema = ModelCapabilities$outboundSchema;
  /** @deprecated use `ModelCapabilities$Outbound` instead. */
  export type Outbound = ModelCapabilities$Outbound;
}

export function modelCapabilitiesToJSON(
  modelCapabilities: ModelCapabilities,
): string {
  return JSON.stringify(
    ModelCapabilities$outboundSchema.parse(modelCapabilities),
  );
}

export function modelCapabilitiesFromJSON(
  jsonString: string,
): SafeParseResult<ModelCapabilities, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ModelCapabilities$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ModelCapabilities' from JSON`,
  );
}
