// @mistralai/mistralai/src/models/components/requestsource.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { ClosedEnum } from "../../types/enums.js";

export const RequestSource = {
  Api: "api",
  Playground: "playground",
  AgentBuilderV1: "agent_builder_v1",
} as const;
export type RequestSource = ClosedEnum<typeof RequestSource>;

/** @internal */
export const RequestSource$inboundSchema: z.ZodNativeEnum<
  typeof RequestSource
> = z.nativeEnum(RequestSource);

/** @internal */
export const RequestSource$outboundSchema: z.ZodNativeEnum<
  typeof RequestSource
> = RequestSource$inboundSchema;

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace RequestSource$ {
  /** @deprecated use `RequestSource$inboundSchema` instead. */
  export const inboundSchema = RequestSource$inboundSchema;
  /** @deprecated use `RequestSource$outboundSchema` instead. */
  export const outboundSchema = RequestSource$outboundSchema;
}
