// @cloudbase/ai/src/models/Yi/type.ts
type SystemMessage = {
  role: 'system'
  content: string
}

type UserMessage = {
  role: 'user'
  content: string
}

type AssistantMessage =
  | {
    role: 'assistant'
    content: string
  }
  | {
    role: 'assistant'
    tool_calls: Array<ToolCall>
    content?: string
  }

type ToolMessage = {
  role: 'tool'
  content: string
  tool_call_id: string
}

type YiMessage = SystemMessage | UserMessage | AssistantMessage | ToolMessage

export type YiInput = {
  model: string
  messages: Array<YiMessage>
  temperature?: number
  top_p?: number
  tools?: Array<{
    type: string
    function: { name: string; description: string; parameters: object }
  }>
}

type ToolCall = {
  id: string
  type: string
  function: { name: string; arguments: string }
}

export type YiGenerateOutput = {
  choices: Array<{
    finish_reason: string
    message: {
      content: string
      role: 'assistant'
      tool_calls?: Array<ToolCall>
    }
  }>
  usage: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}

export type YiStreamOutput = {
  choices: Array<{
    delta: {
      content: string
      tool_calls?: Array<ToolCall>
      role?: 'assistant'
    }
    finish_reason: string
  }>
  usage?: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}
