// @mistralai/mistralai/src/models/operations/agentsapiv1agentsget.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { remap as remap$ } from "../../lib/primitives.js";

export type AgentsApiV1AgentsGetRequest = {
  agentId: string;
  agentVersion?: number | null | undefined;
};

/** @internal */
export type AgentsApiV1AgentsGetRequest$Outbound = {
  agent_id: string;
  agent_version?: number | null | undefined;
};

/** @internal */
export const AgentsApiV1AgentsGetRequest$outboundSchema: z.ZodType<
  AgentsApiV1AgentsGetRequest$Outbound,
  z.ZodTypeDef,
  AgentsApiV1AgentsGetRequest
> = z.object({
  agentId: z.string(),
  agentVersion: z.nullable(z.number().int()).optional(),
}).transform((v) => {
  return remap$(v, {
    agentId: "agent_id",
    agentVersion: "agent_version",
  });
});

export function agentsApiV1AgentsGetRequestToJSON(
  agentsApiV1AgentsGetRequest: AgentsApiV1AgentsGetRequest,
): string {
  return JSON.stringify(
    AgentsApiV1AgentsGetRequest$outboundSchema.parse(
      agentsApiV1AgentsGetRequest,
    ),
  );
}
