// cloudflare/resources/stream/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = exports.Watermarks = exports.WatermarksSinglePage = exports.Videos = exports.Token = exports.Stream = exports.LiveInputs = exports.Keys = exports.KeyGetResponsesSinglePage = exports.Embed = exports.Downloads = exports.DirectUpload = exports.Copy = exports.ClipResource = exports.Captions = exports.CaptionsSinglePage = exports.AudioTracks = exports.AudioSinglePage = void 0;
var audio_tracks_1 = require("./audio-tracks.js");
Object.defineProperty(exports, "AudioSinglePage", { enumerable: true, get: function () { return audio_tracks_1.AudioSinglePage; } });
Object.defineProperty(exports, "AudioTracks", { enumerable: true, get: function () { return audio_tracks_1.AudioTracks; } });
var index_1 = require("./captions/index.js");
Object.defineProperty(exports, "CaptionsSinglePage", { enumerable: true, get: function () { return index_1.CaptionsSinglePage; } });
Object.defineProperty(exports, "Captions", { enumerable: true, get: function () { return index_1.Captions; } });
var clip_1 = require("./clip.js");
Object.defineProperty(exports, "ClipResource", { enumerable: true, get: function () { return clip_1.ClipResource; } });
var copy_1 = require("./copy.js");
Object.defineProperty(exports, "Copy", { enumerable: true, get: function () { return copy_1.Copy; } });
var direct_upload_1 = require("./direct-upload.js");
Object.defineProperty(exports, "DirectUpload", { enumerable: true, get: function () { return direct_upload_1.DirectUpload; } });
var downloads_1 = require("./downloads.js");
Object.defineProperty(exports, "Downloads", { enumerable: true, get: function () { return downloads_1.Downloads; } });
var embed_1 = require("./embed.js");
Object.defineProperty(exports, "Embed", { enumerable: true, get: function () { return embed_1.Embed; } });
var keys_1 = require("./keys.js");
Object.defineProperty(exports, "KeyGetResponsesSinglePage", { enumerable: true, get: function () { return keys_1.KeyGetResponsesSinglePage; } });
Object.defineProperty(exports, "Keys", { enumerable: true, get: function () { return keys_1.Keys; } });
var index_2 = require("./live-inputs/index.js");
Object.defineProperty(exports, "LiveInputs", { enumerable: true, get: function () { return index_2.LiveInputs; } });
var stream_1 = require("./stream.js");
Object.defineProperty(exports, "Stream", { enumerable: true, get: function () { return stream_1.Stream; } });
var token_1 = require("./token.js");
Object.defineProperty(exports, "Token", { enumerable: true, get: function () { return token_1.Token; } });
var videos_1 = require("./videos.js");
Object.defineProperty(exports, "Videos", { enumerable: true, get: function () { return videos_1.Videos; } });
var watermarks_1 = require("./watermarks.js");
Object.defineProperty(exports, "WatermarksSinglePage", { enumerable: true, get: function () { return watermarks_1.WatermarksSinglePage; } });
Object.defineProperty(exports, "Watermarks", { enumerable: true, get: function () { return watermarks_1.Watermarks; } });
var webhooks_1 = require("./webhooks.js");
Object.defineProperty(exports, "Webhooks", { enumerable: true, get: function () { return webhooks_1.Webhooks; } });
