// @cloudbase/app/dist/cjs/libs/adapter.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWxDefaultAdapter = exports.Platform = void 0;
var adapter_wx_mp_1 = __importStar(require("@cloudbase/adapter-wx_mp"));
var adapter_interface_1 = require("@cloudbase/adapter-interface");
exports.Platform = {};
var getWxDefaultAdapter = function () {
    adapter_wx_mp_1.WxRequest.prototype.upload = function (options) {
        var _this = this;
        var self = this;
        return new Promise(function (resolve) {
            var url = options.url, file = options.file, data = options.data, headers = options.headers;
            var fs = wx.getFileSystemManager();
            var task = wx.request({
                url: url,
                method: options.method,
                header: __assign({ 'content-type': ' ' }, headers),
                data: fs.readFileSync(file),
                timeout: _this._timeout,
                success: function (res) {
                    var result = {
                        statusCode: res.statusCode,
                        data: res.data || {},
                    };
                    if (res.statusCode === 200 && (data === null || data === void 0 ? void 0 : data.success_action_status)) {
                        result.statusCode = parseInt(data.success_action_status, 10);
                    }
                    resolve(result);
                },
                fail: function (err) {
                    resolve(err);
                },
                complete: function (err) {
                    if (!(err === null || err === void 0 ? void 0 : err.errMsg)) {
                        return;
                    }
                    if (!self._timeout || self._restrictedMethods.indexOf('upload') === -1) {
                        return;
                    }
                    var errMsg = err.errMsg;
                    if (errMsg === 'request:fail timeout') {
                        console.warn(self._timeoutMsg);
                        try {
                            task.abort();
                        }
                        catch (e) { }
                    }
                },
            });
        });
    };
    function isPlugin() {
        return (typeof App === 'undefined'
            && typeof getApp === 'undefined'
            && !wx.onAppHide
            && !wx.offAppHide
            && !wx.onAppShow
            && !wx.offAppShow);
    }
    adapter_wx_mp_1.default.genAdapter = function genAdapter(options) {
        var adapter = {
            root: { globalThis: {} },
            reqClass: adapter_wx_mp_1.WxRequest,
            wsClass: adapter_wx_mp_1.WxMpWebSocket,
            captchaOptions: {
                openURIWithCallback: function (_url) {
                    var EventBus = options.EventBus;
                    var queryObj = {};
                    var url = _url;
                    console.log('openURIWithCallback', _url);
                    var matched = _url.match(/^(data:.*?)(\?[^#\s]*)?$/);
                    if (matched) {
                        url = matched[1];
                        console.log('openURIWithCallback url', url);
                        var search = matched[2];
                        if (search) {
                            queryObj = (0, adapter_wx_mp_1.parseQueryString)(search);
                        }
                    }
                    console.log('openURIWithCallback queryObj', queryObj);
                    var token = queryObj.token, restQueryObj = __rest(queryObj, ["token"]);
                    if (/^data:/.test(url) && !token) {
                        return Promise.reject({
                            error: 'invalid_argument',
                            error_description: "invalie captcha data: ".concat(_url),
                        });
                    }
                    if (!token) {
                        return Promise.reject({
                            error: 'unimplemented',
                            error_description: 'need to impl captcha data',
                        });
                    }
                    return new Promise(function (resolve) {
                        console.log('wait for captcha...');
                        EventBus.$emit('CAPTCHA_DATA_CHANGE', __assign(__assign({}, restQueryObj), { token: token, url: url }));
                        EventBus.$once('RESOLVE_CAPTCHA_DATA', function (res) {
                            resolve(res);
                        });
                    });
                },
            },
            localStorage: adapter_wx_mp_1.wxMpStorage,
            primaryStorage: adapter_interface_1.StorageType.local,
            getAppSign: function () {
                var info = wx.getAccountInfoSync();
                if (isPlugin()) {
                    return info && info.plugin ? info.plugin.appId : '';
                }
                return info && info.miniProgram ? info.miniProgram.appId : '';
            },
        };
        return adapter;
    };
    return adapter_wx_mp_1.default;
};
exports.getWxDefaultAdapter = getWxDefaultAdapter;
