// @cloudbase/app/dist/cjs/libs/lang.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.i18nProxy = exports.t = exports.langMap = exports.LANG_HEADER_KEY = exports.LANGS = void 0;
var utilities_1 = require("@cloudbase/utilities");
var LANGS;
(function (LANGS) {
    LANGS["ZH"] = "zh-CN";
    LANGS["EN"] = "en-US";
})(LANGS = exports.LANGS || (exports.LANGS = {}));
exports.LANG_HEADER_KEY = 'Accept-Language';
exports.langMap = (_a = {},
    _a[LANGS.EN] = {
        '请检查调用方式，或前往云开发 AI+ 首页查看文档': 'Please check the calling method or visit the AI+ homepage to view the documentation',
        'AI+ 请求出错，错误码': 'AI+ request error, error code',
        错误信息: 'error message',
        未实现: 'not implemented',
        '小程序不支持 wasm 加载': 'wasm loading is not supported in mini programs',
        '小程序不支持动态 js 加载': 'dynamic js loading is not supported in mini programs',
        请求超时: 'request timeout',
        '缺少 privatelink sdk 地址': 'missing privatelink sdk address',
    },
    _a);
var t = function (text, lang) { var _a; return ((_a = exports.langMap[lang]) === null || _a === void 0 ? void 0 : _a[text]) || text; };
exports.t = t;
var getLangCacheKey = function (config) { return "lang_".concat(config.clientId || config.env); };
var i18nProxy = function (platform, config) {
    var _a, _b;
    var cacheKey = getLangCacheKey(config);
    var localStorage = (platform.adapter || {}).localStorage;
    var lang = config.lang || ((_a = localStorage === null || localStorage === void 0 ? void 0 : localStorage.getItem) === null || _a === void 0 ? void 0 : _a.call(localStorage, cacheKey)) || LANGS.ZH;
    (_b = localStorage === null || localStorage === void 0 ? void 0 : localStorage.setItem) === null || _b === void 0 ? void 0 : _b.call(localStorage, cacheKey, lang);
    return new Proxy({
        t: function (text) { return (0, exports.t)(text, lang); },
        LANG_HEADER_KEY: exports.LANG_HEADER_KEY,
        lang: lang,
    }, {
        get: function (target, prop) {
            var _a;
            if (prop === 'lang') {
                return ((_a = localStorage === null || localStorage === void 0 ? void 0 : localStorage.getItem) === null || _a === void 0 ? void 0 : _a.call(localStorage, cacheKey)) || target[prop];
            }
            return target[prop];
        },
        set: function (target, prop, newValue) {
            var _a;
            if (prop === 'lang' && newValue !== target[prop]) {
                target[prop] = newValue;
                target.t = function (text) { return (0, exports.t)(text, newValue); };
                (_a = localStorage === null || localStorage === void 0 ? void 0 : localStorage.setItem) === null || _a === void 0 ? void 0 : _a.call(localStorage, cacheKey, newValue);
                utilities_1.langEvent.bus.fire(utilities_1.langEvent.LANG_CHANGE_EVENT, { i18n: target });
            }
            target[prop] = newValue;
            return true;
        },
    });
};
exports.i18nProxy = i18nProxy;
