// cloudflare/resources/zero-trust/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZeroTrust = exports.Tunnels = exports.TunnelListResponsesV4PagePaginationArray = exports.Seats = exports.SeatsSinglePage = exports.RiskScoring = exports.Organizations = exports.Networks = exports.IdentityProviders = exports.IdentityProviderListResponsesSinglePage = exports.Gateway = exports.Devices = exports.DevicesSinglePage = exports.DLP = exports.DEX = exports.ConnectivitySettings = exports.Access = void 0;
var index_1 = require("./access/index.js");
Object.defineProperty(exports, "Access", { enumerable: true, get: function () { return index_1.Access; } });
var connectivity_settings_1 = require("./connectivity-settings.js");
Object.defineProperty(exports, "ConnectivitySettings", { enumerable: true, get: function () { return connectivity_settings_1.ConnectivitySettings; } });
var index_2 = require("./dex/index.js");
Object.defineProperty(exports, "DEX", { enumerable: true, get: function () { return index_2.DEX; } });
var index_3 = require("./dlp/index.js");
Object.defineProperty(exports, "DLP", { enumerable: true, get: function () { return index_3.DLP; } });
var index_4 = require("./devices/index.js");
Object.defineProperty(exports, "DevicesSinglePage", { enumerable: true, get: function () { return index_4.DevicesSinglePage; } });
Object.defineProperty(exports, "Devices", { enumerable: true, get: function () { return index_4.Devices; } });
var index_5 = require("./gateway/index.js");
Object.defineProperty(exports, "Gateway", { enumerable: true, get: function () { return index_5.Gateway; } });
var index_6 = require("./identity-providers/index.js");
Object.defineProperty(exports, "IdentityProviderListResponsesSinglePage", { enumerable: true, get: function () { return index_6.IdentityProviderListResponsesSinglePage; } });
Object.defineProperty(exports, "IdentityProviders", { enumerable: true, get: function () { return index_6.IdentityProviders; } });
var index_7 = require("./networks/index.js");
Object.defineProperty(exports, "Networks", { enumerable: true, get: function () { return index_7.Networks; } });
var index_8 = require("./organizations/index.js");
Object.defineProperty(exports, "Organizations", { enumerable: true, get: function () { return index_8.Organizations; } });
var index_9 = require("./risk-scoring/index.js");
Object.defineProperty(exports, "RiskScoring", { enumerable: true, get: function () { return index_9.RiskScoring; } });
var seats_1 = require("./seats.js");
Object.defineProperty(exports, "SeatsSinglePage", { enumerable: true, get: function () { return seats_1.SeatsSinglePage; } });
Object.defineProperty(exports, "Seats", { enumerable: true, get: function () { return seats_1.Seats; } });
var index_10 = require("./tunnels/index.js");
Object.defineProperty(exports, "TunnelListResponsesV4PagePaginationArray", { enumerable: true, get: function () { return index_10.TunnelListResponsesV4PagePaginationArray; } });
Object.defineProperty(exports, "Tunnels", { enumerable: true, get: function () { return index_10.Tunnels; } });
var zero_trust_1 = require("./zero-trust.js");
Object.defineProperty(exports, "ZeroTrust", { enumerable: true, get: function () { return zero_trust_1.ZeroTrust; } });
