// @cloudbase/agent-shared/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  assistantMessageSchema: () => assistantMessageSchema,
  clientMessageSchema: () => clientMessageSchema,
  isV4Schema: () => isV4Schema,
  sendMessageEventSchema: () => sendMessageEventSchema,
  sendMessageInputSchema: () => sendMessageInputSchema,
  systemMessageSchema: () => systemMessageSchema,
  toolMessageSchema: () => toolMessageSchema,
  toolSchema: () => toolSchema,
  userMessageSchema: () => userMessageSchema
});
module.exports = __toCommonJS(index_exports);
var z = __toESM(require("zod/v4"));
function isV4Schema(schema) {
  return "_zod" in schema;
}
var systemMessageSchema = z.object({
  role: z.literal("system"),
  content: z.string()
});
var userMessageSchema = z.object({
  role: z.literal("user"),
  content: z.string()
});
var toolMessageSchema = z.object({
  role: z.literal("tool"),
  content: z.string(),
  toolCallId: z.string()
});
var assistantMessageSchema = z.object({
  role: z.literal("assistant"),
  content: z.string().optional(),
  toolCalls: z.array(
    z.object({
      id: z.string(),
      type: z.literal("function"),
      function: z.object({
        name: z.string(),
        arguments: z.string()
      })
    })
  ).optional()
});
var clientMessageSchema = z.union([
  systemMessageSchema,
  userMessageSchema,
  toolMessageSchema,
  assistantMessageSchema
]);
var toolSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.any()
});
var sendMessageInputSchema = z.object({
  conversationId: z.string(),
  messages: z.array(clientMessageSchema).optional(),
  resume: z.object({
    interruptId: z.string(),
    payload: z.any()
  }).optional(),
  tools: z.array(toolSchema).optional()
});
var sendMessageEventSchema = z.union([
  z.object({
    type: z.literal("text"),
    content: z.string()
  }),
  z.object({
    type: z.literal("tool-call-start"),
    toolCallId: z.string(),
    toolCallName: z.string()
  }),
  z.object({
    type: z.literal("tool-call-args"),
    toolCallId: z.string(),
    delta: z.string()
  }),
  z.object({
    type: z.literal("tool-call-end"),
    toolCallId: z.string()
  }),
  z.object({
    type: z.literal("tool-result"),
    toolCallId: z.string(),
    result: z.string()
  }),
  z.object({
    type: z.literal("interrupt"),
    id: z.string(),
    reason: z.string(),
    payload: z.any()
  })
]);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  assistantMessageSchema,
  clientMessageSchema,
  isV4Schema,
  sendMessageEventSchema,
  sendMessageInputSchema,
  systemMessageSchema,
  toolMessageSchema,
  toolSchema,
  userMessageSchema
});
