// @cloudbase/ai/src/models/Moonshot/type.ts
type SystemMessage = {
  role: 'system'
  content: string
}

type UserMessage = {
  role: 'user'
  content: string
}

type AssistantMessage = {
  role: 'assistant'
  content?: string
  tool_calls?: Array<{
    id: string
    type: string
    function?: { name: string; arguments: string }
  }>
}

type ToolMessage = {
  role: 'tool'
  content: string
  tool_call_id: string
}

type MoonshotMessage = SystemMessage | UserMessage | AssistantMessage | ToolMessage

export type MoonshotInputData = {
  messages: Array<MoonshotMessage>
  model: string
  max_tokens?: number
  temperature?: number
  top_p?: number
  n?: number
  tools?: Array<{
    type: string
    function: { name: string; description: string; parameters: object }
  }>
  presence_penalty?: number
  frequency_penalty?: number
  response_format?: object
  stop?: string | Array<string>
}

// todo: more precise
export type MoonshotGenerateTextOutput = {
  choices: Array<{
    finish_reason: string
    message: {
      content: string
      role: 'assistant'
      tool_calls?: Array<ToolCall>
    }
  }>
  usage: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}
// todo: more precise
export type MoonshotStreamTextOutput = {
  choices: Array<{
    delta: {
      content: string
      tool_calls?: Array<ToolCall>
      role: 'assistant'
    }
    finish_reason: string
  }>
  usage?: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}
type ToolCall = {
  id: string
  type: string
  function: { name: string; arguments: string }
}
