// @cloudbase/agent-tools/dist/fs/toolkit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilesystemToolkit = void 0;
const read_tool_1 = require("./read-tool");
const write_tool_1 = require("./write-tool");
const edit_tool_1 = require("./edit-tool");
const glob_tool_1 = require("./glob-tool");
const grep_tool_1 = require("./grep-tool");
const ls_tool_1 = require("./ls-tool");
const str_replace_editor_1 = require("./str-replace-editor");
const multi_read_tool_1 = require("./multi-read-tool");
const multi_write_tool_1 = require("./multi-write-tool");
const multi_edit_tool_1 = require("./multi-edit-tool");
const utils_1 = require("../utils");
/**
 * FilesystemToolkit provides a comprehensive set of file system tools
 * with configurable modes for different use cases.
 */
class FilesystemToolkit extends utils_1.BaseToolkit {
    constructor(options) {
        super({
            name: options.name || "filesystem-toolkit",
            description: options.description || "Comprehensive file system operations toolkit",
            context: options.context,
        });
        this.context = {
            workingDirectory: "",
            fsOperator: null,
        };
        this.context = options.context;
        this.config = {
            mode: options.config?.mode || "multi",
        };
        options.context.fsOperator.setWorkspaceRoot(options.context.workingDirectory);
        this.initialize();
    }
    /**
     * Initialize toolkit and add tools
     */
    async onInitialize() {
        this.initializeTools();
    }
    /**
     * Initialize tools based on configuration
     */
    initializeTools() {
        // Clear existing tools
        this.getTools().forEach((tool) => {
            this.removeTool(tool.name);
        });
        // Add file operation tools based on mode
        if (this.config.mode === "integrated") {
            // Integrated mode: use str-replace-editor
            const strReplaceEditor = (0, str_replace_editor_1.createStrReplaceEditorTool)(this.context);
            this.addTool(strReplaceEditor);
        }
        else if (this.config.mode === "multi") {
            // Always add multi-file tools
            const multiReadTool = (0, multi_read_tool_1.createMultiReadTool)(this.context);
            this.addTool(multiReadTool);
            const multiWriteTool = (0, multi_write_tool_1.createMultiWriteTool)(this.context);
            this.addTool(multiWriteTool);
            const multiEditTool = (0, multi_edit_tool_1.createMultiEditTool)(this.context);
            this.addTool(multiEditTool);
        }
        else {
            // Atomic mode: use separate read, write, edit tools
            const readTool = (0, read_tool_1.createReadTool)(this.context);
            this.addTool(readTool);
            const writeTool = (0, write_tool_1.createWriteTool)(this.context);
            this.addTool(writeTool);
            const editTool = (0, edit_tool_1.createEditTool)(this.context);
            this.addTool(editTool);
        }
        // Always add utility tools
        const globTool = (0, glob_tool_1.createGlobTool)(this.context);
        this.addTool(globTool);
        const grepTool = (0, grep_tool_1.createGrepTool)(this.context);
        this.addTool(grepTool);
        const lsTool = (0, ls_tool_1.createLsTool)(this.context);
        this.addTool(lsTool);
    }
    /**
     * Get tools by category
     */
    getFileOperationTools() {
        if (this.config.mode === "integrated") {
            return this.getTools().filter((tool) => tool.name === "str_replace_editor");
        }
        else {
            return this.getTools().filter((tool) => ["read", "write", "edit"].includes(tool.name));
        }
    }
    /**
     * Get utility tools (glob, grep, ls)
     */
    getUtilityTools() {
        return this.getTools().filter((tool) => ["glob", "grep", "ls"].includes(tool.name));
    }
    /**
     * Get current configuration
     */
    getConfig() {
        return { ...this.config };
    }
    /**
     * Switch between integrated and atomic modes
     */
    switchMode(mode) {
        if (this.config.mode !== mode) {
            this.config.mode = mode;
            this.initializeTools();
        }
    }
    /**
     * Get toolkit statistics
     */
    getStats() {
        return {
            mode: this.config.mode,
        };
    }
}
exports.FilesystemToolkit = FilesystemToolkit;
