// @cloudbase/ai/dist/cjs/eventsource_parser/parse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createParser = void 0;
function createParser(onParse) {
    var isFirstChunk;
    var buffer;
    var startingPosition;
    var startingFieldLength;
    var eventId;
    var eventName;
    var data;
    reset();
    return { feed: feed, reset: reset };
    function reset() {
        isFirstChunk = true;
        buffer = '';
        startingPosition = 0;
        startingFieldLength = -1;
        eventId = undefined;
        eventName = undefined;
        data = '';
    }
    function feed(chunk) {
        buffer = buffer ? buffer + chunk : chunk;
        if (isFirstChunk && hasBom(buffer)) {
            buffer = buffer.slice(BOM.length);
        }
        isFirstChunk = false;
        var length = buffer.length;
        var position = 0;
        var discardTrailingNewline = false;
        while (position < length) {
            if (discardTrailingNewline) {
                if (buffer[position] === '\n') {
                    position += 1;
                }
                discardTrailingNewline = false;
            }
            var lineLength = -1;
            var fieldLength = startingFieldLength;
            var character = void 0;
            for (var index = startingPosition; lineLength < 0 && index < length; index++) {
                character = buffer[index];
                if (character === ':' && fieldLength < 0) {
                    fieldLength = index - position;
                }
                else if (character === '\r') {
                    discardTrailingNewline = true;
                    lineLength = index - position;
                }
                else if (character === '\n') {
                    lineLength = index - position;
                }
            }
            if (lineLength < 0) {
                startingPosition = length - position;
                startingFieldLength = fieldLength;
                break;
            }
            else {
                startingPosition = 0;
                startingFieldLength = -1;
            }
            parseEventStreamLine(buffer, position, fieldLength, lineLength);
            position += lineLength + 1;
        }
        if (position === length) {
            buffer = '';
        }
        else if (position > 0) {
            buffer = buffer.slice(position);
        }
    }
    function parseEventStreamLine(lineBuffer, index, fieldLength, lineLength) {
        if (lineLength === 0) {
            if (data.length > 0) {
                onParse({
                    type: 'event',
                    id: eventId,
                    event: eventName || undefined,
                    data: data.slice(0, -1),
                });
                data = '';
                eventId = undefined;
            }
            eventName = undefined;
            return;
        }
        var noValue = fieldLength < 0;
        var field = lineBuffer.slice(index, index + (noValue ? lineLength : fieldLength));
        var step = 0;
        if (noValue) {
            step = lineLength;
        }
        else if (lineBuffer[index + fieldLength + 1] === ' ') {
            step = fieldLength + 2;
        }
        else {
            step = fieldLength + 1;
        }
        var position = index + step;
        var valueLength = lineLength - step;
        var value = lineBuffer.slice(position, position + valueLength).toString();
        if (field === 'data') {
            data += value ? "".concat(value, "\n") : '\n';
        }
        else if (field === 'event') {
            eventName = value;
        }
        else if (field === 'id' && !value.includes('\u0000')) {
            eventId = value;
        }
        else if (field === 'retry') {
            var retry = parseInt(value, 10);
            if (!Number.isNaN(retry)) {
                onParse({ type: 'reconnect-interval', value: retry });
            }
        }
    }
}
exports.createParser = createParser;
var BOM = [239, 187, 191];
function hasBom(buffer) {
    return BOM.every(function (charCode, index) { return buffer.charCodeAt(index) === charCode; });
}
