// @cloudbase/agent-observability/src/core/constants.ts
/**
 * OTEL attribute constants for AG-Kit observability.
 *
 * Uses OpenInference semantic conventions where applicable:
 * https://github.com/Arize-ai/openinference/tree/main/spec
 *
 * Falls back to AG-Kit specific attributes where OpenInference
 * doesn't define a standard.
 *
 * @module
 */

import {
  SemanticConventions,
  OpenInferenceSpanKind,
} from "@arizeai/openinference-semantic-conventions";

// Re-export OpenInference types for convenience
export { OpenInferenceSpanKind };

/**
 * SDK information
 */
export const OBSERVABILITY_TRACER_NAME = "agkit-tracer";
export const OBSERVABILITY_SDK_NAME = "@agkit/observability";
// Version will be injected from package.json

/**
 * Combined attribute namespace for internal use
 * Provides a single namespace for all OTEL attributes used by AG-Kit
 *
 * Combines OpenInference SemanticConventions with AG-Kit specific attributes
 */
export const OtelSpanAttributes = {
  // OpenInference - re-export all standard conventions
  ...SemanticConventions,

  // AG-Kit Trace attributes (non-standard)
  TRACE_NAME: "trace.name",
  TRACE_TAGS: "trace.tags",
  TRACE_PUBLIC: "trace.public",
  TRACE_METADATA: "trace.metadata",
  TRACE_INPUT: "trace.input",
  TRACE_OUTPUT: "trace.output",

  // AG-Kit Observation attributes (non-standard)
  OBSERVATION_TYPE: "observation.type",
  OBSERVATION_LEVEL: "observation.level",
  OBSERVATION_STATUS_MESSAGE: "observation.status_message",
  OBSERVATION_INPUT: "observation.input",
  OBSERVATION_OUTPUT: "observation.output",
  OBSERVATION_METADATA: "observation.metadata",

  // AG-Kit LLM-specific (non-standard)
  LLM_COMPLETION_START_TIME: "llm.completion_start_time",
  LLM_MODEL_PARAMETERS: "llm.model_parameters",
  LLM_USAGE_DETAILS: "llm.usage_details",
  LLM_COST_DETAILS: "llm.cost_details",

  // AG-Kit Retriever-specific (non-standard)
  RETRIEVER_NAME: "retriever.name",
  RETRIEVER_QUERY: "retriever.query",
  RETRIEVER_INDEX_ID: "retriever.index_id",
  RETRIEVER_TOP_K: "retriever.top_k",

  // AG-Kit General (non-standard)
  ENVIRONMENT: "environment",
  RELEASE: "release",
  VERSION: "version",
} as const;

/**
 * Type for the OtelSpanAttributes object values
 */
export type OtelSpanAttributeValues = typeof OtelSpanAttributes[keyof typeof OtelSpanAttributes];
