// cloudflare/resources/browser-rendering/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Snapshot = exports.Screenshot = exports.Scrape = exports.PDF = exports.Markdown = exports.Links = exports.Json = exports.Content = exports.BrowserRendering = void 0;
var browser_rendering_1 = require("./browser-rendering.js");
Object.defineProperty(exports, "BrowserRendering", { enumerable: true, get: function () { return browser_rendering_1.BrowserRendering; } });
var content_1 = require("./content.js");
Object.defineProperty(exports, "Content", { enumerable: true, get: function () { return content_1.Content; } });
var json_1 = require("./json.js");
Object.defineProperty(exports, "Json", { enumerable: true, get: function () { return json_1.Json; } });
var links_1 = require("./links.js");
Object.defineProperty(exports, "Links", { enumerable: true, get: function () { return links_1.Links; } });
var markdown_1 = require("./markdown.js");
Object.defineProperty(exports, "Markdown", { enumerable: true, get: function () { return markdown_1.Markdown; } });
var pdf_1 = require("./pdf.js");
Object.defineProperty(exports, "PDF", { enumerable: true, get: function () { return pdf_1.PDF; } });
var scrape_1 = require("./scrape.js");
Object.defineProperty(exports, "Scrape", { enumerable: true, get: function () { return scrape_1.Scrape; } });
var screenshot_1 = require("./screenshot.js");
Object.defineProperty(exports, "Screenshot", { enumerable: true, get: function () { return screenshot_1.Screenshot; } });
var snapshot_1 = require("./snapshot.js");
Object.defineProperty(exports, "Snapshot", { enumerable: true, get: function () { return snapshot_1.Snapshot; } });
