// @cloudbase/ai/dist/cjs/bot/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bot = void 0;
var utils_1 = require("../utils");
var Bot = (function () {
    function Bot(req, baseUrl) {
        this.baseUrl = baseUrl;
        var token = arguments[2];
        if (typeof token === 'string') {
            this.req = function (_a) {
                var _b = _a.headers, headers = _b === void 0 ? {} : _b, rest = __rest(_a, ["headers"]);
                return req(__assign(__assign({}, rest), { headers: __assign(__assign({}, headers), { Authorization: "Bearer ".concat(token) }) }));
            };
        }
        else {
            this.req = req;
        }
    }
    Bot.prototype.list = function (props, options) {
        return this.req({
            method: 'get',
            url: this.join('bots'),
            data: props,
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.create = function (_a, options) {
        var botInfo = _a.botInfo;
        return this.req({
            method: 'post',
            url: this.join('bots'),
            data: botInfo,
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.get = function (_a, options) {
        var botId = _a.botId;
        return this.req({
            method: 'get',
            url: this.join("bots/".concat(botId)),
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.update = function (_a, options) {
        var botId = _a.botId, botInfo = _a.botInfo;
        return this.req({
            method: 'PATCH',
            url: this.join("bots/".concat(botId)),
            data: botInfo,
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.delete = function (_a, options) {
        var botId = _a.botId;
        return this.req({ method: 'delete', url: this.join("bots/".concat(botId)), timeout: options === null || options === void 0 ? void 0 : options.timeout });
    };
    Bot.prototype.getChatRecords = function (props, options) {
        return this.req({
            method: 'get',
            url: this.join("bots/".concat(props.botId, "/records")),
            data: props,
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.sendFeedback = function (_a, options) {
        var userFeedback = _a.userFeedback;
        return this.req({
            method: 'post',
            url: this.join("bots/".concat(userFeedback.botId, "/feedback")),
            data: userFeedback,
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.getFeedback = function (props, options) {
        return this.req({
            method: 'get',
            url: this.join("bots/".concat(props.botId, "/feedback")),
            data: props,
            timeout: options === null || options === void 0 ? void 0 : options.timeout,
        });
    };
    Bot.prototype.uploadFiles = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.req({
                        method: 'post',
                        url: this.join("bots/".concat(props.botId, "/files")),
                        data: props,
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.createConversation = function (_a, options) {
        var botId = _a.botId, rest = __rest(_a, ["botId"]);
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2, this.req({
                        method: 'post',
                        url: this.join("bots/".concat(botId, "/conversation")),
                        data: rest,
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.getConversation = function (_a, options) {
        var _b = _a.pageSize, pageSize = _b === void 0 ? 10 : _b, _c = _a.pageNumber, pageNumber = _c === void 0 ? 1 : _c, botId = _a.botId, rest = __rest(_a, ["pageSize", "pageNumber", "botId"]);
        return __awaiter(this, void 0, void 0, function () {
            var offset, limit;
            return __generator(this, function (_d) {
                if (pageNumber < 1)
                    throw new Error('pageNumber must be greater than 0');
                offset = pageSize * (pageNumber - 1);
                limit = pageSize;
                return [2, this.req({
                        method: 'get',
                        url: this.join("bots/".concat(botId, "/conversation")),
                        data: __assign(__assign({}, rest), { offset: offset, limit: limit }),
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.deleteConversation = function (_a, options) {
        var botId = _a.botId, conversationId = _a.conversationId, rest = __rest(_a, ["botId", "conversationId"]);
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2, this.req({
                        method: 'delete',
                        url: this.join("bots/".concat(botId, "/conversation/").concat(conversationId)),
                        data: rest,
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.speechToText = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.req({
                        method: 'post',
                        url: this.join("bots/".concat(props.botId, "/speech-to-text")),
                        data: props,
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.textToSpeech = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.req({
                        method: 'post',
                        url: this.join("bots/".concat(props.botId, "/text-to-speech")),
                        data: props,
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.getTextToSpeechResult = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.req({
                        method: 'get',
                        url: this.join("bots/".concat(props.botId, "/text-to-speech")),
                        data: props,
                        timeout: options === null || options === void 0 ? void 0 : options.timeout,
                    })];
            });
        });
    };
    Bot.prototype.getRecommendQuestions = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.req({
                            method: 'post',
                            url: this.join("bots/".concat(props.botId, "/recommend-questions")),
                            data: props,
                            stream: true,
                            timeout: options === null || options === void 0 ? void 0 : options.timeout,
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, new StreamResult(res)];
                }
            });
        });
    };
    Bot.prototype.generateBot = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.req({
                            method: 'post',
                            url: this.join('generate-bot'),
                            data: props,
                            stream: true,
                            timeout: options === null || options === void 0 ? void 0 : options.timeout,
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, new StreamResult(res)];
                }
            });
        });
    };
    Bot.prototype.getPreview = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.req({
                            method: 'post',
                            url: this.join('preview'),
                            data: props,
                            stream: true,
                            timeout: options === null || options === void 0 ? void 0 : options.timeout,
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, new StreamResult(res)];
                }
            });
        });
    };
    Bot.prototype.generateImage = function (props, options) {
        return this.req({ method: 'post', url: this.join('generate-image'), data: props, timeout: options === null || options === void 0 ? void 0 : options.timeout });
    };
    Bot.prototype.sendMessage = function (props, options) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.req({
                            method: 'post',
                            url: this.join("bots/".concat(props.botId, "/send-message")),
                            data: props,
                            stream: true,
                            timeout: options === null || options === void 0 ? void 0 : options.timeout,
                        })];
                    case 1:
                        res = _a.sent();
                        return [2, new StreamResult(res)];
                }
            });
        });
    };
    Bot.prototype.join = function (url) {
        return "".concat(this.baseUrl, "/").concat(url);
    };
    return Bot;
}());
exports.Bot = Bot;
var StreamResult = (function () {
    function StreamResult(_stream) {
        var stream = (0, utils_1.toPolyfillReadable)(_stream);
        this._eventSourceStream = stream
            .pipeThrough(new utils_1.TextDecoderStream())
            .pipeThrough((0, utils_1.createEventSourceParserTransformStream)());
    }
    Object.defineProperty(StreamResult.prototype, "teeedStream", {
        get: function () {
            var _a = this._eventSourceStream.tee(), s1 = _a[0], s2 = _a[1];
            this._eventSourceStream = s2;
            return s1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StreamResult.prototype, "eventSourceStream", {
        get: function () {
            return (0, utils_1.createAsyncIterable)(this.teeedStream);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StreamResult.prototype, "dataStream", {
        get: function () {
            return (0, utils_1.createAsyncIterable)(this.eventSourceStream.pipeThrough(new utils_1.TransformStream({
                transform: function (chunk, controller) {
                    try {
                        var data = JSON.parse(chunk.data);
                        controller.enqueue(data);
                    }
                    catch (e) {
                        if (chunk.data !== '[DONE]') {
                            console.warn('Error when transforming event source data to json', e, chunk);
                        }
                        else {
                            controller.terminate();
                        }
                    }
                },
            })));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StreamResult.prototype, "textStream", {
        get: function () {
            return (0, utils_1.createAsyncIterable)(this.dataStream.pipeThrough(new utils_1.TransformStream({
                transform: function (chunk, controller) {
                    var _a;
                    controller.enqueue((_a = chunk === null || chunk === void 0 ? void 0 : chunk.content) !== null && _a !== void 0 ? _a : '');
                },
            })));
        },
        enumerable: false,
        configurable: true
    });
    return StreamResult;
}());
