// @cloudbase/agent-tools/dist/mcp/index.js
"use strict";
/**
 * MCP (Model Context Protocol) Integration Module for AG-Kit
 *
 * This module provides integration between AG-Kit's tool system and the
 * Model Context Protocol, allowing AG-Kit tools to work with MCP clients
 * and enabling AG-Kit to use external MCP tools.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPClientManager = exports.createMCPToolkit = exports.MCPToolkit = exports.MemoryTransportRegistry = exports.memoryTransportRegistry = exports.convertAGKitToolToMCPMetadata = exports.zodSchemaToMCPSchema = exports.createAGKitMCPServer = exports.AGKitMCPServer = exports.MCPClientTool = void 0;
// Core MCP integration
var client_tool_1 = require("./client-tool");
Object.defineProperty(exports, "MCPClientTool", { enumerable: true, get: function () { return client_tool_1.MCPClientTool; } });
var agkit_mcp_server_1 = require("./agkit-mcp-server");
Object.defineProperty(exports, "AGKitMCPServer", { enumerable: true, get: function () { return agkit_mcp_server_1.AGKitMCPServer; } });
Object.defineProperty(exports, "createAGKitMCPServer", { enumerable: true, get: function () { return agkit_mcp_server_1.createAGKitMCPServer; } });
Object.defineProperty(exports, "zodSchemaToMCPSchema", { enumerable: true, get: function () { return agkit_mcp_server_1.zodSchemaToMCPSchema; } });
Object.defineProperty(exports, "convertAGKitToolToMCPMetadata", { enumerable: true, get: function () { return agkit_mcp_server_1.convertAGKitToolToMCPMetadata; } });
Object.defineProperty(exports, "memoryTransportRegistry", { enumerable: true, get: function () { return agkit_mcp_server_1.memoryTransportRegistry; } });
Object.defineProperty(exports, "MemoryTransportRegistry", { enumerable: true, get: function () { return agkit_mcp_server_1.MemoryTransportRegistry; } });
var mcp_toolkit_1 = require("./mcp-toolkit");
Object.defineProperty(exports, "MCPToolkit", { enumerable: true, get: function () { return mcp_toolkit_1.MCPToolkit; } });
Object.defineProperty(exports, "createMCPToolkit", { enumerable: true, get: function () { return mcp_toolkit_1.createMCPToolkit; } });
var mcp_client_manager_1 = require("./mcp-client-manager");
Object.defineProperty(exports, "MCPClientManager", { enumerable: true, get: function () { return mcp_client_manager_1.MCPClientManager; } });
