// cloudflare/resources/dns/zone-transfers/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneTransfers = exports.TSIGs = exports.TSIGsSinglePage = exports.Peers = exports.PeersSinglePage = exports.OutgoingResource = exports.IncomingResource = exports.ForceAXFRResource = exports.ACLs = exports.ACLsSinglePage = void 0;
var acls_1 = require("./acls.js");
Object.defineProperty(exports, "ACLsSinglePage", { enumerable: true, get: function () { return acls_1.ACLsSinglePage; } });
Object.defineProperty(exports, "ACLs", { enumerable: true, get: function () { return acls_1.ACLs; } });
var force_axfr_1 = require("./force-axfr.js");
Object.defineProperty(exports, "ForceAXFRResource", { enumerable: true, get: function () { return force_axfr_1.ForceAXFRResource; } });
var incoming_1 = require("./incoming.js");
Object.defineProperty(exports, "IncomingResource", { enumerable: true, get: function () { return incoming_1.IncomingResource; } });
var index_1 = require("./outgoing/index.js");
Object.defineProperty(exports, "OutgoingResource", { enumerable: true, get: function () { return index_1.OutgoingResource; } });
var peers_1 = require("./peers.js");
Object.defineProperty(exports, "PeersSinglePage", { enumerable: true, get: function () { return peers_1.PeersSinglePage; } });
Object.defineProperty(exports, "Peers", { enumerable: true, get: function () { return peers_1.Peers; } });
var tsigs_1 = require("./tsigs.js");
Object.defineProperty(exports, "TSIGsSinglePage", { enumerable: true, get: function () { return tsigs_1.TSIGsSinglePage; } });
Object.defineProperty(exports, "TSIGs", { enumerable: true, get: function () { return tsigs_1.TSIGs; } });
var zone_transfers_1 = require("./zone-transfers.js");
Object.defineProperty(exports, "ZoneTransfers", { enumerable: true, get: function () { return zone_transfers_1.ZoneTransfers; } });
