// @mistralai/mistralai/src/models/operations/agentsapiv1agentsdelete.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type AgentsApiV1AgentsDeleteRequest = {
  agentId: string;
};

/** @internal */
export const AgentsApiV1AgentsDeleteRequest$inboundSchema: z.ZodType<
  AgentsApiV1AgentsDeleteRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  agent_id: z.string(),
}).transform((v) => {
  return remap$(v, {
    "agent_id": "agentId",
  });
});

/** @internal */
export type AgentsApiV1AgentsDeleteRequest$Outbound = {
  agent_id: string;
};

/** @internal */
export const AgentsApiV1AgentsDeleteRequest$outboundSchema: z.ZodType<
  AgentsApiV1AgentsDeleteRequest$Outbound,
  z.ZodTypeDef,
  AgentsApiV1AgentsDeleteRequest
> = z.object({
  agentId: z.string(),
}).transform((v) => {
  return remap$(v, {
    agentId: "agent_id",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace AgentsApiV1AgentsDeleteRequest$ {
  /** @deprecated use `AgentsApiV1AgentsDeleteRequest$inboundSchema` instead. */
  export const inboundSchema = AgentsApiV1AgentsDeleteRequest$inboundSchema;
  /** @deprecated use `AgentsApiV1AgentsDeleteRequest$outboundSchema` instead. */
  export const outboundSchema = AgentsApiV1AgentsDeleteRequest$outboundSchema;
  /** @deprecated use `AgentsApiV1AgentsDeleteRequest$Outbound` instead. */
  export type Outbound = AgentsApiV1AgentsDeleteRequest$Outbound;
}

export function agentsApiV1AgentsDeleteRequestToJSON(
  agentsApiV1AgentsDeleteRequest: AgentsApiV1AgentsDeleteRequest,
): string {
  return JSON.stringify(
    AgentsApiV1AgentsDeleteRequest$outboundSchema.parse(
      agentsApiV1AgentsDeleteRequest,
    ),
  );
}

export function agentsApiV1AgentsDeleteRequestFromJSON(
  jsonString: string,
): SafeParseResult<AgentsApiV1AgentsDeleteRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => AgentsApiV1AgentsDeleteRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'AgentsApiV1AgentsDeleteRequest' from JSON`,
  );
}
