// egg-errors/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// base error
var base_error_1 = require("./base_error");
Object.defineProperty(exports, "EggBaseError", { enumerable: true, get: function () { return base_error_1.default; } });
var base_exception_1 = require("./base_exception");
Object.defineProperty(exports, "EggBaseException", { enumerable: true, get: function () { return base_exception_1.default; } });
var error_options_1 = require("./error_options");
Object.defineProperty(exports, "ErrorOptions", { enumerable: true, get: function () { return error_options_1.default; } });
var error_type_1 = require("./error_type");
Object.defineProperty(exports, "ErrorType", { enumerable: true, get: function () { return error_type_1.default; } });
// error and exception
var error_1 = require("./error");
Object.defineProperty(exports, "EggError", { enumerable: true, get: function () { return error_1.default; } });
var exception_1 = require("./exception");
Object.defineProperty(exports, "EggException", { enumerable: true, get: function () { return exception_1.default; } });
// framework error and formatter
var framework_base_error_1 = require("./framework/framework_base_error");
Object.defineProperty(exports, "FrameworkBaseError", { enumerable: true, get: function () { return framework_base_error_1.FrameworkBaseError; } });
var formatter_1 = require("./framework/formatter");
Object.defineProperty(exports, "FrameworkErrorFormater", { enumerable: true, get: function () { return formatter_1.FrameworkErrorFormater; } });
// http error
var http_error_1 = require("./http/http_error");
Object.defineProperty(exports, "HttpError", { enumerable: true, get: function () { return http_error_1.default; } });
// http error 400 ~ 511
var _400_1 = require("./http/400");
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return _400_1.default; } });
Object.defineProperty(exports, "E400", { enumerable: true, get: function () { return _400_1.default; } });
var _401_1 = require("./http/401");
Object.defineProperty(exports, "UnauthorizedError", { enumerable: true, get: function () { return _401_1.default; } });
Object.defineProperty(exports, "E401", { enumerable: true, get: function () { return _401_1.default; } });
var _402_1 = require("./http/402");
Object.defineProperty(exports, "PaymentRequiredError", { enumerable: true, get: function () { return _402_1.default; } });
Object.defineProperty(exports, "E402", { enumerable: true, get: function () { return _402_1.default; } });
var _403_1 = require("./http/403");
Object.defineProperty(exports, "ForbiddenError", { enumerable: true, get: function () { return _403_1.default; } });
Object.defineProperty(exports, "E403", { enumerable: true, get: function () { return _403_1.default; } });
var _404_1 = require("./http/404");
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return _404_1.default; } });
Object.defineProperty(exports, "E404", { enumerable: true, get: function () { return _404_1.default; } });
var _405_1 = require("./http/405");
Object.defineProperty(exports, "MethodNotAllowedError", { enumerable: true, get: function () { return _405_1.default; } });
Object.defineProperty(exports, "E405", { enumerable: true, get: function () { return _405_1.default; } });
var _406_1 = require("./http/406");
Object.defineProperty(exports, "NotAcceptableError", { enumerable: true, get: function () { return _406_1.default; } });
Object.defineProperty(exports, "E406", { enumerable: true, get: function () { return _406_1.default; } });
var _407_1 = require("./http/407");
Object.defineProperty(exports, "ProxyAuthenticationRequiredError", { enumerable: true, get: function () { return _407_1.default; } });
Object.defineProperty(exports, "E407", { enumerable: true, get: function () { return _407_1.default; } });
var _408_1 = require("./http/408");
Object.defineProperty(exports, "RequestTimeoutError", { enumerable: true, get: function () { return _408_1.default; } });
Object.defineProperty(exports, "E408", { enumerable: true, get: function () { return _408_1.default; } });
var _409_1 = require("./http/409");
Object.defineProperty(exports, "ConflictError", { enumerable: true, get: function () { return _409_1.default; } });
Object.defineProperty(exports, "E409", { enumerable: true, get: function () { return _409_1.default; } });
var _410_1 = require("./http/410");
Object.defineProperty(exports, "GoneError", { enumerable: true, get: function () { return _410_1.default; } });
Object.defineProperty(exports, "E410", { enumerable: true, get: function () { return _410_1.default; } });
var _411_1 = require("./http/411");
Object.defineProperty(exports, "LengthRequiredError", { enumerable: true, get: function () { return _411_1.default; } });
Object.defineProperty(exports, "E411", { enumerable: true, get: function () { return _411_1.default; } });
var _412_1 = require("./http/412");
Object.defineProperty(exports, "PreconditionFailedError", { enumerable: true, get: function () { return _412_1.default; } });
Object.defineProperty(exports, "E412", { enumerable: true, get: function () { return _412_1.default; } });
var _413_1 = require("./http/413");
Object.defineProperty(exports, "PayloadTooLargeError", { enumerable: true, get: function () { return _413_1.default; } });
Object.defineProperty(exports, "E413", { enumerable: true, get: function () { return _413_1.default; } });
var _414_1 = require("./http/414");
Object.defineProperty(exports, "URITooLongError", { enumerable: true, get: function () { return _414_1.default; } });
Object.defineProperty(exports, "E414", { enumerable: true, get: function () { return _414_1.default; } });
var _415_1 = require("./http/415");
Object.defineProperty(exports, "UnsupportedMediaTypeError", { enumerable: true, get: function () { return _415_1.default; } });
Object.defineProperty(exports, "E415", { enumerable: true, get: function () { return _415_1.default; } });
var _416_1 = require("./http/416");
Object.defineProperty(exports, "RangeNotSatisfiableError", { enumerable: true, get: function () { return _416_1.default; } });
Object.defineProperty(exports, "E416", { enumerable: true, get: function () { return _416_1.default; } });
var _417_1 = require("./http/417");
Object.defineProperty(exports, "ExpectationFailedError", { enumerable: true, get: function () { return _417_1.default; } });
Object.defineProperty(exports, "E417", { enumerable: true, get: function () { return _417_1.default; } });
var _418_1 = require("./http/418");
Object.defineProperty(exports, "ImATeapotError", { enumerable: true, get: function () { return _418_1.default; } });
Object.defineProperty(exports, "E418", { enumerable: true, get: function () { return _418_1.default; } });
var _421_1 = require("./http/421");
Object.defineProperty(exports, "MisdirectedRequestError", { enumerable: true, get: function () { return _421_1.default; } });
Object.defineProperty(exports, "E421", { enumerable: true, get: function () { return _421_1.default; } });
var _422_1 = require("./http/422");
Object.defineProperty(exports, "UnprocessableEntityError", { enumerable: true, get: function () { return _422_1.default; } });
Object.defineProperty(exports, "E422", { enumerable: true, get: function () { return _422_1.default; } });
var _423_1 = require("./http/423");
Object.defineProperty(exports, "LockedError", { enumerable: true, get: function () { return _423_1.default; } });
Object.defineProperty(exports, "E423", { enumerable: true, get: function () { return _423_1.default; } });
var _424_1 = require("./http/424");
Object.defineProperty(exports, "FailedDependencyError", { enumerable: true, get: function () { return _424_1.default; } });
Object.defineProperty(exports, "E424", { enumerable: true, get: function () { return _424_1.default; } });
var _425_1 = require("./http/425");
Object.defineProperty(exports, "UnorderedCollectionError", { enumerable: true, get: function () { return _425_1.default; } });
Object.defineProperty(exports, "E425", { enumerable: true, get: function () { return _425_1.default; } });
var _426_1 = require("./http/426");
Object.defineProperty(exports, "UpgradeRequiredError", { enumerable: true, get: function () { return _426_1.default; } });
Object.defineProperty(exports, "E426", { enumerable: true, get: function () { return _426_1.default; } });
var _428_1 = require("./http/428");
Object.defineProperty(exports, "PreconditionRequiredError", { enumerable: true, get: function () { return _428_1.default; } });
Object.defineProperty(exports, "E428", { enumerable: true, get: function () { return _428_1.default; } });
var _429_1 = require("./http/429");
Object.defineProperty(exports, "TooManyRequestsError", { enumerable: true, get: function () { return _429_1.default; } });
Object.defineProperty(exports, "E429", { enumerable: true, get: function () { return _429_1.default; } });
var _431_1 = require("./http/431");
Object.defineProperty(exports, "RequestHeaderFieldsTooLargeError", { enumerable: true, get: function () { return _431_1.default; } });
Object.defineProperty(exports, "E431", { enumerable: true, get: function () { return _431_1.default; } });
var _451_1 = require("./http/451");
Object.defineProperty(exports, "UnavailableForLegalReasonsError", { enumerable: true, get: function () { return _451_1.default; } });
Object.defineProperty(exports, "E451", { enumerable: true, get: function () { return _451_1.default; } });
var _500_1 = require("./http/500");
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return _500_1.default; } });
Object.defineProperty(exports, "E500", { enumerable: true, get: function () { return _500_1.default; } });
var _501_1 = require("./http/501");
Object.defineProperty(exports, "NotImplementedError", { enumerable: true, get: function () { return _501_1.default; } });
Object.defineProperty(exports, "E501", { enumerable: true, get: function () { return _501_1.default; } });
var _502_1 = require("./http/502");
Object.defineProperty(exports, "BadGatewayError", { enumerable: true, get: function () { return _502_1.default; } });
Object.defineProperty(exports, "E502", { enumerable: true, get: function () { return _502_1.default; } });
var _503_1 = require("./http/503");
Object.defineProperty(exports, "ServiceUnavailableError", { enumerable: true, get: function () { return _503_1.default; } });
Object.defineProperty(exports, "E503", { enumerable: true, get: function () { return _503_1.default; } });
var _504_1 = require("./http/504");
Object.defineProperty(exports, "GatewayTimeoutError", { enumerable: true, get: function () { return _504_1.default; } });
Object.defineProperty(exports, "E504", { enumerable: true, get: function () { return _504_1.default; } });
var _505_1 = require("./http/505");
Object.defineProperty(exports, "HTTPVersionNotSupportedError", { enumerable: true, get: function () { return _505_1.default; } });
Object.defineProperty(exports, "E505", { enumerable: true, get: function () { return _505_1.default; } });
var _506_1 = require("./http/506");
Object.defineProperty(exports, "VariantAlsoNegotiatesError", { enumerable: true, get: function () { return _506_1.default; } });
Object.defineProperty(exports, "E506", { enumerable: true, get: function () { return _506_1.default; } });
var _507_1 = require("./http/507");
Object.defineProperty(exports, "InsufficientStorageError", { enumerable: true, get: function () { return _507_1.default; } });
Object.defineProperty(exports, "E507", { enumerable: true, get: function () { return _507_1.default; } });
var _508_1 = require("./http/508");
Object.defineProperty(exports, "LoopDetectedError", { enumerable: true, get: function () { return _508_1.default; } });
Object.defineProperty(exports, "E508", { enumerable: true, get: function () { return _508_1.default; } });
var _509_1 = require("./http/509");
Object.defineProperty(exports, "BandwidthLimitExceededError", { enumerable: true, get: function () { return _509_1.default; } });
Object.defineProperty(exports, "E509", { enumerable: true, get: function () { return _509_1.default; } });
var _510_1 = require("./http/510");
Object.defineProperty(exports, "NotExtendedError", { enumerable: true, get: function () { return _510_1.default; } });
Object.defineProperty(exports, "E510", { enumerable: true, get: function () { return _510_1.default; } });
var _511_1 = require("./http/511");
Object.defineProperty(exports, "NetworkAuthenticationRequiredError", { enumerable: true, get: function () { return _511_1.default; } });
Object.defineProperty(exports, "E511", { enumerable: true, get: function () { return _511_1.default; } });
