// @cloudbase/agent-tools/dist/fs/operator/transaction/transaction-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionController = exports.TransactionExecutor = void 0;
exports.createTransactionExecutor = createTransactionExecutor;
exports.withTransaction = withTransaction;
const transaction_operator_1 = require("./transaction-operator");
/**
 * 事务执行器 - 提供便捷的事务操作方法
 */
class TransactionExecutor {
    constructor(baseOperator) {
        this.operations = [];
        this.transactionOperator = new transaction_operator_1.TransactionFileOperator(baseOperator);
    }
    /**
     * 获取事务操作器
     */
    getOperator() {
        return this.transactionOperator;
    }
    /**
     * 添加文件写入操作
     */
    writeFile(path, data, options) {
        this.operations.push(async (op) => {
            await op.writeFile(path, data, options);
        });
        return this;
    }
    /**
     * 添加目录创建操作
     */
    mkdir(path, options) {
        this.operations.push(async (op) => {
            await op.mkdir(path, options);
        });
        return this;
    }
    /**
     * 添加自定义操作
     */
    custom(operation) {
        this.operations.push(operation);
        return this;
    }
    /**
     * 执行所有操作
     */
    async execute() {
        await this.transactionOperator.beginTransaction();
        try {
            for (const operation of this.operations) {
                await operation(this.transactionOperator);
            }
            await this.transactionOperator.commitTransaction();
            return { success: true };
        }
        catch (error) {
            await this.transactionOperator.rollbackTransaction();
            return { success: false, error: error };
        }
        finally {
            this.operations = []; // 清理操作列表
        }
    }
    /**
     * 在事务中执行操作
     * @param operation 要执行的操作函数
     * @returns 操作结果
     */
    async executeInTransaction(operation) {
        await this.transactionOperator.beginTransaction();
        try {
            const result = await operation(this.transactionOperator);
            await this.transactionOperator.commitTransaction();
            return result;
        }
        catch (error) {
            await this.transactionOperator.rollbackTransaction();
            throw error;
        }
    }
    /**
     * 手动控制的事务 - 返回事务控制器
     */
    async createTransaction() {
        await this.transactionOperator.beginTransaction();
        return new TransactionController(this.transactionOperator);
    }
}
exports.TransactionExecutor = TransactionExecutor;
/**
 * 事务控制器 - 用于手动控制事务的提交和回滚
 */
class TransactionController {
    constructor(transactionOperator) {
        this.isCompleted = false;
        this.transactionOperator = transactionOperator;
    }
    /**
     * 获取事务操作器
     */
    getOperator() {
        if (this.isCompleted) {
            throw new Error("Transaction has been completed");
        }
        return this.transactionOperator;
    }
    /**
     * 提交事务
     */
    async commit() {
        if (this.isCompleted) {
            throw new Error("Transaction has already been completed");
        }
        await this.transactionOperator.commitTransaction();
        this.isCompleted = true;
    }
    /**
     * 回滚事务
     */
    async rollback() {
        if (this.isCompleted) {
            throw new Error("Transaction has already been completed");
        }
        await this.transactionOperator.rollbackTransaction();
        this.isCompleted = true;
    }
    /**
     * 检查事务是否已完成
     */
    isTransactionCompleted() {
        return this.isCompleted;
    }
    /**
     * 获取事务操作历史
     */
    getOperationHistory() {
        return this.transactionOperator.getTransactionOperations();
    }
}
exports.TransactionController = TransactionController;
/**
 * 创建事务执行器的便捷函数
 */
function createTransactionExecutor(baseOperator) {
    return new TransactionExecutor(baseOperator);
}
/**
 * 直接在事务中执行操作的便捷函数
 */
async function withTransaction(baseOperator, operation) {
    const executor = new TransactionExecutor(baseOperator);
    return executor.executeInTransaction(operation);
}
