// @cloudbase/agent-tools/dist/fs/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateWorkspacePath = validateWorkspacePath;
exports.resolveWorkspacePath = resolveWorkspacePath;
exports.createSuccessResponse = createSuccessResponse;
exports.validateRequiredString = validateRequiredString;
exports.validateFileExists = validateFileExists;
exports.validateDirectoryExists = validateDirectoryExists;
const path = __importStar(require("path"));
const utils_1 = require("../utils");
/**
 * Validate if a path is within the workspace directory (supports both absolute and relative paths)
 */
function validateWorkspacePath(filePath, context) {
    try {
        // Prevent directory traversal attacks
        if (filePath.includes("..")) {
            return (0, utils_1.handleToolError)('Path cannot contain ".." for security reasons', "Path validation", "security");
        }
        const normalizedWorkspace = path.normalize(path.resolve(context.workingDirectory));
        // Handle both absolute and relative paths
        let resolvedPath;
        if (path.isAbsolute(filePath)) {
            resolvedPath = path.normalize(filePath);
        }
        else {
            resolvedPath = path.resolve(context.workingDirectory, filePath);
        }
        // Check if path is within workspace boundary
        if (!resolvedPath.startsWith(normalizedWorkspace)) {
            return (0, utils_1.handleToolError)(`Path must be within workspace directory: ${filePath}`, "Security check", "security");
        }
        return null; // No error
    }
    catch (error) {
        return (0, utils_1.handleToolError)(error, "Path validation", "validation");
    }
}
/**
 * Safely resolve a file path (supports both absolute and relative paths)
 */
function resolveWorkspacePath(filePath, context) {
    if (path.isAbsolute(filePath)) {
        return path.normalize(filePath);
    }
    else {
        return path.resolve(context.workingDirectory, filePath);
    }
}
/**
 * Create a success response with typed data
 */
function createSuccessResponse(data) {
    return new utils_1.ToolResult({
        success: true,
        data,
    });
}
/**
 * Validation helper for required string parameters
 */
function validateRequiredString(value, paramName) {
    if (!value || typeof value !== "string" || value.trim() === "") {
        return (0, utils_1.handleToolError)(`${paramName} is required and must be a non-empty string`, "Parameter validation", "validation");
    }
    return null;
}
/**
 * Validation helper for file existence
 */
async function validateFileExists(context, absolutePath, filePath) {
    try {
        if (!(await context.fsOperator.exists(absolutePath))) {
            return (0, utils_1.handleToolError)(`File not found: ${filePath}`, "File existence check", "file_not_found");
        }
        return null;
    }
    catch (error) {
        return (0, utils_1.handleToolError)(error, "File existence check", "permission");
    }
}
/**
 * Validation helper for directory existence
 */
async function validateDirectoryExists(context, absolutePath, dirPath) {
    try {
        if (!(await context.fsOperator.exists(absolutePath))) {
            return (0, utils_1.handleToolError)(`Directory not found: ${dirPath}`, "Directory existence check", "file_not_found");
        }
        const stats = await context.fsOperator.stat(absolutePath);
        if (!stats.isDirectory()) {
            return (0, utils_1.handleToolError)(`Path is not a directory: ${dirPath}`, "Directory validation", "validation");
        }
        return null;
    }
    catch (error) {
        return (0, utils_1.handleToolError)(error, "Directory validation", "permission");
    }
}
