// @cloudbase/agent-tools/dist/code_executers/built_in_code_executor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuiltInCodeExecutor = void 0;
const code_interpreter_1 = require("@e2b/code-interpreter");
const base_code_executor_1 = require("./base_code_executor");
class BuiltInCodeExecutor extends base_code_executor_1.BaseCodeExecutor {
    constructor(fields, options) {
        super();
        this.sandbox = fields?.sandbox;
        this.options = options || {};
        // 如果没有传入 sandbox，则校验 apiKey 和 domain 参数
        if (!this.sandbox) {
            this.apiKey = fields?.apiKey || process.env.AG_KIT_SANDBOX_API_KEY || process.env.E2B_API_KEY;
            this.domain = fields?.domain || process.env.AG_KIT_SANDBOX_DOMAIN || process.env.E2B_DOMAIN;
            if (!this.apiKey) {
                throw new Error('API key is required. Please provide it via constructor parameter, AG_KIT_SANDBOX_API_KEY, or E2B_API_KEY environment variable.');
            }
            if (!this.domain) {
                throw new Error('Domain is required. Please provide it via constructor parameter, AG_KIT_SANDBOX_DOMAIN, or E2B_DOMAIN environment variable.');
            }
        }
    }
    async ensureSandbox() {
        if (!this.sandbox) {
            this.sandbox = await code_interpreter_1.Sandbox.create("code-interpreter-v1", {
                apiKey: this.apiKey,
                domain: this.domain,
                timeoutMs: 3600 * 1000,
            });
        }
        /* else if (!(await this.sandbox.isRunning())) {
          this.sandbox.sandboxId
          this.sandbox = await Sandbox.create();
        }*/
        return this.sandbox;
    }
    /**
     * 自定义实现 createCodeContext 方法
     * 参考 E2B sandbox.createCodeContext 的源码实现
     */
    async createCodeContext(opts) {
        try {
            // 获取 sandbox 的内部属性（这里需要根据实际的 E2B SDK 结构调整）
            const sandbox = await this.ensureSandbox();
            // 构建请求 URL - 需要根据实际的 E2B SDK 获取正确的 jupyterUrl 和 envdAccessToken
            const jupyterUrl = sandbox.jupyterUrl || `https://${this.domain}/jupyter`;
            const envdAccessToken = sandbox.envdAccessToken || sandbox.accessToken;
            const res = await fetch(`${jupyterUrl}/contexts`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "X-Access-Token": envdAccessToken,
                    // 添加其他可能需要的 headers
                    ...sandbox.connectionConfig?.headers,
                },
                body: JSON.stringify({
                    language: opts?.language,
                    cwd: opts?.cwd
                }),
                keepalive: true,
                signal: this.createTimeoutSignal(opts?.requestTimeoutMs)
            });
            if (!res.ok) {
                const errorText = await res.text();
                throw new Error(`Failed to create code context: ${res.status} ${res.statusText} - ${errorText}`);
            }
            const contextResponse = await res.json();
            // 转换为 E2B Context 格式
            return {
                id: contextResponse.id,
                language: contextResponse.language,
                cwd: contextResponse.cwd || process.cwd(), // 提供默认值以满足 Context 类型要求
            };
        }
        catch (error) {
            if (error.name === 'AbortError') {
                throw new Error(`Request timeout: Failed to create code context within ${opts?.requestTimeoutMs || 30000}ms`);
            }
            throw error;
        }
    }
    /**
     * 创建超时信号
     */
    createTimeoutSignal(timeoutMs) {
        if (!timeoutMs)
            return undefined;
        const controller = new AbortController();
        setTimeout(() => controller.abort(), timeoutMs);
        return controller.signal;
    }
    async _invoke(input) {
        let sandbox = await this.ensureSandbox();
        // 使用自定义的 createCodeContext 方法
        const context = await this.createCodeContext({
            language: input.language || "python",
        });
        const res = await sandbox.runCode(input.code, {
            ...this.options,
            context: context,
        });
        return res;
    }
}
exports.BuiltInCodeExecutor = BuiltInCodeExecutor;
