// langchain/dist/index.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const require_chat_models_universal = require('./chat_models/universal.cjs');
const require_errors = require('./agents/errors.cjs');
const require_responses = require('./agents/responses.cjs');
const require_utils = require('./agents/middleware/utils.cjs');
const require_types = require('./agents/middleware/types.cjs');
const require_middleware = require('./agents/middleware.cjs');
const require_utils$1 = require('./agents/tests/utils.cjs');
const require_index = require('./agents/index.cjs');
const require_hitl = require('./agents/middleware/hitl.cjs');
const require_summarization = require('./agents/middleware/summarization.cjs');
const require_dynamicSystemPrompt = require('./agents/middleware/dynamicSystemPrompt.cjs');
const require_llmToolSelector = require('./agents/middleware/llmToolSelector.cjs');
const require_pii = require('./agents/middleware/pii.cjs');
const require_piiRedaction = require('./agents/middleware/piiRedaction.cjs');
const require_contextEditing = require('./agents/middleware/contextEditing.cjs');
const require_toolCallLimit = require('./agents/middleware/toolCallLimit.cjs');
const require_todoListMiddleware = require('./agents/middleware/todoListMiddleware.cjs');
const require_modelCallLimit = require('./agents/middleware/modelCallLimit.cjs');
const require_modelFallback = require('./agents/middleware/modelFallback.cjs');
const require_modelRetry = require('./agents/middleware/modelRetry.cjs');
const require_toolRetry = require('./agents/middleware/toolRetry.cjs');
const require_toolEmulator = require('./agents/middleware/toolEmulator.cjs');
const require_moderation = require('./agents/middleware/provider/openai/moderation.cjs');
const require_promptCaching = require('./agents/middleware/provider/anthropic/promptCaching.cjs');
require('./agents/middleware/index.cjs');
const __langchain_core_messages = require_rolldown_runtime.__toESM(require("@langchain/core/messages"));
const __langchain_core_tools = require_rolldown_runtime.__toESM(require("@langchain/core/tools"));
const __langchain_core_utils_context = require_rolldown_runtime.__toESM(require("@langchain/core/utils/context"));
const __langchain_core_stores = require_rolldown_runtime.__toESM(require("@langchain/core/stores"));
const __langchain_core_documents = require_rolldown_runtime.__toESM(require("@langchain/core/documents"));

//#region src/index.ts
var src_exports = {};
require_rolldown_runtime.__export(src_exports, {
	AIMessage: () => __langchain_core_messages.AIMessage,
	AIMessageChunk: () => __langchain_core_messages.AIMessageChunk,
	BaseMessage: () => __langchain_core_messages.BaseMessage,
	BaseMessageChunk: () => __langchain_core_messages.BaseMessageChunk,
	ClearToolUsesEdit: () => require_contextEditing.ClearToolUsesEdit,
	Document: () => __langchain_core_documents.Document,
	DynamicStructuredTool: () => __langchain_core_tools.DynamicStructuredTool,
	DynamicTool: () => __langchain_core_tools.DynamicTool,
	FakeToolCallingModel: () => require_utils$1.FakeToolCallingModel,
	HumanMessage: () => __langchain_core_messages.HumanMessage,
	HumanMessageChunk: () => __langchain_core_messages.HumanMessageChunk,
	InMemoryStore: () => __langchain_core_stores.InMemoryStore,
	MIDDLEWARE_BRAND: () => require_types.MIDDLEWARE_BRAND,
	MultipleStructuredOutputsError: () => require_errors.MultipleStructuredOutputsError,
	MultipleToolsBoundError: () => require_errors.MultipleToolsBoundError,
	PIIDetectionError: () => require_pii.PIIDetectionError,
	ProviderStrategy: () => require_responses.ProviderStrategy,
	StructuredOutputParsingError: () => require_errors.StructuredOutputParsingError,
	StructuredTool: () => __langchain_core_tools.StructuredTool,
	SystemMessage: () => __langchain_core_messages.SystemMessage,
	SystemMessageChunk: () => __langchain_core_messages.SystemMessageChunk,
	TODO_LIST_MIDDLEWARE_SYSTEM_PROMPT: () => require_todoListMiddleware.TODO_LIST_MIDDLEWARE_SYSTEM_PROMPT,
	Tool: () => __langchain_core_tools.Tool,
	ToolCallLimitExceededError: () => require_toolCallLimit.ToolCallLimitExceededError,
	ToolInvocationError: () => require_errors.ToolInvocationError,
	ToolMessage: () => __langchain_core_messages.ToolMessage,
	ToolMessageChunk: () => __langchain_core_messages.ToolMessageChunk,
	ToolStrategy: () => require_responses.ToolStrategy,
	anthropicPromptCachingMiddleware: () => require_promptCaching.anthropicPromptCachingMiddleware,
	applyStrategy: () => require_pii.applyStrategy,
	context: () => __langchain_core_utils_context.context,
	contextEditingMiddleware: () => require_contextEditing.contextEditingMiddleware,
	countTokensApproximately: () => require_utils.countTokensApproximately,
	createAgent: () => require_index.createAgent,
	createMiddleware: () => require_middleware.createMiddleware,
	detectCreditCard: () => require_pii.detectCreditCard,
	detectEmail: () => require_pii.detectEmail,
	detectIP: () => require_pii.detectIP,
	detectMacAddress: () => require_pii.detectMacAddress,
	detectUrl: () => require_pii.detectUrl,
	dynamicSystemPromptMiddleware: () => require_dynamicSystemPrompt.dynamicSystemPromptMiddleware,
	filterMessages: () => __langchain_core_messages.filterMessages,
	humanInTheLoopMiddleware: () => require_hitl.humanInTheLoopMiddleware,
	initChatModel: () => require_chat_models_universal.initChatModel,
	llmToolSelectorMiddleware: () => require_llmToolSelector.llmToolSelectorMiddleware,
	modelCallLimitMiddleware: () => require_modelCallLimit.modelCallLimitMiddleware,
	modelFallbackMiddleware: () => require_modelFallback.modelFallbackMiddleware,
	modelRetryMiddleware: () => require_modelRetry.modelRetryMiddleware,
	openAIModerationMiddleware: () => require_moderation.openAIModerationMiddleware,
	piiMiddleware: () => require_pii.piiMiddleware,
	piiRedactionMiddleware: () => require_piiRedaction.piiRedactionMiddleware,
	providerStrategy: () => require_responses.providerStrategy,
	resolveRedactionRule: () => require_pii.resolveRedactionRule,
	summarizationMiddleware: () => require_summarization.summarizationMiddleware,
	todoListMiddleware: () => require_todoListMiddleware.todoListMiddleware,
	tool: () => __langchain_core_tools.tool,
	toolCallLimitMiddleware: () => require_toolCallLimit.toolCallLimitMiddleware,
	toolEmulatorMiddleware: () => require_toolEmulator.toolEmulatorMiddleware,
	toolRetryMiddleware: () => require_toolRetry.toolRetryMiddleware,
	toolStrategy: () => require_responses.toolStrategy,
	trimMessages: () => __langchain_core_messages.trimMessages
});

//#endregion
Object.defineProperty(exports, 'AIMessage', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.AIMessage;
  }
});
Object.defineProperty(exports, 'AIMessageChunk', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.AIMessageChunk;
  }
});
Object.defineProperty(exports, 'BaseMessage', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.BaseMessage;
  }
});
Object.defineProperty(exports, 'BaseMessageChunk', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.BaseMessageChunk;
  }
});
exports.ClearToolUsesEdit = require_contextEditing.ClearToolUsesEdit;
Object.defineProperty(exports, 'Document', {
  enumerable: true,
  get: function () {
    return __langchain_core_documents.Document;
  }
});
Object.defineProperty(exports, 'DynamicStructuredTool', {
  enumerable: true,
  get: function () {
    return __langchain_core_tools.DynamicStructuredTool;
  }
});
Object.defineProperty(exports, 'DynamicTool', {
  enumerable: true,
  get: function () {
    return __langchain_core_tools.DynamicTool;
  }
});
exports.FakeToolCallingModel = require_utils$1.FakeToolCallingModel;
Object.defineProperty(exports, 'HumanMessage', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.HumanMessage;
  }
});
Object.defineProperty(exports, 'HumanMessageChunk', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.HumanMessageChunk;
  }
});
Object.defineProperty(exports, 'InMemoryStore', {
  enumerable: true,
  get: function () {
    return __langchain_core_stores.InMemoryStore;
  }
});
exports.MIDDLEWARE_BRAND = require_types.MIDDLEWARE_BRAND;
exports.MultipleStructuredOutputsError = require_errors.MultipleStructuredOutputsError;
exports.MultipleToolsBoundError = require_errors.MultipleToolsBoundError;
exports.PIIDetectionError = require_pii.PIIDetectionError;
exports.ProviderStrategy = require_responses.ProviderStrategy;
exports.StructuredOutputParsingError = require_errors.StructuredOutputParsingError;
Object.defineProperty(exports, 'StructuredTool', {
  enumerable: true,
  get: function () {
    return __langchain_core_tools.StructuredTool;
  }
});
Object.defineProperty(exports, 'SystemMessage', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.SystemMessage;
  }
});
Object.defineProperty(exports, 'SystemMessageChunk', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.SystemMessageChunk;
  }
});
exports.TODO_LIST_MIDDLEWARE_SYSTEM_PROMPT = require_todoListMiddleware.TODO_LIST_MIDDLEWARE_SYSTEM_PROMPT;
Object.defineProperty(exports, 'Tool', {
  enumerable: true,
  get: function () {
    return __langchain_core_tools.Tool;
  }
});
exports.ToolCallLimitExceededError = require_toolCallLimit.ToolCallLimitExceededError;
exports.ToolInvocationError = require_errors.ToolInvocationError;
Object.defineProperty(exports, 'ToolMessage', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.ToolMessage;
  }
});
Object.defineProperty(exports, 'ToolMessageChunk', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.ToolMessageChunk;
  }
});
exports.ToolStrategy = require_responses.ToolStrategy;
exports.anthropicPromptCachingMiddleware = require_promptCaching.anthropicPromptCachingMiddleware;
exports.applyStrategy = require_pii.applyStrategy;
Object.defineProperty(exports, 'context', {
  enumerable: true,
  get: function () {
    return __langchain_core_utils_context.context;
  }
});
exports.contextEditingMiddleware = require_contextEditing.contextEditingMiddleware;
exports.countTokensApproximately = require_utils.countTokensApproximately;
exports.createAgent = require_index.createAgent;
exports.createMiddleware = require_middleware.createMiddleware;
exports.detectCreditCard = require_pii.detectCreditCard;
exports.detectEmail = require_pii.detectEmail;
exports.detectIP = require_pii.detectIP;
exports.detectMacAddress = require_pii.detectMacAddress;
exports.detectUrl = require_pii.detectUrl;
exports.dynamicSystemPromptMiddleware = require_dynamicSystemPrompt.dynamicSystemPromptMiddleware;
Object.defineProperty(exports, 'filterMessages', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.filterMessages;
  }
});
exports.humanInTheLoopMiddleware = require_hitl.humanInTheLoopMiddleware;
exports.initChatModel = require_chat_models_universal.initChatModel;
exports.llmToolSelectorMiddleware = require_llmToolSelector.llmToolSelectorMiddleware;
exports.modelCallLimitMiddleware = require_modelCallLimit.modelCallLimitMiddleware;
exports.modelFallbackMiddleware = require_modelFallback.modelFallbackMiddleware;
exports.modelRetryMiddleware = require_modelRetry.modelRetryMiddleware;
exports.openAIModerationMiddleware = require_moderation.openAIModerationMiddleware;
exports.piiMiddleware = require_pii.piiMiddleware;
exports.piiRedactionMiddleware = require_piiRedaction.piiRedactionMiddleware;
exports.providerStrategy = require_responses.providerStrategy;
exports.resolveRedactionRule = require_pii.resolveRedactionRule;
Object.defineProperty(exports, 'src_exports', {
  enumerable: true,
  get: function () {
    return src_exports;
  }
});
exports.summarizationMiddleware = require_summarization.summarizationMiddleware;
exports.todoListMiddleware = require_todoListMiddleware.todoListMiddleware;
Object.defineProperty(exports, 'tool', {
  enumerable: true,
  get: function () {
    return __langchain_core_tools.tool;
  }
});
exports.toolCallLimitMiddleware = require_toolCallLimit.toolCallLimitMiddleware;
exports.toolEmulatorMiddleware = require_toolEmulator.toolEmulatorMiddleware;
exports.toolRetryMiddleware = require_toolRetry.toolRetryMiddleware;
exports.toolStrategy = require_responses.toolStrategy;
Object.defineProperty(exports, 'trimMessages', {
  enumerable: true,
  get: function () {
    return __langchain_core_messages.trimMessages;
  }
});
