// @mistralai/mistralai/src/models/components/audioformat.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  AudioEncoding,
  AudioEncoding$inboundSchema,
  AudioEncoding$outboundSchema,
} from "./audioencoding.js";

export type AudioFormat = {
  encoding: AudioEncoding;
  sampleRate: number;
};

/** @internal */
export const AudioFormat$inboundSchema: z.ZodType<
  AudioFormat,
  z.ZodTypeDef,
  unknown
> = z.object({
  encoding: AudioEncoding$inboundSchema,
  sample_rate: z.number().int(),
}).transform((v) => {
  return remap$(v, {
    "sample_rate": "sampleRate",
  });
});

/** @internal */
export type AudioFormat$Outbound = {
  encoding: string;
  sample_rate: number;
};

/** @internal */
export const AudioFormat$outboundSchema: z.ZodType<
  AudioFormat$Outbound,
  z.ZodTypeDef,
  AudioFormat
> = z.object({
  encoding: AudioEncoding$outboundSchema,
  sampleRate: z.number().int(),
}).transform((v) => {
  return remap$(v, {
    sampleRate: "sample_rate",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace AudioFormat$ {
  /** @deprecated use `AudioFormat$inboundSchema` instead. */
  export const inboundSchema = AudioFormat$inboundSchema;
  /** @deprecated use `AudioFormat$outboundSchema` instead. */
  export const outboundSchema = AudioFormat$outboundSchema;
  /** @deprecated use `AudioFormat$Outbound` instead. */
  export type Outbound = AudioFormat$Outbound;
}

export function audioFormatToJSON(audioFormat: AudioFormat): string {
  return JSON.stringify(AudioFormat$outboundSchema.parse(audioFormat));
}

export function audioFormatFromJSON(
  jsonString: string,
): SafeParseResult<AudioFormat, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => AudioFormat$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'AudioFormat' from JSON`,
  );
}
