// @mistralai/mistralai/src/models/components/audioencoding.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import {
  catchUnrecognizedEnum,
  OpenEnum,
  Unrecognized,
} from "../../types/enums.js";

export const AudioEncoding = {
  PcmS16le: "pcm_s16le",
  PcmS32le: "pcm_s32le",
  PcmF16le: "pcm_f16le",
  PcmF32le: "pcm_f32le",
  PcmMulaw: "pcm_mulaw",
  PcmAlaw: "pcm_alaw",
} as const;
export type AudioEncoding = OpenEnum<typeof AudioEncoding>;

/** @internal */
export const AudioEncoding$inboundSchema: z.ZodType<
  AudioEncoding,
  z.ZodTypeDef,
  unknown
> = z
  .union([
    z.nativeEnum(AudioEncoding),
    z.string().transform(catchUnrecognizedEnum),
  ]);

/** @internal */
export const AudioEncoding$outboundSchema: z.ZodType<
  AudioEncoding,
  z.ZodTypeDef,
  AudioEncoding
> = z.union([
  z.nativeEnum(AudioEncoding),
  z.string().and(z.custom<Unrecognized<string>>()),
]);

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace AudioEncoding$ {
  /** @deprecated use `AudioEncoding$inboundSchema` instead. */
  export const inboundSchema = AudioEncoding$inboundSchema;
  /** @deprecated use `AudioEncoding$outboundSchema` instead. */
  export const outboundSchema = AudioEncoding$outboundSchema;
}
