// @mistralai/mistralai/src/models/components/batchrequest.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type BatchRequest = {
  customId?: string | null | undefined;
  body: { [k: string]: any };
};

/** @internal */
export const BatchRequest$inboundSchema: z.ZodType<
  BatchRequest,
  z.ZodTypeDef,
  unknown
> = z.object({
  custom_id: z.nullable(z.string()).optional(),
  body: z.record(z.any()),
}).transform((v) => {
  return remap$(v, {
    "custom_id": "customId",
  });
});

/** @internal */
export type BatchRequest$Outbound = {
  custom_id?: string | null | undefined;
  body: { [k: string]: any };
};

/** @internal */
export const BatchRequest$outboundSchema: z.ZodType<
  BatchRequest$Outbound,
  z.ZodTypeDef,
  BatchRequest
> = z.object({
  customId: z.nullable(z.string()).optional(),
  body: z.record(z.any()),
}).transform((v) => {
  return remap$(v, {
    customId: "custom_id",
  });
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace BatchRequest$ {
  /** @deprecated use `BatchRequest$inboundSchema` instead. */
  export const inboundSchema = BatchRequest$inboundSchema;
  /** @deprecated use `BatchRequest$outboundSchema` instead. */
  export const outboundSchema = BatchRequest$outboundSchema;
  /** @deprecated use `BatchRequest$Outbound` instead. */
  export type Outbound = BatchRequest$Outbound;
}

export function batchRequestToJSON(batchRequest: BatchRequest): string {
  return JSON.stringify(BatchRequest$outboundSchema.parse(batchRequest));
}

export function batchRequestFromJSON(
  jsonString: string,
): SafeParseResult<BatchRequest, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => BatchRequest$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'BatchRequest' from JSON`,
  );
}
