// cloudflare/resources/magic-transit/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Sites = exports.SitesSinglePage = exports.Routes = exports.PCAPs = exports.PCAPListResponsesSinglePage = exports.MagicTransit = exports.IPSECTunnels = exports.GRETunnels = exports.Connectors = exports.ConnectorListResponsesSinglePage = exports.CfInterconnects = exports.Apps = exports.AppListResponsesSinglePage = void 0;
var apps_1 = require("./apps.js");
Object.defineProperty(exports, "AppListResponsesSinglePage", { enumerable: true, get: function () { return apps_1.AppListResponsesSinglePage; } });
Object.defineProperty(exports, "Apps", { enumerable: true, get: function () { return apps_1.Apps; } });
var cf_interconnects_1 = require("./cf-interconnects.js");
Object.defineProperty(exports, "CfInterconnects", { enumerable: true, get: function () { return cf_interconnects_1.CfInterconnects; } });
var index_1 = require("./connectors/index.js");
Object.defineProperty(exports, "ConnectorListResponsesSinglePage", { enumerable: true, get: function () { return index_1.ConnectorListResponsesSinglePage; } });
Object.defineProperty(exports, "Connectors", { enumerable: true, get: function () { return index_1.Connectors; } });
var gre_tunnels_1 = require("./gre-tunnels.js");
Object.defineProperty(exports, "GRETunnels", { enumerable: true, get: function () { return gre_tunnels_1.GRETunnels; } });
var ipsec_tunnels_1 = require("./ipsec-tunnels.js");
Object.defineProperty(exports, "IPSECTunnels", { enumerable: true, get: function () { return ipsec_tunnels_1.IPSECTunnels; } });
var magic_transit_1 = require("./magic-transit.js");
Object.defineProperty(exports, "MagicTransit", { enumerable: true, get: function () { return magic_transit_1.MagicTransit; } });
var index_2 = require("./pcaps/index.js");
Object.defineProperty(exports, "PCAPListResponsesSinglePage", { enumerable: true, get: function () { return index_2.PCAPListResponsesSinglePage; } });
Object.defineProperty(exports, "PCAPs", { enumerable: true, get: function () { return index_2.PCAPs; } });
var routes_1 = require("./routes.js");
Object.defineProperty(exports, "Routes", { enumerable: true, get: function () { return routes_1.Routes; } });
var index_3 = require("./sites/index.js");
Object.defineProperty(exports, "SitesSinglePage", { enumerable: true, get: function () { return index_3.SitesSinglePage; } });
Object.defineProperty(exports, "Sites", { enumerable: true, get: function () { return index_3.Sites; } });
