// @mistralai/mistralai/models/operations/agentsapiv1conversationsdelete.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentsApiV1ConversationsDeleteRequest$ = exports.AgentsApiV1ConversationsDeleteRequest$outboundSchema = exports.AgentsApiV1ConversationsDeleteRequest$inboundSchema = void 0;
exports.agentsApiV1ConversationsDeleteRequestToJSON = agentsApiV1ConversationsDeleteRequestToJSON;
exports.agentsApiV1ConversationsDeleteRequestFromJSON = agentsApiV1ConversationsDeleteRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.AgentsApiV1ConversationsDeleteRequest$inboundSchema = z.object({
    conversation_id: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "conversation_id": "conversationId",
    });
});
/** @internal */
exports.AgentsApiV1ConversationsDeleteRequest$outboundSchema = z.object({
    conversationId: z.string(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        conversationId: "conversation_id",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var AgentsApiV1ConversationsDeleteRequest$;
(function (AgentsApiV1ConversationsDeleteRequest$) {
    /** @deprecated use `AgentsApiV1ConversationsDeleteRequest$inboundSchema` instead. */
    AgentsApiV1ConversationsDeleteRequest$.inboundSchema = exports.AgentsApiV1ConversationsDeleteRequest$inboundSchema;
    /** @deprecated use `AgentsApiV1ConversationsDeleteRequest$outboundSchema` instead. */
    AgentsApiV1ConversationsDeleteRequest$.outboundSchema = exports.AgentsApiV1ConversationsDeleteRequest$outboundSchema;
})(AgentsApiV1ConversationsDeleteRequest$ || (exports.AgentsApiV1ConversationsDeleteRequest$ = AgentsApiV1ConversationsDeleteRequest$ = {}));
function agentsApiV1ConversationsDeleteRequestToJSON(agentsApiV1ConversationsDeleteRequest) {
    return JSON.stringify(exports.AgentsApiV1ConversationsDeleteRequest$outboundSchema.parse(agentsApiV1ConversationsDeleteRequest));
}
function agentsApiV1ConversationsDeleteRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.AgentsApiV1ConversationsDeleteRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'AgentsApiV1ConversationsDeleteRequest' from JSON`);
}
