// @cloudbase/agent-tools/dist/fs/multi-write-tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.multiWrite = multiWrite;
exports.createMultiWriteTool = createMultiWriteTool;
const zod_1 = require("zod");
const utils_1 = require("../utils");
const write_tool_1 = require("./write-tool");
/**
 * Write multiple files concurrently
 */
async function multiWrite(context, params) {
    const { writes = [] } = params;
    const startTime = Date.now();
    try {
        context.messageHandler?.currentText.append(`\n[multi-write] Writing ${writes.length} files concurrently...`);
        // Process all files concurrently
        const promises = writes.map(async (fileRequest) => {
            try {
                const res = await (0, write_tool_1.write)(context, fileRequest);
                return {
                    file_path: fileRequest.file_path,
                    ...res,
                };
            }
            catch (error) {
                return {
                    file_path: fileRequest.file_path,
                    success: false,
                    error: error instanceof Error ? error.message : String(error),
                };
            }
        });
        const results = await Promise.all(promises);
        const successfulFiles = results.filter((r) => r.success).length;
        const failedFiles = results.length - successfulFiles;
        const summary = `Successfully wrote ${successfulFiles} file(s), failed ${failedFiles} file(s) out of ${writes.length} total`;
        context.messageHandler?.currentText.append(`\n[multi-write] Completed: ${summary}\n`);
        const multipleFilesResult = {
            files: results,
            total: writes.length,
            successfulCount: successfulFiles,
            failedCount: failedFiles,
        };
        return new utils_1.ToolResult({
            success: true,
            data: multipleFilesResult,
        });
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        context.messageHandler?.currentText.append(`\n[multi-write] Failed: ${errorMessage}\n`);
        return (0, utils_1.handleToolError)(error, "Multi-write tool execution", "execution");
    }
}
/**
 * Create multi-write tool with execution context
 */
function createMultiWriteTool(context) {
    return (0, utils_1.tool)((params) => multiWrite(context, params), {
        name: "MultiWrite",
        description: `Writes multiple files concurrently.
**Usage:**
  - Provide an array of file write requests
  - Each request contains file_path and content.
  - All files are written concurrently
  - Returns detailed results for each file including success/failure status`,
        schema: zod_1.z.object({
            writes: zod_1.z
                .array(write_tool_1.writeParametersSchema)
                .describe("Array of file write requests to process concurrently"),
        }),
        getDisplay: ({ name, input }) => {
            return `> Using ${name} for ${input.writes.length} files`;
        },
    });
}
