// @cloudbase/agent-tools/dist/fs/str-replace-editor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStrReplaceEditorTool = createStrReplaceEditorTool;
const v4_1 = require("zod/v4");
const read_tool_1 = require("./read-tool");
const write_tool_1 = require("./write-tool");
const edit_tool_1 = require("./edit-tool");
const utils_1 = require("../utils");
function createStrReplaceEditorTool(context) {
    return (0, utils_1.tool)(async ({ command, path, view_range = [], old_str, new_str, file_text, insert_line, }) => {
        switch (command) {
            case "view":
                const [startLine = 0, endLine = -1] = view_range;
                return (0, read_tool_1.read)(context, { file_path: path, offset: startLine, endLine });
            case "str_replace":
                return (0, edit_tool_1.editFile)(context, {
                    file_path: path,
                    old_string: old_str,
                    new_string: new_str,
                });
            case "create":
                return (0, write_tool_1.write)(context, { file_path: path, content: file_text });
            case "insert": {
                const content = (await context.fsOperator.readFile(path, "utf-8")).toString();
                const lines = content.split("\n");
                if (insert_line != undefined && insert_line <= lines.length) {
                    lines.splice(insert_line + 1, 0, new_str || "");
                }
                return (0, write_tool_1.write)(context, { file_path: path, content: lines.join("\n") });
            }
            default:
                return {
                    success: false,
                    error: `Unknown command: ${command}`,
                };
        }
    }, {
        name: "str-replace-editor",
        description: "Text editor tool to view and modify text files, helping you debug, fix, and improve code or other text documents.Common indicates the operation type. Different commons require different parameters.",
        schema: v4_1.z.object({
            command: v4_1.z.enum(["view", "str_replace", "create", "insert", "grep"])
                .describe(`Edit type，Must be one of:
1. view: Examine the contents of a file or list the contents of a directory. It can read the entire file or a specific range of lines
2. str_replace: Replace a specific string in a file with a new string. This is used for making precise edits.
3. create: Create a new file with specified content.
4. insert: Insert text at a specific location in a file.
5. grep: Search for text patterns within file contents using regular expressions. Can filter by file types and paths.`),
            path: v4_1.z.string().describe("The path to the file or directory to view."),
            view_range: v4_1.z
                .array(v4_1.z.number())
                .min(2)
                .max(2)
                .optional()
                .describe("Optional: An array of two integers specifying the start and end line numbers to view. Line numbers are 1-indexed, and -1 for the end line means read to the end of the file. This parameter only applies when viewing files, not directories."),
            old_str: v4_1.z
                .string()
                .optional()
                .describe("The text to replace (must match exactly, including whitespace and indentation). Is required when command is str_replace."),
            new_str: v4_1.z
                .string()
                .optional()
                .describe("The new text to insert in place of the old text. Is required when command is str_replace or insert."),
            file_text: v4_1.z
                .string()
                .optional()
                .describe("The content to write to the new file. Is required when command is create."),
            insert_line: v4_1.z
                .number()
                .optional()
                .describe("The line number after which to insert the text (0 for beginning of file). Is required when command is insert."),
        }),
    });
}
