// @cloudbase/agent-tools/dist/fs/write-tool.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeParametersSchema = void 0;
exports.write = write;
exports.createWriteTool = createWriteTool;
const path = __importStar(require("path"));
const v4_1 = require("zod/v4");
const utils_1 = require("./utils");
const utils_2 = require("../utils");
exports.writeParametersSchema = v4_1.z.object({
    file_path: v4_1.z
        .string()
        .describe("The absolute path to the file to write (must be absolute, not relative)"),
    content: v4_1.z.string().describe("Content to write to the file"),
});
async function write(context, { file_path = "", content = "", create_dirs = true }) {
    const startTime = Date.now();
    try {
        // Validate workspace path (handles both absolute and relative paths)
        const pathError = (0, utils_1.validateWorkspacePath)(file_path, context);
        if (pathError) {
            return pathError;
        }
        // Resolve absolute path within workspace
        const absolutePath = (0, utils_1.resolveWorkspacePath)(file_path, context);
        context.messageHandler?.currentText.append(`\n[write] Writing to file: ${file_path}\n`);
        // Check if target is a directory
        if (await context.fsOperator.exists(absolutePath)) {
            const stats = await context.fsOperator.stat(absolutePath);
            if (stats.isDirectory()) {
                return (0, utils_2.handleToolError)(`Target path is a directory, not a file: ${file_path}`, "Path validation", "validation");
            }
        }
        // Create parent directories if needed and requested
        if (create_dirs) {
            const dirName = path.dirname(absolutePath);
            if (!(await context.fsOperator.exists(dirName))) {
                await context.fsOperator.mkdir(dirName, { recursive: true });
                context.messageHandler?.currentText.append(`\n[write] Created parent directories for: ${file_path}\n`);
            }
        }
        // Determine if this is a new file or overwrite
        const isNewFile = !(await context.fsOperator.exists(absolutePath));
        // Write the file
        await context.fsOperator.writeFile(absolutePath, content, "utf8");
        const duration = Date.now() - startTime;
        const lines = content.split("\n").length;
        const size = Buffer.byteLength(content, "utf8");
        context.messageHandler?.currentText.append(`\n[write] ${isNewFile ? "Created" : "Updated"} file: ${file_path} (${lines} lines, ${size} bytes) in ${duration}ms\n`);
        const result = {
            file_path,
            absolute_path: absolutePath,
            is_new_file: isNewFile,
            lines_written: lines,
            bytes_written: size,
        };
        return (0, utils_1.createSuccessResponse)(result);
    }
    catch (error) {
        const duration = Date.now() - startTime;
        const errorMessage = error instanceof Error ? error.message : String(error);
        context.messageHandler?.currentText.append(`\n[write] Error writing file: ${errorMessage} (${duration}ms)\n`);
        return (0, utils_2.handleToolError)(error, "Write tool execution", "execution");
    }
}
/**
 * Create SuperDesign write tool with execution context
 */
function createWriteTool(context) {
    return (0, utils_2.tool)((params) => write(context, params), {
        name: "Write",
        description: "Writes a file to the local filesystem.\n\nUsage:\n- This tool will overwrite the existing file if there is one at the provided path.\n- If this is an existing file, you MUST use the Read tool first to read the file's contents. This tool will fail if you did not read the file first.\n- ALWAYS prefer editing existing files in the codebase. NEVER write new files unless explicitly required.\n- NEVER proactively create documentation files (*.md) or README files. Only create documentation files if explicitly requested by the User.",
        schema: exports.writeParametersSchema,
        getDisplay: ({ name, input }) => {
            return `> Using ${name} for tool call: ${input.file_path || ""}`;
        },
    });
}
