// @cloudbase/agent-observability/src/core/tracerProvider.ts
import { TracerProvider, trace, context } from "@opentelemetry/api";

const OBSERVABILITY_GLOBAL_SYMBOL = Symbol.for("observability");

type ObservabilityGlobalState = {
  isolatedTracerProvider: TracerProvider | null;
};

function createState(): ObservabilityGlobalState {
  return {
    isolatedTracerProvider: null,
  };
}

interface GlobalThis {
  [OBSERVABILITY_GLOBAL_SYMBOL]?: ObservabilityGlobalState;
}

/**
 * Gets the global state for tracing observability.
 *
 * @returns The global state object
 * @internal
 */
function getObservabilityGlobalState(): ObservabilityGlobalState {
  const initialState = createState();

  try {
    const g = globalThis as typeof globalThis & GlobalThis;

    if (typeof g !== "object" || g === null) {
      console.warn(
        "[Observability] globalThis is not available, using fallback state",
      );
      return initialState;
    }

    if (!g[OBSERVABILITY_GLOBAL_SYMBOL]) {
      Object.defineProperty(g, OBSERVABILITY_GLOBAL_SYMBOL, {
        value: initialState,
        writable: false,
        configurable: false,
        enumerable: false,
      });
    }

    return g[OBSERVABILITY_GLOBAL_SYMBOL]!;
  } catch (err) {
    console.error(
      `[Observability] Failed to access global state: ${err instanceof Error ? err.message : String(err)}`,
    );
    return initialState;
  }
}

/**
 * Sets an isolated TracerProvider for tracing tracing operations.
 *
 * This allows tracing to use its own TracerProvider instance, separate from
 * the global OpenTelemetry TracerProvider.
 *
 * Note: While this isolates span processing and export, it does NOT provide
 * complete trace isolation. OpenTelemetry context (trace IDs, parent spans)
 * is still shared between the global and isolated providers.
 *
 * @param provider - The TracerProvider instance to use, or null to clear
 *
 * @example
 * ```typescript
 * import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
 * import { setTracerProvider } from './observability';
 *
 * const provider = new NodeTracerProvider();
 * setTracerProvider(provider);
 * ```
 *
 * @public
 */
export function setTracerProvider(provider: TracerProvider | null) {
  getObservabilityGlobalState().isolatedTracerProvider = provider;
}

/**
 * Gets the TracerProvider for tracing tracing operations.
 *
 * Returns the isolated TracerProvider if one has been set via setTracerProvider(),
 * otherwise falls back to the global OpenTelemetry TracerProvider.
 *
 * @returns The TracerProvider instance to use for tracing tracing
 *
 * @example
 * ```typescript
 * import { getTracerProvider } from './observability';
 *
 * const provider = getTracerProvider();
 * const tracer = provider.getTracer('my-tracer', '1.0.0');
 * ```
 *
 * @public
 */
export function getTracerProvider(): TracerProvider {
  const { isolatedTracerProvider } = getObservabilityGlobalState();

  if (isolatedTracerProvider) return isolatedTracerProvider;

  return trace.getTracerProvider();
}

/**
 * Gets the OpenTelemetry tracer instance for tracing.
 *
 * Returns a tracer specifically configured for tracing with the correct
 * tracer name and version.
 *
 * @returns The tracing OpenTelemetry tracer instance
 *
 * @example
 * ```typescript
 * import { getTracer } from './observability';
 *
 * const tracer = getTracer();
 * const span = tracer.startSpan('my-operation');
 * ```
 *
 * @public
 */
export function getTracer() {
  return getTracerProvider().getTracer(
    OBSERVABILITY_SDK_NAME,
    OBSERVABILITY_SDK_VERSION
  );
}

// SDK version - could be read from package.json in production
const OBSERVABILITY_SDK_NAME = "ag-kit-observability";
const OBSERVABILITY_SDK_VERSION = "0.1.0";
