// cloudflare/resources/zero-trust/dlp/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Profiles = exports.ProfilesSinglePage = exports.PayloadLogs = exports.Patterns = exports.Limits = exports.Entries = exports.EntryListResponsesSinglePage = exports.Email = exports.Datasets = exports.DatasetsSinglePage = exports.DLP = void 0;
var dlp_1 = require("./dlp.js");
Object.defineProperty(exports, "DLP", { enumerable: true, get: function () { return dlp_1.DLP; } });
var index_1 = require("./datasets/index.js");
Object.defineProperty(exports, "DatasetsSinglePage", { enumerable: true, get: function () { return index_1.DatasetsSinglePage; } });
Object.defineProperty(exports, "Datasets", { enumerable: true, get: function () { return index_1.Datasets; } });
var index_2 = require("./email/index.js");
Object.defineProperty(exports, "Email", { enumerable: true, get: function () { return index_2.Email; } });
var index_3 = require("./entries/index.js");
Object.defineProperty(exports, "EntryListResponsesSinglePage", { enumerable: true, get: function () { return index_3.EntryListResponsesSinglePage; } });
Object.defineProperty(exports, "Entries", { enumerable: true, get: function () { return index_3.Entries; } });
var limits_1 = require("./limits.js");
Object.defineProperty(exports, "Limits", { enumerable: true, get: function () { return limits_1.Limits; } });
var patterns_1 = require("./patterns.js");
Object.defineProperty(exports, "Patterns", { enumerable: true, get: function () { return patterns_1.Patterns; } });
var payload_logs_1 = require("./payload-logs.js");
Object.defineProperty(exports, "PayloadLogs", { enumerable: true, get: function () { return payload_logs_1.PayloadLogs; } });
var index_4 = require("./profiles/index.js");
Object.defineProperty(exports, "ProfilesSinglePage", { enumerable: true, get: function () { return index_4.ProfilesSinglePage; } });
Object.defineProperty(exports, "Profiles", { enumerable: true, get: function () { return index_4.Profiles; } });
