// cloudflare/resources/alerting/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Policies = exports.PoliciesSinglePage = exports.HistoryResource = exports.HistoriesV4PagePaginationArray = exports.Destinations = exports.AvailableAlerts = exports.Alerting = void 0;
var alerting_1 = require("./alerting.js");
Object.defineProperty(exports, "Alerting", { enumerable: true, get: function () { return alerting_1.Alerting; } });
var available_alerts_1 = require("./available-alerts.js");
Object.defineProperty(exports, "AvailableAlerts", { enumerable: true, get: function () { return available_alerts_1.AvailableAlerts; } });
var index_1 = require("./destinations/index.js");
Object.defineProperty(exports, "Destinations", { enumerable: true, get: function () { return index_1.Destinations; } });
var history_1 = require("./history.js");
Object.defineProperty(exports, "HistoriesV4PagePaginationArray", { enumerable: true, get: function () { return history_1.HistoriesV4PagePaginationArray; } });
Object.defineProperty(exports, "HistoryResource", { enumerable: true, get: function () { return history_1.HistoryResource; } });
var policies_1 = require("./policies.js");
Object.defineProperty(exports, "PoliciesSinglePage", { enumerable: true, get: function () { return policies_1.PoliciesSinglePage; } });
Object.defineProperty(exports, "Policies", { enumerable: true, get: function () { return policies_1.Policies; } });
