// cloudflare/resources/radar/http/timeseries-groups.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../resource.mjs";
import { isRequestOptions } from "../../../core.mjs";
export class TimeseriesGroups extends APIResource {
    botClass(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.botClass({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/bot_class', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    browser(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.browser({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/browser', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    browserFamily(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.browserFamily({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/browser_family', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    deviceType(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.deviceType({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/device_type', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    httpProtocol(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.httpProtocol({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/http_protocol', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    httpVersion(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.httpVersion({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/http_version', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    ipVersion(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.ipVersion({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/ip_version', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    os(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.os({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/os', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    postQuantum(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.postQuantum({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/post_quantum', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    tlsVersion(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.tlsVersion({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/tls_version', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
}
