// @cloudbase/ai/dist/cjs/AI.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AI = void 0;
var bot_1 = require("./bot");
var models = __importStar(require("./models"));
var utils_1 = require("./utils");
var utilities_1 = require("@cloudbase/utilities");
var MODELS = models.MODELS;
var AI = (function () {
    function AI(req, baseUrl, i18n) {
        var _this = this;
        this.req = req;
        this.baseUrl = baseUrl;
        this.modelRequest = function (_a) {
            var url = _a.url, data = _a.data, headers = _a.headers, stream = _a.stream, timeout = _a.timeout;
            return __awaiter(_this, void 0, void 0, function () {
                var fetchHeaders, _b, responseData, header;
                var _c;
                var _d, _e;
                return __generator(this, function (_f) {
                    switch (_f.label) {
                        case 0:
                            fetchHeaders = {
                                'Content-Type': 'application/json',
                            };
                            stream && Object.assign(fetchHeaders, { Accept: 'text/event-stream' });
                            return [4, this.req.fetch({
                                    method: 'post',
                                    headers: __assign(__assign((_c = {}, _c[(_d = this.i18n) === null || _d === void 0 ? void 0 : _d.LANG_HEADER_KEY] = (_e = this.i18n) === null || _e === void 0 ? void 0 : _e.lang, _c), fetchHeaders), headers),
                                    body: JSON.stringify(data),
                                    url: url,
                                    stream: stream,
                                    timeout: timeout,
                                })];
                        case 1:
                            _b = (_f.sent()), responseData = _b.data, header = _b.header;
                            return [2, this.handleResponseData(responseData, header)];
                    }
                });
            });
        };
        this.botRequest = function (_a) {
            var method = _a.method, url = _a.url, _b = _a.data, data = _b === void 0 ? {} : _b, headers = _a.headers, stream = _a.stream, timeout = _a.timeout;
            return __awaiter(_this, void 0, void 0, function () {
                function objectToParam(obj) {
                    return Object.entries(obj)
                        .map(function (_a) {
                        var key = _a[0], value = _a[1];
                        return "".concat(key, "=").concat(value);
                    })
                        .join('&');
                }
                var _c, fetchHeaders, _d, responseData, header;
                var _e;
                var _f, _g;
                return __generator(this, function (_h) {
                    switch (_h.label) {
                        case 0:
                            if (!(method === 'get')) return [3, 2];
                            _c = this.handleResponseData;
                            return [4, this.req.fetch({ url: "".concat(url, "?").concat(objectToParam(data)), method: method, headers: headers, stream: stream, timeout: timeout })];
                        case 1: return [2, _c.apply(this, [(_h.sent()).data])];
                        case 2:
                            fetchHeaders = {
                                'Content-Type': 'application/json',
                            };
                            stream && Object.assign(fetchHeaders, { Accept: 'text/event-stream' });
                            return [4, this.req.fetch({
                                    url: url,
                                    body: JSON.stringify(data),
                                    headers: __assign(__assign((_e = {}, _e[(_f = this.i18n) === null || _f === void 0 ? void 0 : _f.LANG_HEADER_KEY] = (_g = this.i18n) === null || _g === void 0 ? void 0 : _g.lang, _e), fetchHeaders), headers),
                                    stream: stream,
                                    method: method,
                                    timeout: timeout,
                                })];
                        case 3:
                            _d = (_h.sent()), responseData = _d.data, header = _d.header;
                            return [2, this.handleResponseData(responseData, header)];
                    }
                });
            });
        };
        this.aiBaseUrl = "".concat(baseUrl, "/ai");
        this.aiBotBaseUrl = "".concat(baseUrl, "/aibot");
        this.bot = new bot_1.Bot(this.botRequest, this.aiBotBaseUrl);
        this.i18n = i18n;
        utilities_1.langEvent.bus.on(utilities_1.langEvent.LANG_CHANGE_EVENT, function (params) {
            var _a;
            _this.i18n = ((_a = params.data) === null || _a === void 0 ? void 0 : _a.i18n) || _this.i18n;
        });
    }
    AI.prototype.handleResponseData = function (responseData, header) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var GO_TO_AI_TEXT, json, json;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        GO_TO_AI_TEXT = "".concat(this.i18n.t('请检查调用方式，或前往云开发 AI+ 首页查看文档'), "\uFF1Ahttps://tcb.cloud.tencent.com/dev#/ai");
                        if (!(typeof responseData === 'object' && responseData && 'then' in responseData)) return [3, 2];
                        return [4, responseData];
                    case 1:
                        json = (_c.sent());
                        if (typeof json === 'object' && json && 'code' in json && json.code !== 'NORMAL') {
                            throw new Error("".concat(this.i18n.t('AI+ 请求出错，错误码'), "\uFF1A").concat(json.code, "\uFF0C").concat(this.i18n.t('错误信息'), "\uFF1A").concat(json.message, "\n").concat(GO_TO_AI_TEXT, "\n").concat(JSON.stringify(json, null, 2)));
                        }
                        return [2, responseData];
                    case 2:
                        if (!((_b = (_a = header === null || header === void 0 ? void 0 : header.get) === null || _a === void 0 ? void 0 : _a.call(header, 'content-type')) === null || _b === void 0 ? void 0 : _b.includes('application/json'))) return [3, 4];
                        return [4, (0, utils_1.readableStream2JsonObject)(responseData)];
                    case 3:
                        json = _c.sent();
                        if (typeof json === 'object' && json && 'code' in json && json.code !== 'NORMAL') {
                            throw new Error("".concat(this.i18n.t('AI+ 请求出错，错误码'), "\uFF1A").concat(json.code, "\uFF0C").concat(this.i18n.t('错误信息'), "\uFF1A").concat(json.message, "\n").concat(GO_TO_AI_TEXT, "\n").concat(JSON.stringify(json, null, 2)));
                        }
                        _c.label = 4;
                    case 4: return [2, responseData];
                }
            });
        });
    };
    AI.prototype.createModel = function (model, options) {
        var simpleModel;
        var SimpleModelConstructor = MODELS[model];
        if (SimpleModelConstructor) {
            simpleModel = new SimpleModelConstructor(this.modelRequest, this.aiBaseUrl);
        }
        else {
            var subUrl = typeof (options === null || options === void 0 ? void 0 : options.defaultModelSubUrl) === 'string' ? options.defaultModelSubUrl : '/chat/completions';
            simpleModel = new models.DefaultSimpleModel(this.modelRequest, this.aiBaseUrl, "".concat(model).concat(subUrl));
        }
        var reactModel = new models.ReactModel(simpleModel);
        return reactModel;
    };
    AI.prototype.registerModel = function (name, model) {
        if (MODELS[name] != null) {
            console.warn("AI model ".concat(name, " already exists!"));
            return;
        }
        MODELS[name] = model;
    };
    AI.prototype.createImageModel = function (provider) {
        return new models.DefaultImageModel(this.modelRequest, this.aiBaseUrl, provider);
    };
    AI.prototype.registerFunctionTool = function (functionTool) {
        if (models.toolMap.has(functionTool.name)) {
            console.warn("AI function tool ".concat(functionTool.name, " already exists and will be overwritten!"));
        }
        models.toolMap.set(functionTool.name, functionTool.fn);
    };
    return AI;
}());
exports.AI = AI;
