// cloudflare/resources/radar/http/timeseries-groups.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeseriesGroups = void 0;
const resource_1 = require("../../../resource.js");
const core_1 = require("../../../core.js");
class TimeseriesGroups extends resource_1.APIResource {
    botClass(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.botClass({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/bot_class', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    browser(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.browser({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/browser', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    browserFamily(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.browserFamily({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/browser_family', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    deviceType(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.deviceType({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/device_type', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    httpProtocol(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.httpProtocol({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/http_protocol', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    httpVersion(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.httpVersion({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/http_version', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    ipVersion(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.ipVersion({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/ip_version', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    os(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.os({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/os', { query, ...options })._thenUnwrap((obj) => obj.result);
    }
    postQuantum(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.postQuantum({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/post_quantum', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
    tlsVersion(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.tlsVersion({}, query);
        }
        return this._client.get('/radar/http/timeseries_groups/tls_version', {
            query,
            ...options,
        })._thenUnwrap((obj) => obj.result);
    }
}
exports.TimeseriesGroups = TimeseriesGroups;
