// @cloudbase/agent-tools/dist/fs/operator/transaction/transaction-operator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionFileOperator = exports.EOperatorType = void 0;
const base_operator_1 = require("../base-operator");
const local_operator_1 = require("../local-operator");
const in_memory_operator_1 = require("../in-memory-operator");
const sanbox_operator_1 = require("../sanbox-operator");
const snapshot_strategies_1 = require("./snapshot-strategies");
var EOperatorType;
(function (EOperatorType) {
    EOperatorType["LOCAL"] = "local";
    EOperatorType["MEMORY"] = "memory";
    EOperatorType["SANDBOX"] = "sandbox";
})(EOperatorType || (exports.EOperatorType = EOperatorType = {}));
class TransactionFileOperator extends base_operator_1.BaseFileOperator {
    constructor(baseOperator, operatorType) {
        super();
        // 事务状态
        this.isInTransaction = false;
        this.transactionOperations = [];
        this.fileStates = new Map();
        // 配置
        this.config = {
            batchSize: 100,
            maxConcurrency: 10,
            enableLazyLoading: true,
        };
        this.baseOperator = baseOperator;
        this.operatorType = operatorType || this.detectOperatorType(baseOperator);
        this.snapshotStrategy = this.createSnapshotStrategy();
        this.workspaceRoot = baseOperator.workspaceRoot;
    }
    detectOperatorType(operator) {
        if (operator instanceof local_operator_1.LocalFileOperator)
            return EOperatorType.LOCAL;
        if (operator instanceof in_memory_operator_1.InMemoryFileOperator)
            return EOperatorType.MEMORY;
        if (operator instanceof sanbox_operator_1.SandboxFileOperator)
            return EOperatorType.SANDBOX;
        return EOperatorType.LOCAL;
    }
    createSnapshotStrategy() {
        switch (this.operatorType) {
            case EOperatorType.LOCAL:
                return new snapshot_strategies_1.LocalSnapshotStrategy();
            case EOperatorType.MEMORY:
                return new snapshot_strategies_1.MemorySnapshotStrategy(this.baseOperator);
            case EOperatorType.SANDBOX:
                return new snapshot_strategies_1.SandboxSnapshotStrategy(this.baseOperator);
            default:
                return new snapshot_strategies_1.LocalSnapshotStrategy();
        }
    }
    // ==================== 事务管理 ====================
    async beginTransaction() {
        if (this.isInTransaction) {
            throw new Error("事务已经开始");
        }
        this.isInTransaction = true;
        await this.snapshotStrategy.initialize();
        this.transactionOperations = [];
        this.fileStates.clear();
    }
    async commitTransaction() {
        if (!this.isInTransaction) {
            throw new Error("没有活跃的事务");
        }
        await this.snapshotStrategy.cleanup();
        this.resetTransactionState();
    }
    async rollbackTransaction() {
        if (!this.isInTransaction) {
            throw new Error("没有活跃的事务");
        }
        await this.snapshotStrategy.rollback(this.transactionOperations);
        await this.snapshotStrategy.cleanup();
        this.resetTransactionState();
    }
    resetTransactionState() {
        this.isInTransaction = false;
        this.transactionOperations = [];
        this.fileStates.clear();
    }
    // ==================== 文件状态管理 ====================
    async saveFileState(path) {
        if (!this.isInTransaction || this.fileStates.has(path))
            return;
        try {
            const exists = await this.baseOperator.exists(path);
            const fileState = { exists, isDirectory: false };
            if (exists) {
                const stat = await this.baseOperator.stat(path);
                fileState.isDirectory = stat.isDirectory();
                // 延迟加载内容
                if (this.config.enableLazyLoading && !fileState.isDirectory) {
                    fileState.contentLoader = async () => {
                        const content = await this.baseOperator.readFile(path);
                        return Buffer.isBuffer(content) ? content : Buffer.from(content);
                    };
                }
            }
            this.fileStates.set(path, fileState);
            await this.snapshotStrategy.saveFile(path, fileState);
        }
        catch (error) {
            this.fileStates.set(path, { exists: false, isDirectory: false });
        }
    }
    recordOperation(type, path, oldPath) {
        this.transactionOperations.push({
            type,
            path,
            oldPath,
            timestamp: Date.now(),
        });
    }
    // ==================== BaseFileOperator 实现 ====================
    resolvePath(filePath) {
        return this.baseOperator.resolvePath(filePath);
    }
    async access(path) {
        return this.baseOperator.access(path);
    }
    async readFile(path, encoding) {
        // 直接读取，因为所有操作都已应用到磁盘
        return this.baseOperator.readFile(path, encoding);
    }
    async writeFile(path, data, options) {
        if (this.isInTransaction) {
            const pathStr = this.resolvePath(path);
            await this.saveFileState(pathStr);
            this.recordOperation("write", pathStr);
        }
        return this.baseOperator.writeFile(path, data, options);
    }
    async mkdir(path, options) {
        if (this.isInTransaction) {
            const pathStr = this.resolvePath(path);
            await this.saveFileState(pathStr);
            this.recordOperation("mkdir", pathStr);
        }
        return this.baseOperator.mkdir(path, options);
    }
    async readdir(path, options) {
        return this.baseOperator.readdir(path, options);
    }
    async stat(path) {
        return this.baseOperator.stat(path);
    }
    async exists(path) {
        // 直接检查，因为所有操作都已应用到磁盘
        return this.baseOperator.exists(path);
    }
    async unlink(path) {
        if (this.isInTransaction) {
            const pathStr = this.resolvePath(path);
            await this.saveFileState(pathStr);
            this.recordOperation("unlink", pathStr);
        }
        return this.baseOperator.unlink(path);
    }
    async rmdir(path) {
        if (this.isInTransaction) {
            const pathStr = this.resolvePath(path);
            await this.saveFileState(pathStr);
            this.recordOperation("rmdir", pathStr);
        }
        return this.baseOperator.rmdir(path);
    }
    async rename(oldPath, newPath) {
        if (this.isInTransaction) {
            const oldPathStr = this.resolvePath(oldPath);
            const newPathStr = this.resolvePath(newPath);
            await this.saveFileState(oldPathStr);
            await this.saveFileState(newPathStr);
            this.recordOperation("rename", newPathStr, oldPathStr);
        }
        return this.baseOperator.rename(oldPath, newPath);
    }
    // ==================== 公共接口 ====================
    isTransactionActive() {
        return this.isInTransaction;
    }
    getTransactionOperations() {
        return [...this.transactionOperations];
    }
    setConfig(config) {
        this.config = { ...this.config, ...config };
    }
    getConfig() {
        return { ...this.config };
    }
}
exports.TransactionFileOperator = TransactionFileOperator;
