// cloudflare/resources/radar/bgp/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Top = exports.Routes = exports.Leaks = exports.IPs = exports.Hijacks = exports.BGP = void 0;
var bgp_1 = require("./bgp.js");
Object.defineProperty(exports, "BGP", { enumerable: true, get: function () { return bgp_1.BGP; } });
var index_1 = require("./hijacks/index.js");
Object.defineProperty(exports, "Hijacks", { enumerable: true, get: function () { return index_1.Hijacks; } });
var ips_1 = require("./ips.js");
Object.defineProperty(exports, "IPs", { enumerable: true, get: function () { return ips_1.IPs; } });
var index_2 = require("./leaks/index.js");
Object.defineProperty(exports, "Leaks", { enumerable: true, get: function () { return index_2.Leaks; } });
var routes_1 = require("./routes.js");
Object.defineProperty(exports, "Routes", { enumerable: true, get: function () { return routes_1.Routes; } });
var index_3 = require("./top/index.js");
Object.defineProperty(exports, "Top", { enumerable: true, get: function () { return index_3.Top; } });
