// cloudflare/resources/workers-for-platforms/dispatch/namespaces/scripts/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tags = exports.TagListResponsesSinglePage = exports.TagUpdateResponsesSinglePage = exports.Settings = exports.Secrets = exports.SecretListResponsesSinglePage = exports.Scripts = exports.Content = exports.Bindings = exports.BindingGetResponsesSinglePage = exports.AssetUpload = void 0;
var asset_upload_1 = require("./asset-upload.js");
Object.defineProperty(exports, "AssetUpload", { enumerable: true, get: function () { return asset_upload_1.AssetUpload; } });
var bindings_1 = require("./bindings.js");
Object.defineProperty(exports, "BindingGetResponsesSinglePage", { enumerable: true, get: function () { return bindings_1.BindingGetResponsesSinglePage; } });
Object.defineProperty(exports, "Bindings", { enumerable: true, get: function () { return bindings_1.Bindings; } });
var content_1 = require("./content.js");
Object.defineProperty(exports, "Content", { enumerable: true, get: function () { return content_1.Content; } });
var scripts_1 = require("./scripts.js");
Object.defineProperty(exports, "Scripts", { enumerable: true, get: function () { return scripts_1.Scripts; } });
var secrets_1 = require("./secrets.js");
Object.defineProperty(exports, "SecretListResponsesSinglePage", { enumerable: true, get: function () { return secrets_1.SecretListResponsesSinglePage; } });
Object.defineProperty(exports, "Secrets", { enumerable: true, get: function () { return secrets_1.Secrets; } });
var settings_1 = require("./settings.js");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return settings_1.Settings; } });
var tags_1 = require("./tags.js");
Object.defineProperty(exports, "TagUpdateResponsesSinglePage", { enumerable: true, get: function () { return tags_1.TagUpdateResponsesSinglePage; } });
Object.defineProperty(exports, "TagListResponsesSinglePage", { enumerable: true, get: function () { return tags_1.TagListResponsesSinglePage; } });
Object.defineProperty(exports, "Tags", { enumerable: true, get: function () { return tags_1.Tags; } });
