// cloudflare/resources/zero-trust/tunnels/cloudflared/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Token = exports.Management = exports.Connectors = exports.Configurations = exports.Cloudflared = exports.CloudflaredListResponsesV4PagePaginationArray = exports.Connections = exports.ClientsSinglePage = void 0;
var connections_1 = require("./connections.js");
Object.defineProperty(exports, "ClientsSinglePage", { enumerable: true, get: function () { return connections_1.ClientsSinglePage; } });
Object.defineProperty(exports, "Connections", { enumerable: true, get: function () { return connections_1.Connections; } });
var cloudflared_1 = require("./cloudflared.js");
Object.defineProperty(exports, "CloudflaredListResponsesV4PagePaginationArray", { enumerable: true, get: function () { return cloudflared_1.CloudflaredListResponsesV4PagePaginationArray; } });
Object.defineProperty(exports, "Cloudflared", { enumerable: true, get: function () { return cloudflared_1.Cloudflared; } });
var configurations_1 = require("./configurations.js");
Object.defineProperty(exports, "Configurations", { enumerable: true, get: function () { return configurations_1.Configurations; } });
var connectors_1 = require("./connectors.js");
Object.defineProperty(exports, "Connectors", { enumerable: true, get: function () { return connectors_1.Connectors; } });
var management_1 = require("./management.js");
Object.defineProperty(exports, "Management", { enumerable: true, get: function () { return management_1.Management; } });
var token_1 = require("./token.js");
Object.defineProperty(exports, "Token", { enumerable: true, get: function () { return token_1.Token; } });
