// @jsonjoy.com/fs-node/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fsCommonObjectsList = exports.fsSynchronousApiList = exports.fsCallbackApiList = exports.FsPromises = exports.Dir = exports.FileHandle = exports.StatFs = exports.Dirent = exports.Stats = exports.toUnixTimestamp = exports.FSWatcher = exports.StatWatcher = exports.Volume = void 0;
const tslib_1 = require("tslib");
var volume_1 = require("./volume");
Object.defineProperty(exports, "Volume", { enumerable: true, get: function () { return volume_1.Volume; } });
Object.defineProperty(exports, "StatWatcher", { enumerable: true, get: function () { return volume_1.StatWatcher; } });
Object.defineProperty(exports, "FSWatcher", { enumerable: true, get: function () { return volume_1.FSWatcher; } });
Object.defineProperty(exports, "toUnixTimestamp", { enumerable: true, get: function () { return volume_1.toUnixTimestamp; } });
var Stats_1 = require("./Stats");
Object.defineProperty(exports, "Stats", { enumerable: true, get: function () { return Stats_1.default; } });
var Dirent_1 = require("./Dirent");
Object.defineProperty(exports, "Dirent", { enumerable: true, get: function () { return Dirent_1.default; } });
var StatFs_1 = require("./StatFs");
Object.defineProperty(exports, "StatFs", { enumerable: true, get: function () { return StatFs_1.default; } });
var FileHandle_1 = require("./FileHandle");
Object.defineProperty(exports, "FileHandle", { enumerable: true, get: function () { return FileHandle_1.FileHandle; } });
var Dir_1 = require("./Dir");
Object.defineProperty(exports, "Dir", { enumerable: true, get: function () { return Dir_1.Dir; } });
var FsPromises_1 = require("./FsPromises");
Object.defineProperty(exports, "FsPromises", { enumerable: true, get: function () { return FsPromises_1.FsPromises; } });
tslib_1.__exportStar(require("./options"), exports);
tslib_1.__exportStar(require("./util"), exports);
tslib_1.__exportStar(require("./glob"), exports);
var fsCallbackApiList_1 = require("./lists/fsCallbackApiList");
Object.defineProperty(exports, "fsCallbackApiList", { enumerable: true, get: function () { return fsCallbackApiList_1.fsCallbackApiList; } });
var fsSynchronousApiList_1 = require("./lists/fsSynchronousApiList");
Object.defineProperty(exports, "fsSynchronousApiList", { enumerable: true, get: function () { return fsSynchronousApiList_1.fsSynchronousApiList; } });
var fsCommonObjectsList_1 = require("./lists/fsCommonObjectsList");
Object.defineProperty(exports, "fsCommonObjectsList", { enumerable: true, get: function () { return fsCommonObjectsList_1.fsCommonObjectsList; } });
