// cloudflare/resources/dns/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneTransfers = exports.Settings = exports.Records = exports.RecordResponsesV4PagePaginationArray = exports.DNSSECResource = exports.DNS = exports.Analytics = void 0;
var index_1 = require("./analytics/index.js");
Object.defineProperty(exports, "Analytics", { enumerable: true, get: function () { return index_1.Analytics; } });
var dns_1 = require("./dns.js");
Object.defineProperty(exports, "DNS", { enumerable: true, get: function () { return dns_1.DNS; } });
var dnssec_1 = require("./dnssec.js");
Object.defineProperty(exports, "DNSSECResource", { enumerable: true, get: function () { return dnssec_1.DNSSECResource; } });
var records_1 = require("./records.js");
Object.defineProperty(exports, "RecordResponsesV4PagePaginationArray", { enumerable: true, get: function () { return records_1.RecordResponsesV4PagePaginationArray; } });
Object.defineProperty(exports, "Records", { enumerable: true, get: function () { return records_1.Records; } });
var index_2 = require("./settings/index.js");
Object.defineProperty(exports, "Settings", { enumerable: true, get: function () { return index_2.Settings; } });
var index_3 = require("./zone-transfers/index.js");
Object.defineProperty(exports, "ZoneTransfers", { enumerable: true, get: function () { return index_3.ZoneTransfers; } });
