// @cloudbase/ai/dist/cjs/models/Yi/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YiSimpleModel = void 0;
var utils_1 = require("../../utils");
function processInput(input) {
    var messages = input.messages, model = input.model, temperature = input.temperature, tools = input.tools, top_p = input.top_p;
    return __assign(__assign({}, input), { messages: messages, model: model, tools: tools, top_p: top_p, temperature: temperature });
}
var YiSimpleModel = (function () {
    function YiSimpleModel(req, baseUrl, subUrl) {
        this.req = req;
        this.baseUrl = baseUrl;
        this.subUrl = '01-ai/v1/chat/completions';
        if (subUrl != null) {
            this.subUrl = subUrl;
        }
    }
    Object.defineProperty(YiSimpleModel.prototype, "url", {
        get: function () {
            return "".concat(this.baseUrl, "/").concat(this.subUrl);
        },
        enumerable: false,
        configurable: true
    });
    YiSimpleModel.prototype.doGenerate = function (data, options) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.req({
                            url: this.url,
                            data: __assign(__assign({}, processInput(data)), { stream: false }),
                            stream: false,
                            timeout: options === null || options === void 0 ? void 0 : options.timeout,
                        })];
                    case 1:
                        res = (_a.sent());
                        return [2, __assign(__assign({}, res), { rawResponse: res })];
                }
            });
        });
    };
    YiSimpleModel.prototype.doStream = function (data, options) {
        return __awaiter(this, void 0, void 0, function () {
            var isToolCall, _stream, stream, arkStream, streamWithRaw;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        isToolCall = null;
                        return [4, this.req({
                                url: this.url,
                                data: __assign(__assign({}, processInput(data)), { stream: true }),
                                stream: true,
                                timeout: options === null || options === void 0 ? void 0 : options.timeout,
                            })];
                    case 1:
                        _stream = _a.sent();
                        stream = (0, utils_1.toPolyfillReadable)(_stream);
                        arkStream = (0, utils_1.intoStandardStream)(stream);
                        streamWithRaw = arkStream.pipeThrough(new utils_1.TransformStream({
                            transform: function (chunk, controller) {
                                var _a, _b, _c, _d, _e, _f;
                                if (!(((_c = (_b = (_a = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.delta) === null || _c === void 0 ? void 0 : _c.content) || ((_f = (_e = (_d = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _d === void 0 ? void 0 : _d[0]) === null || _e === void 0 ? void 0 : _e.delta) === null || _f === void 0 ? void 0 : _f.tool_calls)))
                                    return;
                                var newChoices = chunk.choices.map(function (choice) {
                                    var message = Object.assign(choice.delta, { role: 'assistant' });
                                    if (isToolCall == null)
                                        isToolCall = (0, utils_1.isToolCallAssistantMessage)(message);
                                    if (isToolCall) {
                                        return __assign(__assign({}, choice), { finish_reason: 'tool_calls', delta: message });
                                    }
                                    return __assign(__assign({}, choice), { delta: message });
                                });
                                var newChunk = __assign(__assign({}, chunk), { choices: newChoices });
                                controller.enqueue(__assign(__assign({}, newChunk), { rawResponse: chunk }));
                            },
                        }));
                        return [2, (0, utils_1.createAsyncIterable)(streamWithRaw)];
                }
            });
        });
    };
    return YiSimpleModel;
}());
exports.YiSimpleModel = YiSimpleModel;
