// @cloudbase/agent-tools/dist/bash/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandBuilders = void 0;
exports.createLocalBashContext = createLocalBashContext;
exports.createSandboxBashContext = createSandboxBashContext;
exports.validateCommand = validateCommand;
exports.escapeShellArg = escapeShellArg;
exports.buildCommand = buildCommand;
exports.parseCommandOutput = parseCommandOutput;
exports.formatExecutionTime = formatExecutionTime;
const local_operator_1 = require("./operator/local-operator");
const sandbox_operator_1 = require("./operator/sandbox-operator");
/**
 * Create a local bash execution context
 */
function createLocalBashContext(options) {
    const bashOperator = new local_operator_1.LocalBashOperator({
        cwd: options?.cwd,
        env: options?.env,
    });
    return {
        bashOperator,
        workingDirectory: options?.cwd || process.cwd(),
        environmentVariables: options?.env || {},
        defaultTimeout: options?.defaultTimeout || 30000,
    };
}
/**
 * Create a sandbox bash execution context
 */
function createSandboxBashContext(options) {
    const bashOperator = new sandbox_operator_1.SandboxBashOperator({
        sandbox: options.sandbox,
        cwd: options.cwd,
        env: options.env,
    });
    return {
        bashOperator,
        workingDirectory: options.cwd || "/home/user",
        environmentVariables: options.env || {},
        defaultTimeout: options.defaultTimeout || 30000,
    };
}
/**
 * Validate command for security (basic checks)
 */
function validateCommand(command) {
    // Basic security checks
    const dangerousPatterns = [
        /rm\s+-rf\s+\//, // rm -rf /
        /:\(\)\{.*\}/, // Fork bomb pattern
        /sudo\s+rm/, // sudo rm
        /mkfs/, // Format filesystem
        /dd\s+if=.*of=\/dev/, // Direct disk write
        />\s*\/dev\/sd[a-z]/, // Write to disk device
    ];
    for (const pattern of dangerousPatterns) {
        if (pattern.test(command)) {
            return {
                isValid: false,
                reason: `Command contains potentially dangerous pattern: ${pattern.source}`,
            };
        }
    }
    return { isValid: true };
}
/**
 * Escape shell arguments
 */
function escapeShellArg(arg) {
    return `'${arg.replace(/'/g, "'\"'\"'")}'`;
}
/**
 * Build command with escaped arguments
 */
function buildCommand(command, args = []) {
    if (args.length === 0) {
        return command;
    }
    const escapedArgs = args.map(escapeShellArg);
    return `${command} ${escapedArgs.join(" ")}`;
}
/**
 * Parse command output for common patterns
 */
function parseCommandOutput(stdout, stderr) {
    return {
        lines: stdout.split("\n").filter((line) => line.trim()),
        errorLines: stderr.split("\n").filter((line) => line.trim()),
        isEmpty: !stdout.trim() && !stderr.trim(),
        hasErrors: stderr.trim().length > 0,
        exitedCleanly: stderr.trim().length === 0,
    };
}
/**
 * Format execution time
 */
function formatExecutionTime(ms) {
    if (ms < 1000) {
        return `${ms}ms`;
    }
    else if (ms < 60000) {
        return `${(ms / 1000).toFixed(2)}s`;
    }
    else {
        const minutes = Math.floor(ms / 60000);
        const seconds = ((ms % 60000) / 1000).toFixed(2);
        return `${minutes}m ${seconds}s`;
    }
}
/**
 * Common command builders
 */
exports.CommandBuilders = {
    /**
     * List files with details
     */
    listFiles: (path = ".", options) => {
        let cmd = "ls";
        if (options?.all)
            cmd += " -a";
        if (options?.long)
            cmd += " -l";
        if (options?.human)
            cmd += " -h";
        return `${cmd} ${escapeShellArg(path)}`;
    },
    /**
     * Find files by pattern
     */
    findFiles: (pattern, path = ".") => {
        return `find ${escapeShellArg(path)} -name ${escapeShellArg(pattern)}`;
    },
    /**
     * Search in files
     */
    grep: (pattern, files = "*", options) => {
        let cmd = "grep";
        if (options?.recursive)
            cmd += " -r";
        if (options?.ignoreCase)
            cmd += " -i";
        if (options?.lineNumbers)
            cmd += " -n";
        return `${cmd} ${escapeShellArg(pattern)} ${escapeShellArg(files)}`;
    },
    /**
     * Check if file/directory exists
     */
    exists: (path) => {
        return `test -e ${escapeShellArg(path)} && echo "exists" || echo "not found"`;
    },
    /**
     * Get file info
     */
    stat: (path) => {
        return `stat ${escapeShellArg(path)}`;
    },
    /**
     * Create directory
     */
    mkdir: (path, options) => {
        let cmd = "mkdir";
        if (options?.parents)
            cmd += " -p";
        return `${cmd} ${escapeShellArg(path)}`;
    },
    /**
     * Copy files
     */
    copy: (source, destination, options) => {
        let cmd = "cp";
        if (options?.recursive)
            cmd += " -r";
        if (options?.preserve)
            cmd += " -p";
        return `${cmd} ${escapeShellArg(source)} ${escapeShellArg(destination)}`;
    },
    /**
     * Move/rename files
     */
    move: (source, destination) => {
        return `mv ${escapeShellArg(source)} ${escapeShellArg(destination)}`;
    },
    /**
     * Remove files/directories
     */
    remove: (path, options) => {
        let cmd = "rm";
        if (options?.recursive)
            cmd += " -r";
        if (options?.force)
            cmd += " -f";
        return `${cmd} ${escapeShellArg(path)}`;
    },
};
