// cloudflare/resources/ssl/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationResource = exports.Universal = exports.SSL = exports.Recommendations = exports.CertificatePacks = exports.CertificatePackListResponsesSinglePage = exports.Analyze = void 0;
var analyze_1 = require("./analyze.js");
Object.defineProperty(exports, "Analyze", { enumerable: true, get: function () { return analyze_1.Analyze; } });
var index_1 = require("./certificate-packs/index.js");
Object.defineProperty(exports, "CertificatePackListResponsesSinglePage", { enumerable: true, get: function () { return index_1.CertificatePackListResponsesSinglePage; } });
Object.defineProperty(exports, "CertificatePacks", { enumerable: true, get: function () { return index_1.CertificatePacks; } });
var recommendations_1 = require("./recommendations.js");
Object.defineProperty(exports, "Recommendations", { enumerable: true, get: function () { return recommendations_1.Recommendations; } });
var ssl_1 = require("./ssl.js");
Object.defineProperty(exports, "SSL", { enumerable: true, get: function () { return ssl_1.SSL; } });
var index_2 = require("./universal/index.js");
Object.defineProperty(exports, "Universal", { enumerable: true, get: function () { return index_2.Universal; } });
var verification_1 = require("./verification.js");
Object.defineProperty(exports, "VerificationResource", { enumerable: true, get: function () { return verification_1.VerificationResource; } });
