// cloudflare/resources/email-security/settings/settings.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../resource.mjs";
import * as AllowPoliciesAPI from "./allow-policies.mjs";
import { AllowPolicies, AllowPolicyListResponsesV4PagePaginationArray, } from "./allow-policies.mjs";
import * as BlockSendersAPI from "./block-senders.mjs";
import { BlockSenderListResponsesV4PagePaginationArray, BlockSenders, } from "./block-senders.mjs";
import * as DomainsAPI from "./domains.mjs";
import { DomainBulkDeleteResponsesSinglePage, DomainListResponsesV4PagePaginationArray, Domains, } from "./domains.mjs";
import * as ImpersonationRegistryAPI from "./impersonation-registry.mjs";
import { ImpersonationRegistry, ImpersonationRegistryListResponsesV4PagePaginationArray, } from "./impersonation-registry.mjs";
import * as TrustedDomainsAPI from "./trusted-domains.mjs";
import { TrustedDomainListResponsesV4PagePaginationArray, TrustedDomains, } from "./trusted-domains.mjs";
export class Settings extends APIResource {
    constructor() {
        super(...arguments);
        this.allowPolicies = new AllowPoliciesAPI.AllowPolicies(this._client);
        this.blockSenders = new BlockSendersAPI.BlockSenders(this._client);
        this.domains = new DomainsAPI.Domains(this._client);
        this.impersonationRegistry = new ImpersonationRegistryAPI.ImpersonationRegistry(this._client);
        this.trustedDomains = new TrustedDomainsAPI.TrustedDomains(this._client);
    }
}
Settings.AllowPolicies = AllowPolicies;
Settings.AllowPolicyListResponsesV4PagePaginationArray = AllowPolicyListResponsesV4PagePaginationArray;
Settings.BlockSenders = BlockSenders;
Settings.BlockSenderListResponsesV4PagePaginationArray = BlockSenderListResponsesV4PagePaginationArray;
Settings.Domains = Domains;
Settings.DomainListResponsesV4PagePaginationArray = DomainListResponsesV4PagePaginationArray;
Settings.DomainBulkDeleteResponsesSinglePage = DomainBulkDeleteResponsesSinglePage;
Settings.ImpersonationRegistry = ImpersonationRegistry;
Settings.ImpersonationRegistryListResponsesV4PagePaginationArray =
    ImpersonationRegistryListResponsesV4PagePaginationArray;
Settings.TrustedDomains = TrustedDomains;
Settings.TrustedDomainListResponsesV4PagePaginationArray = TrustedDomainListResponsesV4PagePaginationArray;
