// @mistralai/mistralai/src/models/components/agenthandoffdoneevent.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ClosedEnum } from "../../types/enums.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export const AgentHandoffDoneEventType = {
  AgentHandoffDone: "agent.handoff.done",
} as const;
export type AgentHandoffDoneEventType = ClosedEnum<
  typeof AgentHandoffDoneEventType
>;

export type AgentHandoffDoneEvent = {
  type: AgentHandoffDoneEventType | undefined;
  createdAt?: Date | undefined;
  outputIndex: number | undefined;
  id: string;
  nextAgentId: string;
  nextAgentName: string;
};

/** @internal */
export const AgentHandoffDoneEventType$inboundSchema: z.ZodNativeEnum<
  typeof AgentHandoffDoneEventType
> = z.nativeEnum(AgentHandoffDoneEventType);

/** @internal */
export const AgentHandoffDoneEvent$inboundSchema: z.ZodType<
  AgentHandoffDoneEvent,
  z.ZodTypeDef,
  unknown
> = z.object({
  type: AgentHandoffDoneEventType$inboundSchema.default("agent.handoff.done"),
  created_at: z.string().datetime({ offset: true }).transform(v => new Date(v))
    .optional(),
  output_index: z.number().int().default(0),
  id: z.string(),
  next_agent_id: z.string(),
  next_agent_name: z.string(),
}).transform((v) => {
  return remap$(v, {
    "created_at": "createdAt",
    "output_index": "outputIndex",
    "next_agent_id": "nextAgentId",
    "next_agent_name": "nextAgentName",
  });
});

export function agentHandoffDoneEventFromJSON(
  jsonString: string,
): SafeParseResult<AgentHandoffDoneEvent, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => AgentHandoffDoneEvent$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'AgentHandoffDoneEvent' from JSON`,
  );
}
