// @mistralai/mistralai/src/models/components/archiveftmodelout.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { safeParse } from "../../lib/schemas.js";
import { ClosedEnum } from "../../types/enums.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export const ArchiveFTModelOutObject = {
  Model: "model",
} as const;
export type ArchiveFTModelOutObject = ClosedEnum<
  typeof ArchiveFTModelOutObject
>;

export type ArchiveFTModelOut = {
  id: string;
  object: ArchiveFTModelOutObject | undefined;
  archived: boolean | undefined;
};

/** @internal */
export const ArchiveFTModelOutObject$inboundSchema: z.ZodNativeEnum<
  typeof ArchiveFTModelOutObject
> = z.nativeEnum(ArchiveFTModelOutObject);

/** @internal */
export const ArchiveFTModelOut$inboundSchema: z.ZodType<
  ArchiveFTModelOut,
  z.ZodTypeDef,
  unknown
> = z.object({
  id: z.string(),
  object: ArchiveFTModelOutObject$inboundSchema.default("model"),
  archived: z.boolean().default(true),
});

export function archiveFTModelOutFromJSON(
  jsonString: string,
): SafeParseResult<ArchiveFTModelOut, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ArchiveFTModelOut$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ArchiveFTModelOut' from JSON`,
  );
}
