// @mistralai/mistralai/src/models/components/conversationappendrequest.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { remap as remap$ } from "../../lib/primitives.js";
import { ClosedEnum } from "../../types/enums.js";
import {
  CompletionArgs,
  CompletionArgs$Outbound,
  CompletionArgs$outboundSchema,
} from "./completionargs.js";
import {
  ConversationInputs,
  ConversationInputs$Outbound,
  ConversationInputs$outboundSchema,
} from "./conversationinputs.js";

export const ConversationAppendRequestHandoffExecution = {
  Client: "client",
  Server: "server",
} as const;
export type ConversationAppendRequestHandoffExecution = ClosedEnum<
  typeof ConversationAppendRequestHandoffExecution
>;

export type ConversationAppendRequest = {
  inputs: ConversationInputs;
  stream?: boolean | undefined;
  /**
   * Whether to store the results into our servers or not.
   */
  store?: boolean | undefined;
  handoffExecution?: ConversationAppendRequestHandoffExecution | undefined;
  /**
   * White-listed arguments from the completion API
   */
  completionArgs?: CompletionArgs | undefined;
};

/** @internal */
export const ConversationAppendRequestHandoffExecution$outboundSchema:
  z.ZodNativeEnum<typeof ConversationAppendRequestHandoffExecution> = z
    .nativeEnum(ConversationAppendRequestHandoffExecution);

/** @internal */
export type ConversationAppendRequest$Outbound = {
  inputs: ConversationInputs$Outbound;
  stream: boolean;
  store: boolean;
  handoff_execution: string;
  completion_args?: CompletionArgs$Outbound | undefined;
};

/** @internal */
export const ConversationAppendRequest$outboundSchema: z.ZodType<
  ConversationAppendRequest$Outbound,
  z.ZodTypeDef,
  ConversationAppendRequest
> = z.object({
  inputs: ConversationInputs$outboundSchema,
  stream: z.boolean().default(false),
  store: z.boolean().default(true),
  handoffExecution: ConversationAppendRequestHandoffExecution$outboundSchema
    .default("server"),
  completionArgs: CompletionArgs$outboundSchema.optional(),
}).transform((v) => {
  return remap$(v, {
    handoffExecution: "handoff_execution",
    completionArgs: "completion_args",
  });
});

export function conversationAppendRequestToJSON(
  conversationAppendRequest: ConversationAppendRequest,
): string {
  return JSON.stringify(
    ConversationAppendRequest$outboundSchema.parse(conversationAppendRequest),
  );
}
