// @mistralai/mistralai/src/models/components/processingstatusout.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export type ProcessingStatusOut = {
  documentId: string;
  processingStatus: string;
};

/** @internal */
export const ProcessingStatusOut$inboundSchema: z.ZodType<
  ProcessingStatusOut,
  z.ZodTypeDef,
  unknown
> = z.object({
  document_id: z.string(),
  processing_status: z.string(),
}).transform((v) => {
  return remap$(v, {
    "document_id": "documentId",
    "processing_status": "processingStatus",
  });
});

export function processingStatusOutFromJSON(
  jsonString: string,
): SafeParseResult<ProcessingStatusOut, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ProcessingStatusOut$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ProcessingStatusOut' from JSON`,
  );
}
