// @mistralai/mistralai/src/models/components/toolexecutiondeltaevent.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ClosedEnum } from "../../types/enums.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";
import {
  BuiltInConnectors,
  BuiltInConnectors$inboundSchema,
} from "./builtinconnectors.js";

export const ToolExecutionDeltaEventType = {
  ToolExecutionDelta: "tool.execution.delta",
} as const;
export type ToolExecutionDeltaEventType = ClosedEnum<
  typeof ToolExecutionDeltaEventType
>;

export type ToolExecutionDeltaEventName = BuiltInConnectors | string;

export type ToolExecutionDeltaEvent = {
  type: ToolExecutionDeltaEventType | undefined;
  createdAt?: Date | undefined;
  outputIndex: number | undefined;
  id: string;
  name: BuiltInConnectors | string;
  arguments: string;
};

/** @internal */
export const ToolExecutionDeltaEventType$inboundSchema: z.ZodNativeEnum<
  typeof ToolExecutionDeltaEventType
> = z.nativeEnum(ToolExecutionDeltaEventType);

/** @internal */
export const ToolExecutionDeltaEventName$inboundSchema: z.ZodType<
  ToolExecutionDeltaEventName,
  z.ZodTypeDef,
  unknown
> = z.union([BuiltInConnectors$inboundSchema, z.string()]);

export function toolExecutionDeltaEventNameFromJSON(
  jsonString: string,
): SafeParseResult<ToolExecutionDeltaEventName, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ToolExecutionDeltaEventName$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ToolExecutionDeltaEventName' from JSON`,
  );
}

/** @internal */
export const ToolExecutionDeltaEvent$inboundSchema: z.ZodType<
  ToolExecutionDeltaEvent,
  z.ZodTypeDef,
  unknown
> = z.object({
  type: ToolExecutionDeltaEventType$inboundSchema.default(
    "tool.execution.delta",
  ),
  created_at: z.string().datetime({ offset: true }).transform(v => new Date(v))
    .optional(),
  output_index: z.number().int().default(0),
  id: z.string(),
  name: z.union([BuiltInConnectors$inboundSchema, z.string()]),
  arguments: z.string(),
}).transform((v) => {
  return remap$(v, {
    "created_at": "createdAt",
    "output_index": "outputIndex",
  });
});

export function toolExecutionDeltaEventFromJSON(
  jsonString: string,
): SafeParseResult<ToolExecutionDeltaEvent, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => ToolExecutionDeltaEvent$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'ToolExecutionDeltaEvent' from JSON`,
  );
}
