// @mistralai/mistralai/src/models/components/unarchiveftmodelout.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod/v3";
import { safeParse } from "../../lib/schemas.js";
import { ClosedEnum } from "../../types/enums.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

export const UnarchiveFTModelOutObject = {
  Model: "model",
} as const;
export type UnarchiveFTModelOutObject = ClosedEnum<
  typeof UnarchiveFTModelOutObject
>;

export type UnarchiveFTModelOut = {
  id: string;
  object: UnarchiveFTModelOutObject | undefined;
  archived: boolean | undefined;
};

/** @internal */
export const UnarchiveFTModelOutObject$inboundSchema: z.ZodNativeEnum<
  typeof UnarchiveFTModelOutObject
> = z.nativeEnum(UnarchiveFTModelOutObject);

/** @internal */
export const UnarchiveFTModelOut$inboundSchema: z.ZodType<
  UnarchiveFTModelOut,
  z.ZodTypeDef,
  unknown
> = z.object({
  id: z.string(),
  object: UnarchiveFTModelOutObject$inboundSchema.default("model"),
  archived: z.boolean().default(false),
});

export function unarchiveFTModelOutFromJSON(
  jsonString: string,
): SafeParseResult<UnarchiveFTModelOut, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => UnarchiveFTModelOut$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'UnarchiveFTModelOut' from JSON`,
  );
}
