// cloudflare/resources/zero-trust/dex/commands/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Quota = exports.Downloads = exports.Devices = exports.DeviceListResponsesV4PagePagination = exports.Commands = exports.CommandListResponsesV4PagePagination = void 0;
var commands_1 = require("./commands.js");
Object.defineProperty(exports, "CommandListResponsesV4PagePagination", { enumerable: true, get: function () { return commands_1.CommandListResponsesV4PagePagination; } });
Object.defineProperty(exports, "Commands", { enumerable: true, get: function () { return commands_1.Commands; } });
var devices_1 = require("./devices.js");
Object.defineProperty(exports, "DeviceListResponsesV4PagePagination", { enumerable: true, get: function () { return devices_1.DeviceListResponsesV4PagePagination; } });
Object.defineProperty(exports, "Devices", { enumerable: true, get: function () { return devices_1.Devices; } });
var downloads_1 = require("./downloads.js");
Object.defineProperty(exports, "Downloads", { enumerable: true, get: function () { return downloads_1.Downloads; } });
var quota_1 = require("./quota.js");
Object.defineProperty(exports, "Quota", { enumerable: true, get: function () { return quota_1.Quota; } });
