// cloudflare/resources/logpush/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validate = exports.Ownership = exports.Jobs = exports.LogpushJobsSinglePage = exports.Logpush = exports.Edge = exports.InstantLogpushJobsSinglePage = exports.Datasets = void 0;
var index_1 = require("./datasets/index.js");
Object.defineProperty(exports, "Datasets", { enumerable: true, get: function () { return index_1.Datasets; } });
var edge_1 = require("./edge.js");
Object.defineProperty(exports, "InstantLogpushJobsSinglePage", { enumerable: true, get: function () { return edge_1.InstantLogpushJobsSinglePage; } });
Object.defineProperty(exports, "Edge", { enumerable: true, get: function () { return edge_1.Edge; } });
var logpush_1 = require("./logpush.js");
Object.defineProperty(exports, "Logpush", { enumerable: true, get: function () { return logpush_1.Logpush; } });
var jobs_1 = require("./jobs.js");
Object.defineProperty(exports, "LogpushJobsSinglePage", { enumerable: true, get: function () { return jobs_1.LogpushJobsSinglePage; } });
Object.defineProperty(exports, "Jobs", { enumerable: true, get: function () { return jobs_1.Jobs; } });
var ownership_1 = require("./ownership.js");
Object.defineProperty(exports, "Ownership", { enumerable: true, get: function () { return ownership_1.Ownership; } });
var validate_1 = require("./validate.js");
Object.defineProperty(exports, "Validate", { enumerable: true, get: function () { return validate_1.Validate; } });
