// cloudflare/resources/alerting/destinations/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = exports.WebhooksSinglePage = exports.PagerdutyResource = exports.PagerdutiesSinglePage = exports.Eligible = exports.Destinations = void 0;
var destinations_1 = require("./destinations.js");
Object.defineProperty(exports, "Destinations", { enumerable: true, get: function () { return destinations_1.Destinations; } });
var eligible_1 = require("./eligible.js");
Object.defineProperty(exports, "Eligible", { enumerable: true, get: function () { return eligible_1.Eligible; } });
var pagerduty_1 = require("./pagerduty.js");
Object.defineProperty(exports, "PagerdutiesSinglePage", { enumerable: true, get: function () { return pagerduty_1.PagerdutiesSinglePage; } });
Object.defineProperty(exports, "PagerdutyResource", { enumerable: true, get: function () { return pagerduty_1.PagerdutyResource; } });
var webhooks_1 = require("./webhooks.js");
Object.defineProperty(exports, "WebhooksSinglePage", { enumerable: true, get: function () { return webhooks_1.WebhooksSinglePage; } });
Object.defineProperty(exports, "Webhooks", { enumerable: true, get: function () { return webhooks_1.Webhooks; } });
