// cloudflare/resources/magic-transit/sites/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.WANs = exports.WANsSinglePage = exports.Sites = exports.SitesSinglePage = exports.LANs = exports.LANsSinglePage = exports.ACLs = exports.ACLsSinglePage = void 0;
var acls_1 = require("./acls.js");
Object.defineProperty(exports, "ACLsSinglePage", { enumerable: true, get: function () { return acls_1.ACLsSinglePage; } });
Object.defineProperty(exports, "ACLs", { enumerable: true, get: function () { return acls_1.ACLs; } });
var lans_1 = require("./lans.js");
Object.defineProperty(exports, "LANsSinglePage", { enumerable: true, get: function () { return lans_1.LANsSinglePage; } });
Object.defineProperty(exports, "LANs", { enumerable: true, get: function () { return lans_1.LANs; } });
var sites_1 = require("./sites.js");
Object.defineProperty(exports, "SitesSinglePage", { enumerable: true, get: function () { return sites_1.SitesSinglePage; } });
Object.defineProperty(exports, "Sites", { enumerable: true, get: function () { return sites_1.Sites; } });
var wans_1 = require("./wans.js");
Object.defineProperty(exports, "WANsSinglePage", { enumerable: true, get: function () { return wans_1.WANsSinglePage; } });
Object.defineProperty(exports, "WANs", { enumerable: true, get: function () { return wans_1.WANs; } });
