// @mistralai/mistralai/src/models/components/ocrtableobject.ts
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */

import * as z from "zod";
import { safeParse } from "../../lib/schemas.js";
import { ClosedEnum } from "../../types/enums.js";
import { Result as SafeParseResult } from "../../types/fp.js";
import { SDKValidationError } from "../errors/sdkvalidationerror.js";

/**
 * Format of the table
 */
export const Format = {
  Markdown: "markdown",
  Html: "html",
} as const;
/**
 * Format of the table
 */
export type Format = ClosedEnum<typeof Format>;

export type OCRTableObject = {
  /**
   * Table ID for extracted table in a page
   */
  id: string;
  /**
   * Content of the table in the given format
   */
  content: string;
  /**
   * Format of the table
   */
  format: Format;
};

/** @internal */
export const Format$inboundSchema: z.ZodNativeEnum<typeof Format> = z
  .nativeEnum(Format);

/** @internal */
export const Format$outboundSchema: z.ZodNativeEnum<typeof Format> =
  Format$inboundSchema;

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace Format$ {
  /** @deprecated use `Format$inboundSchema` instead. */
  export const inboundSchema = Format$inboundSchema;
  /** @deprecated use `Format$outboundSchema` instead. */
  export const outboundSchema = Format$outboundSchema;
}

/** @internal */
export const OCRTableObject$inboundSchema: z.ZodType<
  OCRTableObject,
  z.ZodTypeDef,
  unknown
> = z.object({
  id: z.string(),
  content: z.string(),
  format: Format$inboundSchema,
});

/** @internal */
export type OCRTableObject$Outbound = {
  id: string;
  content: string;
  format: string;
};

/** @internal */
export const OCRTableObject$outboundSchema: z.ZodType<
  OCRTableObject$Outbound,
  z.ZodTypeDef,
  OCRTableObject
> = z.object({
  id: z.string(),
  content: z.string(),
  format: Format$outboundSchema,
});

/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
export namespace OCRTableObject$ {
  /** @deprecated use `OCRTableObject$inboundSchema` instead. */
  export const inboundSchema = OCRTableObject$inboundSchema;
  /** @deprecated use `OCRTableObject$outboundSchema` instead. */
  export const outboundSchema = OCRTableObject$outboundSchema;
  /** @deprecated use `OCRTableObject$Outbound` instead. */
  export type Outbound = OCRTableObject$Outbound;
}

export function ocrTableObjectToJSON(ocrTableObject: OCRTableObject): string {
  return JSON.stringify(OCRTableObject$outboundSchema.parse(ocrTableObject));
}

export function ocrTableObjectFromJSON(
  jsonString: string,
): SafeParseResult<OCRTableObject, SDKValidationError> {
  return safeParse(
    jsonString,
    (x) => OCRTableObject$inboundSchema.parse(JSON.parse(x)),
    `Failed to parse 'OCRTableObject' from JSON`,
  );
}
