// cloudflare/resources/r2/buckets/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SippyResource = exports.Metrics = exports.Locks = exports.Lifecycle = exports.EventNotifications = exports.Domains = exports.CORS = exports.Buckets = void 0;
var buckets_1 = require("./buckets.js");
Object.defineProperty(exports, "Buckets", { enumerable: true, get: function () { return buckets_1.Buckets; } });
var cors_1 = require("./cors.js");
Object.defineProperty(exports, "CORS", { enumerable: true, get: function () { return cors_1.CORS; } });
var index_1 = require("./domains/index.js");
Object.defineProperty(exports, "Domains", { enumerable: true, get: function () { return index_1.Domains; } });
var event_notifications_1 = require("./event-notifications.js");
Object.defineProperty(exports, "EventNotifications", { enumerable: true, get: function () { return event_notifications_1.EventNotifications; } });
var lifecycle_1 = require("./lifecycle.js");
Object.defineProperty(exports, "Lifecycle", { enumerable: true, get: function () { return lifecycle_1.Lifecycle; } });
var locks_1 = require("./locks.js");
Object.defineProperty(exports, "Locks", { enumerable: true, get: function () { return locks_1.Locks; } });
var metrics_1 = require("./metrics.js");
Object.defineProperty(exports, "Metrics", { enumerable: true, get: function () { return metrics_1.Metrics; } });
var sippy_1 = require("./sippy.js");
Object.defineProperty(exports, "SippyResource", { enumerable: true, get: function () { return sippy_1.SippyResource; } });
