// @cloudbase/ai/src/models/Ark/type.ts
type UserMessage = {
  role: 'user'
  content: string
}

type SystemMessage = {
  role: 'system'
  content: string
}

type ToolMessage = {
  role: 'tool'
  content: string
  tool_call_id: string
}

type AssistantMessage =
  | {
    role: 'assistant'
    content: string
  }
  | {
    role: 'assistant'
    tool_calls: Array<ToolCall>
    content?: string
  }

type ToolCall = {
  id: string
  type: string
  function: {
    name: string
    arguments: string
  }
}

type ArkMessage = UserMessage | SystemMessage | AssistantMessage | ToolMessage

export type ArkInputData = {
  model: string
  messages: Array<ArkMessage>
  max_tokens?: number
  stop?: string | Array<string>
  frequency_penalty?: number
  temperature?: number
  top_p?: number
  tools?: Array<Tool>
}

type Tool = {
  type: string
  function: {
    name: string
    description?: string
    parameters?: object
  }
}

export type ArkGenerateTextOutput = {
  choices?: Array<{
    finish_reason: string
    message?: AssistantMessage
  }>
  usage?: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}

// todo: more precise
export type ArkStreamTextOutput = {
  choices?: Array<{
    finish_reason: string
    delta: AssistantMessage
  }>
  usage?: {
    prompt_tokens: number
    completion_tokens: number
    total_tokens: number
  }
}
