// @cloudbase/manager-node/lib/constant.js
"use strict";
// // cloudbase cli 配置的字段名
// export class ConfigItems {
//     static credentail = 'credential'
// }
Object.defineProperty(exports, "__esModule", { value: true });
exports.SCF_TEMP_COS = exports.COS_SDK_KEEPALIVE = exports.COS_SDK_PROTOCOL = exports.INTERNAL_ENDPOINT_REGION = exports.USE_INTERNAL_ENDPOINT = exports.SCF_STATUS = exports.ROLE_NAME = exports.PUBLIC_RSA_KEY = exports.ERROR = exports.SERVICE_TYPE = exports.ENDPOINT = exports.RUN_ENV = exports.SDK_VERSION = exports.ENV_NAME = void 0;
exports.ENV_NAME = {
    ENV_SECRETID: 'TENCENTCLOUD_SECRETID',
    ENV_SECRETKEY: 'TENCENTCLOUD_SECRETKEY',
    ENV_SESSIONTOKEN: 'TENCENTCLOUD_SESSIONTOKEN',
    ENV_TCB_ENV_ID: 'TENCENTCLOUD_TCB_ENVID',
    ENV_RUNENV: 'TENCENTCLOUD_RUNENV',
    ENV_RUNENV_SCF: 'TENCENTCLOUD_RUNENV=SCF'
};
exports.SDK_VERSION = 'TCB-NODE-MANAGER/1.0.O';
exports.RUN_ENV = {
    SCF: 'SCF'
};
exports.ENDPOINT = {
    TCB: 'tcb.tencentcloudapi.com',
    SCF: 'scf.tencentcloudapi.com',
    COS: 'cos.tencentcloudapi.com',
    FLEXDB: 'flexdb.tencentcloudapi.com'
};
exports.SERVICE_TYPE = {
    TCB: 'tcb'
};
exports.ERROR = {
    MISS_SECRET_INFO_IN_ENV: 'MISS_SECRET_INFO_IN_ENV',
    MISS_SECRET_INFO_IN_ARGS: 'MISS_SECRET_INFO_IN_ARGS',
    CURRENT_ENVIRONMENT_IS_NULL: 'CURRENT_ENVIRONMENT_IS_NULL',
    ENV_ID_NOT_EXISTS: 'ENV_ID_NOT_EXISTS'
};
exports.PUBLIC_RSA_KEY = `
-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC0ZLB0ZpWWFsHPnDDw++Nc2wI3
nl2uyOrIJ5FUfxt4GAmt1Faf5pgMxAnL9exEUrrUDUX8Ri1R0KyfnHQQwCvKt8T8
bgILIJe9UB8e9dvFqgqH2oA8Vqwi0YqDcvFLFJk2BJbm/0QYtZ563FumW8LEXAgu
UeHi/0OZN9vQ33jWMQIDAQAB
-----END PUBLIC KEY-----
`;
exports.ROLE_NAME = {
    TCB: 'TCB_QcsRole'
};
exports.SCF_STATUS = {
    ACTIVE: 'Active',
    CREATING: 'Creating',
    UPDATING: 'Updating',
    CREATE_FAILED: 'CreateFailed',
    UPDATE_FAILED: 'UpdateFailed',
    PUBLISHING: 'Publishing',
    PUBLISH_FAILED: 'PublishFailed',
    DELETING: 'Deleting',
    DELETE_FAILED: 'DeleteFailed'
};
// 是否使用内网域名
exports.USE_INTERNAL_ENDPOINT = "USE_INTERNAL_ENDPOINT" in process.env;
exports.INTERNAL_ENDPOINT_REGION = process.env.INTERNAL_ENDPOINT_REGION;
exports.COS_SDK_PROTOCOL = process.env.COS_SDK_PROTOCOL;
exports.COS_SDK_KEEPALIVE = process.env.COS_SDK_KEEPALIVE;
// SCF 临时 COS 配置（用于函数代码上传）
exports.SCF_TEMP_COS = {
    APPID: '1253665819',
    REGION: 'ap-shanghai'
};
