// @authing/nextjs/dist/api/callback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCallbackApi = void 0;
require("../types");
function createCallbackApi({ appDomain, clientId, clientSecret, failureRedirect, successRedirect }) {
    return async function handler(req, res) {
        var _a, _b;
        const { code } = req.query;
        if (!code) {
            res.redirect(302, failureRedirect !== null && failureRedirect !== void 0 ? failureRedirect : '/login');
            return;
        }
        const body = {
            client_id: clientId,
            client_secret: clientSecret,
            grant_type: 'authorization_code',
            code
        };
        const formBody = [];
        // eslint-disable-next-line
        for (const property in body) {
            const encodedKey = encodeURIComponent(property);
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
            const encodedValue = encodeURIComponent(body[property]);
            formBody.push(`${encodedKey}=${encodedValue}`);
        }
        const oidcRes = await fetch(`${appDomain}/oidc/token`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
            },
            body: formBody.join('&')
        });
        const oidcToken = (await oidcRes.json());
        if (oidcToken.error) {
            console.error(oidcToken);
            res.redirect(302, failureRedirect !== null && failureRedirect !== void 0 ? failureRedirect : '/login');
            return;
        }
        const resInfo = await fetch(`${appDomain}/oidc/me?access_token=${oidcToken.access_token}`);
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        Object.assign(req === null || req === void 0 ? void 0 : req.session, { user: await resInfo.json() });
        await ((_b = (_a = req.session) === null || _a === void 0 ? void 0 : _a.save) === null || _b === void 0 ? void 0 : _b.call(_a));
        res.redirect(302, successRedirect !== null && successRedirect !== void 0 ? successRedirect : '/');
    };
}
exports.createCallbackApi = createCallbackApi;
