// @cloudbase/agent-tools/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsafeLocalCodeExecutor = exports.BuiltInCodeExecutor = exports.ToolResult = exports.ToolkitManager = exports.toolkitManager = exports.tool = exports.DynamicTool = exports.BaseToolkit = exports.BaseTool = void 0;
// Core tool system
var utils_1 = require("./utils");
Object.defineProperty(exports, "BaseTool", { enumerable: true, get: function () { return utils_1.BaseTool; } });
Object.defineProperty(exports, "BaseToolkit", { enumerable: true, get: function () { return utils_1.BaseToolkit; } });
Object.defineProperty(exports, "DynamicTool", { enumerable: true, get: function () { return utils_1.DynamicTool; } });
Object.defineProperty(exports, "tool", { enumerable: true, get: function () { return utils_1.tool; } });
Object.defineProperty(exports, "toolkitManager", { enumerable: true, get: function () { return utils_1.toolkitManager; } });
Object.defineProperty(exports, "ToolkitManager", { enumerable: true, get: function () { return utils_1.ToolkitManager; } });
Object.defineProperty(exports, "ToolResult", { enumerable: true, get: function () { return utils_1.ToolResult; } });
// Code executors
var built_in_code_executor_1 = require("./code_executers/built_in_code_executor");
Object.defineProperty(exports, "BuiltInCodeExecutor", { enumerable: true, get: function () { return built_in_code_executor_1.BuiltInCodeExecutor; } });
var unsafe_local_code_executor_1 = require("./code_executers/unsafe_local_code_executor");
Object.defineProperty(exports, "UnsafeLocalCodeExecutor", { enumerable: true, get: function () { return unsafe_local_code_executor_1.UnsafeLocalCodeExecutor; } });
// Bash tools
__exportStar(require("./bash"), exports);
// File system tools
__exportStar(require("./fs"), exports);
// Browser tools
__exportStar(require("./browser"), exports);
// MCP integration
__exportStar(require("./mcp"), exports);
