// @copilotkitnext/shared/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  DEFAULT_AGENT_ID: () => DEFAULT_AGENT_ID,
  finalizeRunEvents: () => finalizeRunEvents,
  logger: () => logger,
  partialJSONParse: () => partialJSONParse,
  randomUUID: () => randomUUID
});
module.exports = __toCommonJS(index_exports);

// src/utils.ts
var import_uuid = require("uuid");
var PartialJSON = __toESM(require("partial-json"));
function randomUUID() {
  return (0, import_uuid.v4)();
}
function partialJSONParse(json) {
  try {
    return PartialJSON.parse(json);
  } catch (error) {
    return {};
  }
}

// src/logger.ts
var logger = console;

// src/constants.ts
var DEFAULT_AGENT_ID = "default";

// src/finalize-events.ts
var import_client = require("@ag-ui/client");
var defaultStopMessage = "Run stopped by user";
var defaultAbruptEndMessage = "Run ended without emitting a terminal event";
function finalizeRunEvents(events, options = {}) {
  const { stopRequested = false, interruptionMessage } = options;
  const resolvedStopMessage = interruptionMessage ?? defaultStopMessage;
  const resolvedAbruptMessage = interruptionMessage && interruptionMessage !== defaultStopMessage ? interruptionMessage : defaultAbruptEndMessage;
  const appended = [];
  const openMessageIds = /* @__PURE__ */ new Set();
  const openToolCalls = /* @__PURE__ */ new Map();
  for (const event of events) {
    switch (event.type) {
      case import_client.EventType.TEXT_MESSAGE_START: {
        const messageId = event.messageId;
        if (typeof messageId === "string") {
          openMessageIds.add(messageId);
        }
        break;
      }
      case import_client.EventType.TEXT_MESSAGE_END: {
        const messageId = event.messageId;
        if (typeof messageId === "string") {
          openMessageIds.delete(messageId);
        }
        break;
      }
      case import_client.EventType.TOOL_CALL_START: {
        const toolCallId = event.toolCallId;
        if (typeof toolCallId === "string") {
          openToolCalls.set(toolCallId, {
            hasEnd: false,
            hasResult: false
          });
        }
        break;
      }
      case import_client.EventType.TOOL_CALL_END: {
        const toolCallId = event.toolCallId;
        const info = toolCallId ? openToolCalls.get(toolCallId) : void 0;
        if (info) {
          info.hasEnd = true;
        }
        break;
      }
      case import_client.EventType.TOOL_CALL_RESULT: {
        const toolCallId = event.toolCallId;
        const info = toolCallId ? openToolCalls.get(toolCallId) : void 0;
        if (info) {
          info.hasResult = true;
        }
        break;
      }
      default:
        break;
    }
  }
  const hasRunFinished = events.some((event) => event.type === import_client.EventType.RUN_FINISHED);
  const hasRunError = events.some((event) => event.type === import_client.EventType.RUN_ERROR);
  const hasTerminalEvent = hasRunFinished || hasRunError;
  const terminalEventMissing = !hasTerminalEvent;
  for (const messageId of openMessageIds) {
    const endEvent = {
      type: import_client.EventType.TEXT_MESSAGE_END,
      messageId
    };
    events.push(endEvent);
    appended.push(endEvent);
  }
  for (const [toolCallId, info] of openToolCalls) {
    if (!info.hasEnd) {
      const endEvent = {
        type: import_client.EventType.TOOL_CALL_END,
        toolCallId
      };
      events.push(endEvent);
      appended.push(endEvent);
    }
    if (terminalEventMissing && !info.hasResult) {
      const resultEvent = {
        type: import_client.EventType.TOOL_CALL_RESULT,
        toolCallId,
        messageId: `${toolCallId ?? randomUUID()}-result`,
        role: "tool",
        content: JSON.stringify(
          stopRequested ? {
            status: "stopped",
            reason: "stop_requested",
            message: resolvedStopMessage
          } : {
            status: "error",
            reason: "missing_terminal_event",
            message: resolvedAbruptMessage
          }
        )
      };
      events.push(resultEvent);
      appended.push(resultEvent);
    }
  }
  if (terminalEventMissing) {
    if (stopRequested) {
      const finishedEvent = {
        type: import_client.EventType.RUN_FINISHED
      };
      events.push(finishedEvent);
      appended.push(finishedEvent);
    } else {
      const errorEvent = {
        type: import_client.EventType.RUN_ERROR,
        message: resolvedAbruptMessage,
        code: "INCOMPLETE_STREAM"
      };
      events.push(errorEvent);
      appended.push(errorEvent);
    }
  }
  return appended;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_AGENT_ID,
  finalizeRunEvents,
  logger,
  partialJSONParse,
  randomUUID
});
