// @mistralai/mistralai/models/components/documentupdatein.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentUpdateIn$ = exports.DocumentUpdateIn$outboundSchema = exports.DocumentUpdateIn$inboundSchema = exports.Attributes$ = exports.Attributes$outboundSchema = exports.Attributes$inboundSchema = void 0;
exports.attributesToJSON = attributesToJSON;
exports.attributesFromJSON = attributesFromJSON;
exports.documentUpdateInToJSON = documentUpdateInToJSON;
exports.documentUpdateInFromJSON = documentUpdateInFromJSON;
const z = __importStar(require("zod"));
const schemas_js_1 = require("../../lib/schemas.js");
/** @internal */
exports.Attributes$inboundSchema = z.union([
    z.boolean(),
    z.string(),
    z.number().int(),
    z.number(),
    z.string().datetime({ offset: true }).transform(v => new Date(v)),
    z.array(z.string()),
    z.array(z.number().int()),
    z.array(z.number()),
    z.array(z.boolean()),
]);
/** @internal */
exports.Attributes$outboundSchema = z.union([
    z.boolean(),
    z.string(),
    z.number().int(),
    z.number(),
    z.date().transform(v => v.toISOString()),
    z.array(z.string()),
    z.array(z.number().int()),
    z.array(z.number()),
    z.array(z.boolean()),
]);
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var Attributes$;
(function (Attributes$) {
    /** @deprecated use `Attributes$inboundSchema` instead. */
    Attributes$.inboundSchema = exports.Attributes$inboundSchema;
    /** @deprecated use `Attributes$outboundSchema` instead. */
    Attributes$.outboundSchema = exports.Attributes$outboundSchema;
})(Attributes$ || (exports.Attributes$ = Attributes$ = {}));
function attributesToJSON(attributes) {
    return JSON.stringify(exports.Attributes$outboundSchema.parse(attributes));
}
function attributesFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.Attributes$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'Attributes' from JSON`);
}
/** @internal */
exports.DocumentUpdateIn$inboundSchema = z.object({
    name: z.nullable(z.string()).optional(),
    attributes: z.nullable(z.record(z.union([
        z.boolean(),
        z.string(),
        z.number().int(),
        z.number(),
        z.string().datetime({ offset: true }).transform(v => new Date(v)),
        z.array(z.string()),
        z.array(z.number().int()),
        z.array(z.number()),
        z.array(z.boolean()),
    ]))).optional(),
});
/** @internal */
exports.DocumentUpdateIn$outboundSchema = z.object({
    name: z.nullable(z.string()).optional(),
    attributes: z.nullable(z.record(z.union([
        z.boolean(),
        z.string(),
        z.number().int(),
        z.number(),
        z.date().transform(v => v.toISOString()),
        z.array(z.string()),
        z.array(z.number().int()),
        z.array(z.number()),
        z.array(z.boolean()),
    ]))).optional(),
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var DocumentUpdateIn$;
(function (DocumentUpdateIn$) {
    /** @deprecated use `DocumentUpdateIn$inboundSchema` instead. */
    DocumentUpdateIn$.inboundSchema = exports.DocumentUpdateIn$inboundSchema;
    /** @deprecated use `DocumentUpdateIn$outboundSchema` instead. */
    DocumentUpdateIn$.outboundSchema = exports.DocumentUpdateIn$outboundSchema;
})(DocumentUpdateIn$ || (exports.DocumentUpdateIn$ = DocumentUpdateIn$ = {}));
function documentUpdateInToJSON(documentUpdateIn) {
    return JSON.stringify(exports.DocumentUpdateIn$outboundSchema.parse(documentUpdateIn));
}
function documentUpdateInFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.DocumentUpdateIn$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'DocumentUpdateIn' from JSON`);
}
