// @cloudbase/ai/dist/cjs/models/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolMap = exports.ReactModel = exports.DefaultImageModel = exports.DefaultSimpleModel = exports.HunYuanOpenSimpleModel = exports.HunYuanExpSimpleModel = exports.MoonshotSimpleModel = exports.YiSimpleModel = exports.DSSimpleModel = exports.ArkSimpleModel = exports.HunYuanSimpleModel = exports.HunYuanBetaSimpleModel = exports.ZhiPuSimpleModel = exports.MODELS = void 0;
var index_1 = require("./ZhiPu/index");
Object.defineProperty(exports, "ZhiPuSimpleModel", { enumerable: true, get: function () { return index_1.ZhiPuSimpleModel; } });
var index_2 = require("./HunYuanBeta/index");
Object.defineProperty(exports, "HunYuanBetaSimpleModel", { enumerable: true, get: function () { return index_2.HunYuanBetaSimpleModel; } });
var index_3 = require("./HunYuan/index");
Object.defineProperty(exports, "HunYuanSimpleModel", { enumerable: true, get: function () { return index_3.HunYuanSimpleModel; } });
var index_4 = require("./Ark/index");
Object.defineProperty(exports, "ArkSimpleModel", { enumerable: true, get: function () { return index_4.ArkSimpleModel; } });
var index_5 = require("./DashScope/index");
Object.defineProperty(exports, "DSSimpleModel", { enumerable: true, get: function () { return index_5.DSSimpleModel; } });
var index_6 = require("./Yi/index");
Object.defineProperty(exports, "YiSimpleModel", { enumerable: true, get: function () { return index_6.YiSimpleModel; } });
var index_7 = require("./Moonshot/index");
Object.defineProperty(exports, "MoonshotSimpleModel", { enumerable: true, get: function () { return index_7.MoonshotSimpleModel; } });
var index_8 = require("./HunYuanExp/index");
Object.defineProperty(exports, "HunYuanExpSimpleModel", { enumerable: true, get: function () { return index_8.HunYuanExpSimpleModel; } });
var index_9 = require("./HunYuanOpen/index");
Object.defineProperty(exports, "HunYuanOpenSimpleModel", { enumerable: true, get: function () { return index_9.HunYuanOpenSimpleModel; } });
var DeepSeek_1 = require("./DeepSeek");
var Default_1 = require("./Default");
Object.defineProperty(exports, "DefaultSimpleModel", { enumerable: true, get: function () { return Default_1.DefaultSimpleModel; } });
var image_model_1 = require("./image-model");
Object.defineProperty(exports, "DefaultImageModel", { enumerable: true, get: function () { return image_model_1.DefaultImageModel; } });
exports.MODELS = {
    hunyuan: index_3.HunYuanSimpleModel,
    'hunyuan-beta': index_2.HunYuanBetaSimpleModel,
    ark: index_4.ArkSimpleModel,
    dashscope: index_5.DSSimpleModel,
    '01-ai': index_6.YiSimpleModel,
    moonshot: index_7.MoonshotSimpleModel,
    zhipu: index_1.ZhiPuSimpleModel,
    'hunyuan-exp': index_8.HunYuanExpSimpleModel,
    'hunyuan-open': index_9.HunYuanOpenSimpleModel,
    deepseek: DeepSeek_1.DeepSeekSimpleModel,
};
var model_1 = require("./model");
Object.defineProperty(exports, "ReactModel", { enumerable: true, get: function () { return model_1.ReactModel; } });
Object.defineProperty(exports, "toolMap", { enumerable: true, get: function () { return model_1.toolMap; } });
