// @cloudbase/ai/dist/cjs/models/model.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toolMap = exports.ReactModel = void 0;
var utils_1 = require("../utils");
function processInput(obj) {
    var _a;
    var onStepFinish = obj.onStepFinish, abortSignal = obj.abortSignal, maxSteps = obj.maxSteps, topP = obj.topP, toolChoice = obj.toolChoice, b = __rest(obj, ["onStepFinish", "abortSignal", "maxSteps", "topP", "toolChoice"]);
    if (maxSteps != null && maxSteps < 1) {
        throw new Error('`maxSteps` muse be greater than 0.');
    }
    return [
        { onStepFinish: onStepFinish, abortSignal: abortSignal, maxSteps: maxSteps },
        __assign(__assign({}, b), { tools: (_a = b.tools) === null || _a === void 0 ? void 0 : _a.map(function (tool) {
                if ('fn' in tool) {
                    return (0, utils_1.functionToolToModelTool)(tool);
                }
                return tool;
            }), top_p: topP != null ? topP : b.top_p, tool_choice: toolChoice != null ? toolChoice : b.tool_choice }),
    ];
}
var ReactModel = (function () {
    function ReactModel(model) {
        this.model = model;
    }
    ReactModel.prototype.generateText = function (_input, options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var rawResponses, totalUsage, _c, _d, onStepFinish, _e, maxSteps, input, doGenerate, currentRes, currentStep, toolCall, stepUsage_1, toolCallResult, choice, e_1, lastChoice, lastMessage, text, messages, stepUsage;
            var _this = this;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        rawResponses = [];
                        totalUsage = { completion_tokens: 0, prompt_tokens: 0, total_tokens: 0 };
                        _c = processInput(_input), _d = _c[0], onStepFinish = _d.onStepFinish, _e = _d.maxSteps, maxSteps = _e === void 0 ? 10 : _e, input = _c[1];
                        doGenerate = function () { return _this.model.doGenerate(input, options); };
                        return [4, doGenerate()];
                    case 1:
                        currentRes = _f.sent();
                        currentStep = 1;
                        currentRes.rawResponse && rawResponses.push(currentRes.rawResponse);
                        toolCall = null;
                        _f.label = 2;
                    case 2:
                        if (!(currentStep < maxSteps && (toolCall = getToolCallFromGenerate(currentRes)) != null)) return [3, 9];
                        stepUsage_1 = createSolidUsage(currentRes.usage);
                        addToUsage(totalUsage, stepUsage_1);
                        _f.label = 3;
                    case 3:
                        _f.trys.push([3, 7, , 8]);
                        return [4, callTool(toolCall)];
                    case 4:
                        toolCallResult = _f.sent();
                        choice = currentRes.choices[0];
                        return [4, (onStepFinish === null || onStepFinish === void 0 ? void 0 : onStepFinish({
                                finishReason: choice.finish_reason,
                                messages: input.messages.slice(),
                                text: choice.message.content,
                                toolCall: toolCall,
                                toolResult: toolCallResult,
                                stepUsage: stepUsage_1,
                                totalUsage: Object.assign({}, totalUsage),
                            }))];
                    case 5:
                        _f.sent();
                        pushNewMessages(input.messages, choice.message, toolCallResult);
                        return [4, doGenerate()];
                    case 6:
                        currentRes = _f.sent();
                        currentRes.rawResponse && rawResponses.push(currentRes.rawResponse);
                        currentStep += 1;
                        return [3, 8];
                    case 7:
                        e_1 = _f.sent();
                        return [2, {
                                text: '',
                                messages: input.messages,
                                usage: totalUsage,
                                error: e_1,
                                rawResponses: rawResponses,
                            }];
                    case 8: return [3, 2];
                    case 9:
                        lastChoice = (_a = currentRes === null || currentRes === void 0 ? void 0 : currentRes.choices) === null || _a === void 0 ? void 0 : _a[0];
                        lastMessage = lastChoice === null || lastChoice === void 0 ? void 0 : lastChoice.message;
                        text = (_b = lastMessage === null || lastMessage === void 0 ? void 0 : lastMessage.content) !== null && _b !== void 0 ? _b : '';
                        messages = lastMessage ? __spreadArray(__spreadArray([], input.messages, true), [lastMessage], false) : input.messages;
                        stepUsage = createSolidUsage(currentRes.usage);
                        addToUsage(totalUsage, stepUsage);
                        return [4, (onStepFinish === null || onStepFinish === void 0 ? void 0 : onStepFinish({
                                finishReason: lastChoice.finish_reason,
                                messages: messages.slice(),
                                text: text,
                                toolCall: getToolCallFromGenerate(currentRes),
                                toolResult: null,
                                stepUsage: stepUsage,
                                totalUsage: Object.assign({}, totalUsage),
                            }))];
                    case 10:
                        _f.sent();
                        return [2, {
                                text: text,
                                messages: messages,
                                usage: totalUsage,
                                rawResponses: rawResponses,
                            }];
                }
            });
        });
    };
    ReactModel.prototype.streamText = function (_input, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var totalUsage, _b, _c, onStepFinish, _d, maxSteps, input, doStream, currentRes, currentStep, readResult, readCurrentStream, _e, assistantMessage, stepUsage_2, toolCall, toolCallResult, e_2, _f, s1_1, s2_1, message_1, usage, messages, _g, s1_2, s2_2, messagePromise, usagePromise, message, finishReason, stepUsage, originStream, _h, s1, s2;
            var _this = this;
            return __generator(this, function (_j) {
                switch (_j.label) {
                    case 0:
                        totalUsage = { completion_tokens: 0, prompt_tokens: 0, total_tokens: 0 };
                        _b = processInput(_input), _c = _b[0], onStepFinish = _c.onStepFinish, _d = _c.maxSteps, maxSteps = _d === void 0 ? 10 : _d, input = _b[1];
                        doStream = function () { return _this.model.doStream(input, options); };
                        return [4, doStream()];
                    case 1:
                        currentRes = _j.sent();
                        currentStep = 1;
                        readResult = null;
                        readCurrentStream = function () {
                            var _a = currentRes.tee(), oldStream = _a[0], newStream = _a[1];
                            currentRes = (0, utils_1.createAsyncIterable)(oldStream);
                            return readFunctionCallStream(newStream);
                        };
                        _j.label = 2;
                    case 2:
                        _e = currentStep < maxSteps;
                        if (!_e) return [3, 4];
                        return [4, readCurrentStream()];
                    case 3:
                        _e = (readResult = _j.sent()) != null;
                        _j.label = 4;
                    case 4:
                        if (!_e) return [3, 11];
                        assistantMessage = readResult.message, stepUsage_2 = readResult.usage;
                        addToUsage(totalUsage, stepUsage_2);
                        toolCall = (_a = assistantMessage.tool_calls) === null || _a === void 0 ? void 0 : _a[0];
                        _j.label = 5;
                    case 5:
                        _j.trys.push([5, 9, , 10]);
                        return [4, callTool(toolCall)];
                    case 6:
                        toolCallResult = _j.sent();
                        return [4, (onStepFinish === null || onStepFinish === void 0 ? void 0 : onStepFinish({
                                finishReason: 'tool_calls',
                                messages: input.messages.slice(),
                                text: assistantMessage.content,
                                toolCall: toolCall,
                                toolResult: toolCallResult,
                                stepUsage: stepUsage_2,
                                totalUsage: Object.assign({}, totalUsage),
                            }))];
                    case 7:
                        _j.sent();
                        pushNewMessages(input.messages, assistantMessage, toolCallResult);
                        return [4, doStream()];
                    case 8:
                        currentRes = _j.sent();
                        return [3, 10];
                    case 9:
                        e_2 = _j.sent();
                        _f = currentRes.tee(), s1_1 = _f[0], s2_1 = _f[1];
                        return [2, {
                                messages: Promise.resolve(input.messages),
                                dataStream: (0, utils_1.createAsyncIterable)(s1_1),
                                textStream: (0, utils_1.createAsyncIterable)(s2_1.pipeThrough(new utils_1.TransformStream({
                                    transform: function (chunk, controller) {
                                        var _a, _b, _c;
                                        var str = (_c = (_b = (_a = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.delta) === null || _c === void 0 ? void 0 : _c.content;
                                        if (typeof str === 'string')
                                            controller.enqueue(str);
                                    },
                                }))),
                                usage: Promise.resolve(totalUsage),
                                error: e_2,
                            }];
                    case 10: return [3, 2];
                    case 11: return [4, readCurrentStream()];
                    case 12:
                        readResult = _j.sent();
                        if (readResult) {
                            message_1 = readResult.message, usage = readResult.usage;
                            addToUsage(totalUsage, usage);
                            messages = __spreadArray(__spreadArray([], input.messages, true), [message_1], false);
                            onStepFinish({
                                messages: messages.slice(),
                                finishReason: 'tool_call',
                                stepUsage: usage,
                                text: message_1.content,
                                toolCall: message_1.tool_calls[0],
                                totalUsage: Object.assign({}, totalUsage),
                            });
                            _g = currentRes.tee(), s1_2 = _g[0], s2_2 = _g[1];
                            return [2, {
                                    messages: Promise.resolve(__spreadArray(__spreadArray([], input.messages, true), [message_1], false)),
                                    dataStream: (0, utils_1.createAsyncIterable)(s1_2),
                                    textStream: (0, utils_1.createAsyncIterable)(s2_2.pipeThrough(new utils_1.TransformStream({
                                        transform: function (chunk, controller) {
                                            var _a, _b, _c;
                                            var str = (_c = (_b = (_a = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.delta) === null || _c === void 0 ? void 0 : _c.content;
                                            if (typeof str === 'string')
                                                controller.enqueue(str);
                                        },
                                    }))),
                                    usage: Promise.resolve(totalUsage),
                                }];
                        }
                        messagePromise = (0, utils_1.createPromise)();
                        usagePromise = (0, utils_1.createPromise)();
                        message = {
                            role: 'assistant',
                            content: '',
                        };
                        finishReason = '';
                        stepUsage = { completion_tokens: 0, prompt_tokens: 0, total_tokens: 0 };
                        originStream = currentRes.pipeThrough(new utils_1.TransformStream({
                            transform: function (chunk, controller) {
                                var _a, _b, _c, _d, _e, _f, _g, _h;
                                var content = (_c = (_b = (_a = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.delta) === null || _c === void 0 ? void 0 : _c.content;
                                if (typeof content === 'string') {
                                    message.content += content;
                                }
                                var reason = (_e = (_d = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _d === void 0 ? void 0 : _d[0]) === null || _e === void 0 ? void 0 : _e.finish_reason;
                                if (reason)
                                    finishReason = reason;
                                if ((_f = chunk === null || chunk === void 0 ? void 0 : chunk.usage) === null || _f === void 0 ? void 0 : _f.completion_tokens)
                                    stepUsage.completion_tokens = chunk.usage.completion_tokens;
                                if ((_g = chunk === null || chunk === void 0 ? void 0 : chunk.usage) === null || _g === void 0 ? void 0 : _g.prompt_tokens)
                                    stepUsage.prompt_tokens = chunk.usage.prompt_tokens;
                                if ((_h = chunk === null || chunk === void 0 ? void 0 : chunk.usage) === null || _h === void 0 ? void 0 : _h.total_tokens)
                                    stepUsage.total_tokens = chunk.usage.total_tokens;
                                controller.enqueue(chunk);
                            },
                            flush: function () {
                                messagePromise.res(__spreadArray(__spreadArray([], input.messages, true), [message], false));
                                addToUsage(totalUsage, stepUsage);
                                usagePromise.res(Object.assign({}, totalUsage));
                                onStepFinish === null || onStepFinish === void 0 ? void 0 : onStepFinish({
                                    messages: __spreadArray(__spreadArray([], input.messages, true), [message], false),
                                    finishReason: finishReason,
                                    text: message.content,
                                    stepUsage: stepUsage,
                                    totalUsage: Object.assign({}, totalUsage),
                                });
                            },
                        }));
                        _h = originStream.tee(), s1 = _h[0], s2 = _h[1];
                        return [2, {
                                messages: messagePromise.promise,
                                dataStream: (0, utils_1.createAsyncIterable)(s1),
                                textStream: (0, utils_1.createAsyncIterable)(s2.pipeThrough(new utils_1.TransformStream({
                                    transform: function (chunk, controller) {
                                        var _a, _b, _c;
                                        var content = (_c = (_b = (_a = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.delta) === null || _c === void 0 ? void 0 : _c.content;
                                        if (typeof content === 'string') {
                                            controller.enqueue(content);
                                        }
                                    },
                                }))),
                                usage: usagePromise.promise,
                            }];
                }
            });
        });
    };
    return ReactModel;
}());
exports.ReactModel = ReactModel;
function getToolCallFromGenerate(output) {
    var _a;
    var choice = (_a = output === null || output === void 0 ? void 0 : output.choices) === null || _a === void 0 ? void 0 : _a[0];
    if (!choice)
        return null;
    var finish_reason = choice.finish_reason, message = choice.message;
    if (finish_reason !== 'tool_calls')
        return null;
    if (!message)
        return null;
    if (!(0, utils_1.isToolCallAssistantMessage)(message))
        return null;
    return message.tool_calls[0];
}
function pushNewMessages(messages, assistantMessage, toolCallResult) {
    messages.push(assistantMessage, {
        role: 'tool',
        tool_call_id: assistantMessage.tool_calls[0].id,
        content: JSON.stringify(toolCallResult),
    });
}
function readFunctionCallStream(stream) {
    var _a, e_3, _b, _c;
    var _d, _e, _f, _g, _h, _j;
    return __awaiter(this, void 0, void 0, function () {
        var stepUsage, aStream, retToolCall, retMessage, _k, aStream_1, aStream_1_1, chunk, choice, finish_reason, delta, toolCall, e_3_1;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0:
                    stepUsage = { completion_tokens: 0, prompt_tokens: 0, total_tokens: 0 };
                    aStream = (0, utils_1.createAsyncIterable)(stream);
                    retToolCall = {
                        id: '',
                        function: {
                            name: '',
                            arguments: '',
                        },
                        type: '',
                    };
                    retMessage = {
                        role: 'assistant',
                        content: '',
                        tool_calls: [retToolCall],
                    };
                    _l.label = 1;
                case 1:
                    _l.trys.push([1, 6, 7, 12]);
                    _k = true, aStream_1 = __asyncValues(aStream);
                    _l.label = 2;
                case 2: return [4, aStream_1.next()];
                case 3:
                    if (!(aStream_1_1 = _l.sent(), _a = aStream_1_1.done, !_a)) return [3, 5];
                    _c = aStream_1_1.value;
                    _k = false;
                    try {
                        chunk = _c;
                        choice = chunk === null || chunk === void 0 ? void 0 : chunk.choices[0];
                        if (!choice)
                            return [2, null];
                        finish_reason = choice.finish_reason, delta = choice.delta;
                        if (finish_reason !== 'tool_calls')
                            return [2, null];
                        if (!delta)
                            return [3, 4];
                        if (delta.content)
                            retMessage.content += delta.content;
                        if (!('tool_calls' in delta))
                            return [3, 4];
                        toolCall = (_d = delta === null || delta === void 0 ? void 0 : delta.tool_calls) === null || _d === void 0 ? void 0 : _d[0];
                        if (toolCall === null || toolCall === void 0 ? void 0 : toolCall.id)
                            retToolCall.id = toolCall.id;
                        if (toolCall === null || toolCall === void 0 ? void 0 : toolCall.type)
                            retToolCall.type = toolCall.type;
                        if ((_e = toolCall === null || toolCall === void 0 ? void 0 : toolCall.function) === null || _e === void 0 ? void 0 : _e.name)
                            retToolCall.function.name = toolCall.function.name;
                        if ((_f = toolCall === null || toolCall === void 0 ? void 0 : toolCall.function) === null || _f === void 0 ? void 0 : _f.arguments)
                            retToolCall.function.arguments += toolCall.function.arguments;
                        if ((_g = chunk === null || chunk === void 0 ? void 0 : chunk.usage) === null || _g === void 0 ? void 0 : _g.completion_tokens)
                            stepUsage.completion_tokens = chunk.usage.completion_tokens;
                        if ((_h = chunk === null || chunk === void 0 ? void 0 : chunk.usage) === null || _h === void 0 ? void 0 : _h.prompt_tokens)
                            stepUsage.prompt_tokens = chunk.usage.prompt_tokens;
                        if ((_j = chunk === null || chunk === void 0 ? void 0 : chunk.usage) === null || _j === void 0 ? void 0 : _j.total_tokens)
                            stepUsage.total_tokens = chunk.usage.total_tokens;
                    }
                    finally {
                        _k = true;
                    }
                    _l.label = 4;
                case 4: return [3, 2];
                case 5: return [3, 12];
                case 6:
                    e_3_1 = _l.sent();
                    e_3 = { error: e_3_1 };
                    return [3, 12];
                case 7:
                    _l.trys.push([7, , 10, 11]);
                    if (!(!_k && !_a && (_b = aStream_1.return))) return [3, 9];
                    return [4, _b.call(aStream_1)];
                case 8:
                    _l.sent();
                    _l.label = 9;
                case 9: return [3, 11];
                case 10:
                    if (e_3) throw e_3.error;
                    return [7];
                case 11: return [7];
                case 12: return [2, {
                        message: retMessage,
                        usage: stepUsage,
                    }];
            }
        });
    });
}
exports.toolMap = new Map();
function callTool(toolCall) {
    return exports.toolMap.get(toolCall.function.name)(JSON.parse(toolCall.function.arguments));
}
function createSolidUsage(usage) {
    var _a, _b, _c;
    return {
        completion_tokens: (_a = usage === null || usage === void 0 ? void 0 : usage.completion_tokens) !== null && _a !== void 0 ? _a : 0,
        prompt_tokens: (_b = usage === null || usage === void 0 ? void 0 : usage.prompt_tokens) !== null && _b !== void 0 ? _b : 0,
        total_tokens: (_c = usage === null || usage === void 0 ? void 0 : usage.total_tokens) !== null && _c !== void 0 ? _c : 0,
    };
}
function addToUsage(targetUsage, sourceUsage) {
    targetUsage.completion_tokens += sourceUsage.completion_tokens;
    targetUsage.prompt_tokens += sourceUsage.prompt_tokens;
    targetUsage.total_tokens += sourceUsage.total_tokens;
}
